Index: MANIFEST
===================================================================
RCS file: /cvsroot/xmltv/xmltv/MANIFEST,v
retrieving revision 1.100
diff -u -B -b -w -r1.100 MANIFEST
--- MANIFEST	10 Apr 2004 22:00:29 -0000	1.100
+++ MANIFEST	2 Jun 2004 09:36:59 -0000
@@ -4,8 +4,14 @@
 Uninstall.pm
 README
 lib/Ask.pm
-lib/AskTerm.pm
-lib/AskTk.pm
+lib/Ask/GDialog.pm
+lib/Ask/Term.pm
+lib/Ask/Tk.pm
+lib/ProgressBar.pm
+lib/ProgressBar/GDialog.pm
+lib/ProgressBar/None.pm
+lib/ProgressBar/Term.pm
+lib/ProgressBar/Tk.pm
 lib/XMLTV.pm.PL
 lib/XMLTV.pm.in
 lib/TZ.pm
Index: Makefile.PL
===================================================================
RCS file: /cvsroot/xmltv/xmltv/Makefile.PL,v
retrieving revision 1.199
diff -u -B -b -w -r1.199 Makefile.PL
--- Makefile.PL	13 Apr 2004 22:20:54 -0000	1.199
+++ Makefile.PL	2 Jun 2004 09:37:00 -0000
@@ -107,9 +107,16 @@
      'lib/Usage.pm'               => '$(INST_LIBDIR)/XMLTV/Usage.pm',
      'lib/Date.pm'                => '$(INST_LIBDIR)/XMLTV/Date.pm',
      'lib/Version.pm'             => '$(INST_LIBDIR)/XMLTV/Version.pm',
+     'lib/GUI.pm'                 => '$(INST_LIBDIR)/XMLTV/GUI.pm',
      'lib/Ask.pm'                 => '$(INST_LIBDIR)/XMLTV/Ask.pm',
-     'lib/AskTk.pm'               => '$(INST_LIBDIR)/XMLTV/AskTk.pm',
-     'lib/AskTerm.pm'             => '$(INST_LIBDIR)/XMLTV/AskTerm.pm',
+     'lib/Ask/GDialog.pm'         => '$(INST_LIBDIR)/XMLTV/Ask/GDialog.pm',
+     'lib/Ask/Term.pm'            => '$(INST_LIBDIR)/XMLTV/Ask/Term.pm',
+     'lib/Ask/Tk.pm'              => '$(INST_LIBDIR)/XMLTV/Ask/Tk.pm',
+     'lib/ProgressBar.pm'         => '$(INST_LIBDIR)/XMLTV/ProgressBar.pm',
+     'lib/ProgressBar/GDialog.pm' => '$(INST_LIBDIR)/XMLTV/ProgressBar/GDialog.pm',
+     'lib/ProgressBar/None.pm'    => '$(INST_LIBDIR)/XMLTV/ProgressBar/None.pm',
+     'lib/ProgressBar/Term.pm'    => '$(INST_LIBDIR)/XMLTV/ProgressBar/Term.pm',
+     'lib/ProgressBar/Tk.pm'      => '$(INST_LIBDIR)/XMLTV/ProgressBar/Tk.pm',
      'lib/Summarize.pm'           => '$(INST_LIBDIR)/XMLTV/Summarize.pm',
      'lib/IMDB.pm'                => '$(INST_LIBDIR)/XMLTV/IMDB.pm',
      'lib/Gunzip.pm'              => '$(INST_LIBDIR)/XMLTV/Gunzip.pm',
@@ -179,8 +186,8 @@
     *ask = sub { print "$_[0] yes\n"; 1 };
 }
 else {
-    do 'lib/AskTerm.pm' or die 'could not load lib/AskTerm.pm, aborting';
-    *ask = \&XMLTV::AskTerm::askBooleanQuestion;
+    do 'lib/Ask/Term.pm' or die 'could not load lib/Ask/Term.pm, aborting';
+    *ask = \&XMLTV::Ask::Term::ask_boolean;
 }
 
 # Weird shit happens when you change things like PREFIX without
@@ -194,7 +201,7 @@
 
 END
   ;
-    if (! ask("Do you wish to continue anyway?", 0)) {
+    if (! ask(0, "Do you wish to continue anyway?", 0)) {
 	exit(1);
     }
 }
@@ -223,6 +230,11 @@
        prereqs => { 'HTML::Entities' => 1.27 } ,
      },
 
+     { name => 'tv_grab_uk_bleb',
+       blurb => 'Fast alternative grabber for the UK',
+       exes => [ 'grab/uk_bleb/tv_grab_uk_bleb' ],
+       prereqs => { 'IO::Scalar' => 0, 'Archive::Zip' => 0 } },
+     
      { name => 'tv_grab_it',
        blurb => 'Grabber for Italy',
        exes => [ 'grab/it/tv_grab_it' ],
@@ -436,7 +448,7 @@
 		  "\n");
 }
 print STDERR "\n";
-if (not ask('Do you want to proceed with this configuration?', 1)) {
+if (not ask(0,'Do you want to proceed with this configuration?', 1)) {
     # Need to set {install} for each component by prompting.
     foreach my $info (@opt_components) {
 	my $missing = $info->{missing};
@@ -466,7 +478,7 @@
 	else { die }
 	
 	$info->{install} =
-	  ask($msg, not $missing);
+	  ask(0, $msg, not $missing);
     }
 }
 
Index: grab/Config_file.pm
===================================================================
RCS file: /cvsroot/xmltv/xmltv/grab/Config_file.pm,v
retrieving revision 1.6
diff -u -B -b -w -r1.6 Config_file.pm
--- grab/Config_file.pm	3 Jan 2004 14:52:53 -0000	1.6
+++ grab/Config_file.pm	2 Jun 2004 09:37:00 -0000
@@ -41,7 +41,7 @@
 sub check_no_overwrite( $ ) {
     my $f = shift;
     if (-e $f) {
-	if (not askBooleanQuestion
+	if (not ask_boolean
 	    ("The configuration file $f already exists.  There is
 currently no support for altering an existing configuration: you have
 to reconfigure from scratch.
Index: grab/Grab_XML.pm
===================================================================
RCS file: /cvsroot/xmltv/xmltv/grab/Grab_XML.pm,v
retrieving revision 1.11
diff -u -B -b -w -r1.11 Grab_XML.pm
--- grab/Grab_XML.pm	3 Jan 2004 14:52:53 -0000	1.11
+++ grab/Grab_XML.pm	2 Jun 2004 09:37:00 -0000
@@ -6,6 +6,8 @@
 use XMLTV;
 use XMLTV::Usage;
 use XMLTV::Memoize;
+use XMLTV::ProgressBar;
+use XMLTV::Ask;
 use XMLTV::TZ qw(parse_local_date);
 use XMLTV::Get_nice qw();
 use XMLTV::Date;
@@ -24,9 +26,6 @@
     }
 }
 
-# Use Term::ProgressBar if installed.
-use constant Have_bar => eval { require Term::ProgressBar; 1 };
-
 =pod
 
 =head1 NAME
@@ -74,8 +73,10 @@
 date can be downloaded.  This method is abstract, you must override
 it.
 
+Arguments: the command line options for --config-file and --quiet.
+
 =cut
-sub urls_by_date( $ ) {
+sub urls_by_date( $$$ ) {
     my $pkg = shift;
     die 'abstract class method: override in subclass';
 }
@@ -95,6 +96,21 @@
     return shift; # leave unchanged
 }
 
+=pod
+
+=item XMLTV::Grab_XML->configure()
+
+Configure the grabber if needed.
+
+=cut
+sub configure( $ ) {
+    # Default is no config file
+    print STDERR "no configuration necessary\n";
+    exit();
+}
+
+=pod
+
 =item XMLTV::Grab_XML->nextday(day)
 
 Bump a YYYYMMDD date by one.  You probably shouldnE<39>t override this.
@@ -167,6 +183,7 @@
     my ($opt_days,
 	$opt_help,
 	$opt_output,
+    $opt_gui,
 	$opt_offset,
 	$opt_quiet,
 	$opt_configure,
@@ -177,6 +194,7 @@
     GetOptions('days=i'        => \$opt_days,
 	       'help'          => \$opt_help,
 	       'output=s'      => \$opt_output,
+           'gui:s'         => \$opt_gui,
 	       'offset=i'      => \$opt_offset,
 	       'quiet'         => \$opt_quiet,
 	       'configure'     => \$opt_configure,
@@ -187,13 +205,13 @@
       if (defined $opt_days && $opt_days < 0);
     usage(1, $pkg->usage_msg()) if $opt_help;
     usage(0, $pkg->usage_msg()) if @ARGV;
+    
+    XMLTV::Ask::init($opt_gui);
+    XMLTV::ProgressBar::init($opt_gui);
+    
     if ($opt_configure) {
-	print STDERR "no configuration necessary\n";
-	exit();
-    }
-    for ($opt_config_file) {
-	warn "this grabber has no configuration, so ignoring --config-file $_\n"
-	  if defined;
+        $pkg->configure( $opt_config_file, $opt_quiet );
+        exit;
     }
 
     # Need to call parse_local_date() before any resetting of
@@ -204,7 +222,7 @@
     $pkg->date_init();
     my $today = UnixDate($now, '%Q');
 
-    my %urls = $pkg->urls_by_date();
+    my %urls = $pkg->urls_by_date( $opt_config_file, $opt_quiet );
     t 'URLs by date: ' . d \%urls;
 
     my @to_get;
@@ -227,8 +245,8 @@
 	warn "couldn't get any listings from the site for today or later\n";
     }
 
-    my $bar = new Term::ProgressBar('downloading listings', scalar @to_get)
-      if Have_bar && not $opt_quiet;
+    my $bar = new XMLTV::ProgressBar('downloading listings', scalar @to_get)
+      if not $opt_quiet;
     my @listingses;
     foreach my $url (@to_get) {
 	my $xml;
@@ -263,8 +281,9 @@
 	}
 
 	push @listingses, XMLTV::parse($xml);
-	update $bar if Have_bar && not $opt_quiet;
+	update $bar if not $opt_quiet;
     }
+    $bar->finish();
     my %w_args = ();
     if (defined $opt_output) {
 	my $fh = new IO::File ">$opt_output";
@@ -290,6 +309,45 @@
 
 =pod
 
+=item XMLTV::Grab_XML->remove_early_stop_times()
+
+Checks each stop time and removes it if it's before the start time.
+
+Argument: the XML to correct
+Returns: the corrected XML
+
+=cut
+
+my $warned_bad_stop_time = 0;
+sub remove_early_stop_times( $$ ) {
+    my $pkg = shift;
+    my @lines = split /\n/, shift;
+    foreach (@lines) {
+	if (/<programme/) {
+	    # First change to numeric timezones.
+	    s{(start|stop)="(\d+) ([A-Z]+)"}
+	    {qq'$1="$2 ' . tz_to_num($3) . '"'}eg;
+	    
+	    # Now remove stop times before start.  Only worry about
+	    # cases where the timezone is the same - we hope the
+	    # upstream data will be fixed by the next TZ changeover.
+	    #
+	    /start="(\d+) (\S+)"/ or next;
+	    my ($start, $tz) = ($1, $2);
+	    /stop="(\d+) \Q$tz\E"/ or next;
+	    my $stop = $1;
+	    if ($stop lt $start) {
+		warn "removing stop time before start time: $_"
+		  unless $warned_bad_stop_time++;
+		s/stop="[^""]+"\s*// or die;
+	    }
+	}
+    }
+    return join('', @lines);
+}
+
+=pod
+
 =back
 
 =head1 AUTHOR
Index: grab/de_tvtoday/tv_grab_de_tvtoday.in
===================================================================
RCS file: /cvsroot/xmltv/xmltv/grab/de_tvtoday/tv_grab_de_tvtoday.in,v
retrieving revision 1.9
diff -u -B -b -w -r1.9 tv_grab_de_tvtoday.in
--- grab/de_tvtoday/tv_grab_de_tvtoday.in	10 Apr 2004 22:00:30 -0000	1.9
+++ grab/de_tvtoday/tv_grab_de_tvtoday.in	2 Jun 2004 09:37:01 -0000
@@ -11,7 +11,7 @@
 
 tv_grab_de_tvtoday --help
 
-tv_grab_de_tvtoday [--config-file FILE] --configure
+tv_grab_de_tvtoday [--config-file FILE] --configure [--gui OPTION]
 
 tv_grab_de_tvtoday [--config-file FILE] [--output FILE] 
                    [--days N] [--offset N]
@@ -37,6 +37,11 @@
 default is B<~/.xmltv/tv_grab_de_tvtoday.conf>.  This is the file 
 written by B<--configure> and read when grabbing.
 
+B<--gui OPTION> Use this option to enable a graphical interface to be used.
+OPTION may be 'Tk' or 'GDialog', or left blank for the best available choice.
+Additional allowed values of OPTION are 'Term' for normal terminal output
+(default) and 'TermNoProgressBar' to disable the use of Term::ProgressBar.
+
 B<--output FILE> write to FILE rather than standard output.
 
 B<--days N> grab N days.  The default is seven.
@@ -88,6 +93,8 @@
 use URI::Escape;
 use XMLTV;
 use XMLTV::Ask;
+use XMLTV::ProgressBar;
+use XMLTV::Europe_TZ;
 use XMLTV::DST;
 use XMLTV::Config_file;
 use XMLTV::Mode;
@@ -95,7 +102,7 @@
 use XMLTV::Memoize;
 use XMLTV::Usage <<END
 $0: get German television listings from www.tvtoday.de in XMLTV format
-To configure: $0 --configure [--config-file FILE]
+To configure: $0 --configure [--config-file FILE] [--gui OPTION]
 To grab data: $0 [--config-file FILE] [--output FILE] 
                  [--days N] [--offset N]
                  [--quiet] [--slow] [--nosqueezeout]
@@ -137,9 +144,6 @@
 my $debug = 0;
 $XMLTV::Get_nice::Delay = 0 if($debug);
 
-#-- Use Term::ProgressBar if installed.
-use constant Have_bar => eval { require Term::ProgressBar; 1; };
-
 #-- attributes of xmltv root element
 my $head = { 
     'source-data-url'      => 'http://www.tvtoday.de/tv/programm/programm.php',
@@ -157,6 +161,7 @@
 
 my $opt_configure;
 my $opt_config_file;
+my $opt_gui;
 my $opt_output;
 my $opt_days;
 my $opt_offset = 0;
@@ -170,6 +175,7 @@
 GetOptions(
     'configure'      => \$opt_configure,
     'config-file=s'  => \$opt_config_file,
+    'gui:s'          => \$opt_gui,
     'output=s'       => \$opt_output,
     'days=i'         => \$opt_days,
     'offset=i'       => \$opt_offset,
@@ -183,6 +189,9 @@
 
 usage(1) if $opt_help;
 
+XMLTV::Ask::init($opt_gui);
+XMLTV::ProgressBar::init($opt_gui);
+
 #-- make sure offset+days arguments are within range
 die "offset mustn't be larger than six"
   if($opt_offset > 6);
@@ -285,7 +294,7 @@
     my @chs = sort keys %channels;
     my @names = map { $channels{$_} } @chs;
     my @qs = map { "add channel $_?" } @names;
-    my @want = askManyBooleanQuestions(1, @qs);
+    my @want = ask_many_boolean(1, @qs);
 
     foreach (@chs) {
 	my $w = shift @want;
@@ -353,16 +362,18 @@
 
 #-- write out <programme> tags
 my $numdays = $opt_days + $opt_offset - 1;
-my $bar = new Term::ProgressBar('grabbing', scalar(@requests) * $opt_days)
-  if Have_bar && not $opt_quiet;
+  
+my $bar = new XMLTV::ProgressBar('grabbing', scalar(@requests) * $opt_days)
+  if not $opt_quiet;
 
 foreach my $channel (@requests) {
     for (my $day = $opt_offset; $day <= $numdays; $day ++) {
 	grab_data($channel, $day, $day == $numdays);
-	update $bar if Have_bar && not $opt_quiet;
+	update $bar if not $opt_quiet;
     }
 }
 
+$bar->finish();
 
 #-- hey, looks like we've finished ...
 $writer->end();
@@ -387,7 +398,7 @@
     my $lday = shift @_;    #- true: last day to grab in row
 
     print STDERR "grabbing channel $channels{$ch} (offset=$offset) ...\n"
-      unless(Have_bar || $opt_quiet);
+      unless($opt_quiet);
 	    
     #- we got to send ztag=8 to retrieve data for yesterday
     $offset = 8 if($offset < 0);
Index: grab/dk/tv_grab_dk
===================================================================
RCS file: /cvsroot/xmltv/xmltv/grab/dk/tv_grab_dk,v
retrieving revision 1.14
diff -u -B -b -w -r1.14 tv_grab_dk
--- grab/dk/tv_grab_dk	10 Apr 2004 22:00:30 -0000	1.14
+++ grab/dk/tv_grab_dk	2 Jun 2004 09:37:02 -0000
@@ -10,7 +10,7 @@
 
 tv_grab_dk --help
 
-tv_grab_dk [--config-file FILE] --configure
+tv_grab_dk [--config-file FILE] --configure [--gui OPTION]
 
 tv_grab_dk [--config-file FILE] [--output FILE] [--days N]
            [--offset N] [--quiet]
@@ -32,6 +32,11 @@
 default is B<~/.xmltv/tv_grab_dk.conf>.  This is the file written by
 B<--configure> and read when grabbing.
 
+B<--gui OPTION> Use this option to enable a graphical interface to be used.
+OPTION may be 'Tk' or 'GDialog', or left blank for the best available choice.
+Additional allowed values of OPTION are 'Term' for normal terminal output
+(default) and 'TermNoProgressBar' to disable the use of Term::ProgressBar.
+
 B<--output FILE> write to FILE rather than standard output.
 
 B<--days N> grab N days.  The default is one week.
@@ -79,6 +84,7 @@
 
 use XMLTV;
 use XMLTV::Memoize;
+use XMLTV::ProgressBar;
 use XMLTV::Ask;
 use XMLTV::Mode;
 use XMLTV::Config_file;
@@ -109,8 +115,7 @@
     }
 }
 
-# Use Term::ProgressBar if installed.
-use constant Have_bar => eval { require Term::ProgressBar; 1 };
+
 
 # Whether zero-length programmes should be included in the output.
 my $WRITE_ZERO_LENGTH = 0;
@@ -130,8 +135,8 @@
 # Get options
 XMLTV::Memoize::check_argv('XMLTV::Get_nice::get_nice_aux');
 my ($opt_days, $opt_offset, $opt_help, $opt_output,
-    $opt_configure, $opt_config_file, $opt_quiet,
-    $opt_list_channels);
+    $opt_configure, $opt_config_file, $opt_gui,
+    $opt_quiet, $opt_list_channels);
 $opt_days   = 7; # default
 $opt_offset = 0; # default
 GetOptions('days=i'        => \$opt_days,
@@ -139,6 +144,7 @@
 	   'help'          => \$opt_help,
 	   'configure'     => \$opt_configure,
 	   'config-file=s' => \$opt_config_file,
+           'gui:s'         => \$opt_gui,
 	   'output=s'      => \$opt_output,
 	   'quiet'         => \$opt_quiet,
 	   'list-channels' => \$opt_list_channels,
@@ -148,6 +154,9 @@
   if (defined $opt_days && $opt_days < 0);
 usage(1) if $opt_help;
 
+XMLTV::Ask::init($opt_gui);
+XMLTV::ProgressBar::init($opt_gui);
+
 my $mode = XMLTV::Mode::mode('grab', # default
 			     $opt_configure => 'configure',
 			     $opt_list_channels => 'list-channels',
@@ -161,17 +170,17 @@
     XMLTV::Config_file::check_no_overwrite($config_file);
     open(CONF, ">$config_file") or die "cannot write to $config_file: $!";
     # find list of available channels
-    my $bar = new Term::ProgressBar('getting list of channels', 1)
-      if Have_bar && not $opt_quiet;
+    my $bar = new XMLTV::ProgressBar('getting list of channels', 1)
+      if not $opt_quiet;
     my %channels = get_channels();
     die 'no channels could be found' if (scalar(keys(%channels)) == 0);
-    update $bar if Have_bar && not $opt_quiet;
+    update $bar if not $opt_quiet;
 
     # Ask about each channel.
     my @chs = sort keys %channels;
     my @names = map { $channels{$_} } @chs;
     my @qs = map { "add channel $_?" } @names;
-    my @want = askManyBooleanQuestions(1, @qs);
+    my @want = ask_many_boolean(1, @qs);
     foreach (@chs) {
 	my $w = shift @want;
 	warn("cannot read input, stopping channel questions"), last
@@ -185,6 +194,8 @@
 	# TODO don't store display-name in config file.
     }
 
+    $bar->finish() if not $opt_quiet;
+    
     close CONF or warn "cannot close $config_file: $!";
     say("Finished configuration.");
 
@@ -219,11 +230,11 @@
    });
 
 if ($opt_list_channels) {
-    my $bar = new Term::ProgressBar('getting list of channels', 1)
-      if Have_bar && not $opt_quiet;
+    my $bar = new XMLTV::ProgressBar('getting list of channels', 1)
+      if not $opt_quiet;
     my %channels = get_channels();
     die 'no channels could be found' if (scalar(keys(%channels)) == 0);
-    update $bar if Have_bar && not $opt_quiet;
+    update $bar if not $opt_quiet;
 
     foreach my $ch_did (sort(keys %channels)) {
 	my $ch_name = $channels{$ch_did};
@@ -231,6 +242,7 @@
 	$writer->write_channel({ id => $ch_xid,
 				 'display-name' => [ [ $ch_name ] ] });
     }
+    $bar->finish() if not $opt_quiet;
     $writer->end();
     exit();
 }
@@ -290,16 +302,17 @@
 
 my %warned_ch_name; # suppress duplicate warnings
 
-my $bar = new Term::ProgressBar('fetching data', scalar @to_get)
-  if Have_bar && not $opt_quiet;
+my $bar = new XMLTV::ProgressBar('fetching data', scalar @to_get)
+  if not $opt_quiet;
 my @to_get_detailed;
 my $num_detailed = 0;
 foreach (@to_get) {
     my ($url, $date, $ch_xmltv_id, $ch_tvgids_id) = @$_;
     t "going to get $ch_xmltv_id for $date";
     process_listings_page($writer, $ch_xmltv_id, $url, $date);
-    update $bar if Have_bar && not $opt_quiet;
+    update $bar if not $opt_quiet;
 }
+$bar->finish() if not $opt_quiet;
 $writer->end();
 
 ######################################################################
Index: grab/es/tv_grab_es
===================================================================
RCS file: /cvsroot/xmltv/xmltv/grab/es/tv_grab_es,v
retrieving revision 1.26
diff -u -B -b -w -r1.26 tv_grab_es
--- grab/es/tv_grab_es	10 Apr 2004 22:00:30 -0000	1.26
+++ grab/es/tv_grab_es	2 Jun 2004 09:37:03 -0000
@@ -10,7 +10,7 @@
 
 tv_grab_es --help
 
-tv_grab_es [--config-file FILE] --configure
+tv_grab_es [--config-file FILE] --configure [--gui OPTION]
 
 tv_grab_es [--config-file FILE] [--output FILE] [--days N]
            [--offset N] [--quiet]
@@ -38,6 +38,11 @@
 default is B<~/.xmltv/tv_grab_es.conf>.  This is the file written by
 B<--configure> and read when grabbing.
 
+B<--gui OPTION> Use this option to enable a graphical interface to be used.
+OPTION may be 'Tk' or 'GDialog', or left blank for the best available choice.
+Additional allowed values of OPTION are 'Term' for normal terminal output
+(default) and 'TermNoProgressBar' to disable the use of XMLTV::ProgressBar.
+
 B<--output FILE> write to FILE rather than standard output.
 
 B<--days N> grab N days.  The default is 3.
@@ -97,6 +102,7 @@
 
 use XMLTV;
 use XMLTV::Memoize;
+use XMLTV::ProgressBar;
 use XMLTV::Ask;
 use XMLTV::Config_file;
 use XMLTV::DST;
@@ -113,9 +119,6 @@
 END
   ;
 
-# Use Term::ProgressBar if installed.
-use constant Have_bar => eval { require Term::ProgressBar; 1 };
-
 # Attributes of the root element in output.
 my $HEAD = { 'source-info-url'     => 'http://www.elpais.es/parrillatv/portada.html',
 	     'source-data-url'     => "http://www.elpais.es/parrillatv/resultados.html",
@@ -138,8 +141,8 @@
 # Get options, including undocumented --cache option.
 XMLTV::Memoize::check_argv('XMLTV::Get_nice::get_nice_aux');
 my ($opt_days, $opt_offset, $opt_help, $opt_output,
-    $opt_configure, $opt_config_file, $opt_quiet,
-    $opt_list_channels);
+    $opt_configure, $opt_config_file, $opt_gui,
+    $opt_quiet, $opt_list_channels);
 $opt_days  = 3; # default
 $opt_offset = 0; # default
 $opt_quiet  = 0; # default
@@ -148,6 +151,7 @@
 	   'help'          => \$opt_help,
 	   'configure'     => \$opt_configure,
 	   'config-file=s' => \$opt_config_file,
+           'gui:s'         => \$opt_gui,
 	   'output=s'      => \$opt_output,
 	   'quiet'         => \$opt_quiet,
 	   'list-channels' => \$opt_list_channels
@@ -157,6 +161,9 @@
   if (defined $opt_days && $opt_days < 0);
 usage(1) if $opt_help;
 
+XMLTV::Ask::init($opt_gui);
+XMLTV::ProgressBar::init($opt_gui);
+
 my $mode = XMLTV::Mode::mode('grab', # default
 			     $opt_configure => 'configure',
 			     $opt_list_channels => 'list-channels',
@@ -192,7 +199,7 @@
     my @chs = sort keys %channels;
     my @names = map { $channels{$_} } @chs;
     my @qs = map { "add channel $_?" } @names;
-    my @want = askManyBooleanQuestions(1, @qs);
+    my @want = ask_many_boolean(1, @qs);
     foreach (@chs) {
 	my $w = shift @want;
 	warn("cannot read input, stopping channel questions"), last
@@ -285,14 +292,15 @@
 # This progress bar is for both downloading and parsing.  Maybe
 # they could be separate.
 #
-my $bar = new Term::ProgressBar('getting listings', scalar @to_get)
-  if Have_bar && not $opt_quiet;
+my $bar = new XMLTV::ProgressBar('getting listings', scalar @to_get)
+  if not $opt_quiet;
 foreach (@to_get) {
 	foreach (process_table($_->[0], $_->[1], $_->[2])) {
 		$writer->write_programme($_);
 	}
-	update $bar if Have_bar && not $opt_quiet;
+	update $bar if not $opt_quiet;
 }
+$bar->finish() if not $opt_quiet;
 $writer->end();
 
 ######################################################################
@@ -491,8 +499,8 @@
 
 # get channel listing
 sub get_channels {
-    my $bar = new Term::ProgressBar('getting list of channels', 1)
-	if Have_bar && not $opt_quiet;
+    my $bar = new XMLTV::ProgressBar('getting list of channels', 1)
+	if not $opt_quiet;
     my %channels;
     my $url="http://www.elpais.es/parrillatv/portada.html";
     t $url;
@@ -527,7 +535,8 @@
 	}
     }
     die "no channels could be found" if not keys %channels;
-    update $bar if Have_bar && not $opt_quiet;
+    update $bar if not $opt_quiet;
+    $bar->finish() if not $opt_quiet;
     return %channels;
 }
 
Index: grab/es_digital/tv_grab_es_digital
===================================================================
RCS file: /cvsroot/xmltv/xmltv/grab/es_digital/tv_grab_es_digital,v
retrieving revision 1.3
diff -u -B -b -w -r1.3 tv_grab_es_digital
--- grab/es_digital/tv_grab_es_digital	10 Apr 2004 22:00:30 -0000	1.3
+++ grab/es_digital/tv_grab_es_digital	2 Jun 2004 09:37:03 -0000
@@ -10,7 +10,7 @@
 
 tv_grab_es_digital --help
 
-tv_grab_es_digital [--config-file FILE] --configure
+tv_grab_es_digital [--config-file FILE] --configure [--gui OPTION]
 
 tv_grab_es_digital [--config-file FILE] [--output FILE] [--days N]
            [--offset N] [--quiet]
@@ -36,6 +36,11 @@
 default is B<~/.xmltv/tv_grab_es_digital.conf>.  This is the file written by
 B<--configure> and read when grabbing.
 
+B<--gui OPTION> Use this option to enable a graphical interface to be used.
+OPTION may be 'Tk' or 'GDialog', or left blank for the best available choice.
+Additional allowed values of OPTION are 'Term' for normal terminal output
+(default) and 'TermNoProgressBar' to disable the use of Term::ProgressBar.
+
 B<--output FILE> write to FILE rather than standard output.
 
 B<--days N> grab N days.  The default is 3.
@@ -93,6 +98,7 @@
 
 use XMLTV;
 use XMLTV::Memoize;
+use XMLTV::ProgressBar;
 use XMLTV::Ask;
 use XMLTV::Config_file;
 use XMLTV::DST;
@@ -108,9 +114,6 @@
 END
   ;
 
-# Use Term::ProgressBar if installed.
-use constant Have_bar => eval { require Term::ProgressBar; 1 };
-
 # Attributes of the root element in output.
 my $HEAD = { 'source-info-url'     => 'http://http://www.plus.es/codigo/television/rejilla_dia_dia.asp',
 	     'source-data-url'     => "http://http://www.plus.es/codigo/television/rejilla_dia_dia.asp",
@@ -133,8 +136,8 @@
 # Get options, including undocumented --cache option.
 XMLTV::Memoize::check_argv('XMLTV::Get_nice::get_nice_aux');
 my ($opt_days, $opt_offset, $opt_help, $opt_output,
-    $opt_configure, $opt_config_file, $opt_quiet,
-    $opt_list_channels);
+    $opt_configure, $opt_config_file, $opt_gui,
+    $opt_quiet, $opt_list_channels);
 $opt_days  = 3; # default
 $opt_offset = 0; # default
 $opt_quiet  = 0; # default
@@ -143,6 +146,7 @@
 	   'help'          => \$opt_help,
 	   'configure'     => \$opt_configure,
 	   'config-file=s' => \$opt_config_file,
+       'gui:s'         => \$opt_gui,
 	   'output=s'      => \$opt_output,
 	   'quiet'         => \$opt_quiet,
 	   'list-channels' => \$opt_list_channels
@@ -152,6 +156,9 @@
   if (defined $opt_days && $opt_days < 0);
 usage(1) if $opt_help;
 
+XMLTV::Ask::init($opt_gui);
+XMLTV::ProgressBar::init($opt_gui);
+
 my $mode = XMLTV::Mode::mode('grab', # default
 			     $opt_configure => 'configure',
 			     $opt_list_channels => 'list-channels',
@@ -187,7 +194,7 @@
     my @chs = sort keys %channels;
     my @names = map { $channels{$_} } @chs;
     my @qs = map { "add channel $_?" } @names;
-    my @want = askManyBooleanQuestions(1, @qs);
+    my @want = ask_many_boolean(1, @qs);
     foreach (@chs) {
 	my $w = shift @want;
 	warn("cannot read input, stopping channel questions"), last
@@ -278,16 +285,18 @@
 # This progress bar is for both downloading and parsing.  Maybe
 # they could be separate.
 #
-my $bar = new Term::ProgressBar('getting listings', scalar @to_get)
-  if Have_bar && not $opt_quiet;
+my $bar = new XMLTV::ProgressBar('getting listings', scalar @to_get)
+  if not $opt_quiet;
 foreach (@to_get) {
 	foreach (process_table($_->[0], $_->[1], $_->[2])) {
 		$writer->write_programme($_);
 	}
-	update $bar if Have_bar && not $opt_quiet;
+	update $bar if not $opt_quiet;
 }
 $writer->end();
 
+$bar->finish();
+
 ######################################################################
 # subroutine definitions
 
@@ -480,8 +489,8 @@
 
 # get channel listing
 sub get_channels {
-    my $bar = new Term::ProgressBar('getting list of channels', 1)
-	if Have_bar && not $opt_quiet;
+    my $bar = new XMLTV::ProgressBar('getting list of channels', 1)
+	if not $opt_quiet;
     my %channels;
     my $url="http://www.plus.es/codigo/television/plataformas/digitalplus/portada2.asp";
     t $url;
@@ -517,7 +526,8 @@
 	}
     }
     die "no channels could be found" if not keys %channels;
-    update $bar if Have_bar && not $opt_quiet;
+    update $bar if not $opt_quiet;
+    $bar->finish();
     return %channels;
 }
 
Index: grab/fi/tv_grab_fi
===================================================================
RCS file: /cvsroot/xmltv/xmltv/grab/fi/tv_grab_fi,v
retrieving revision 1.31
diff -u -B -b -w -r1.31 tv_grab_fi
--- grab/fi/tv_grab_fi	10 Apr 2004 22:00:31 -0000	1.31
+++ grab/fi/tv_grab_fi	2 Jun 2004 09:37:03 -0000
@@ -10,7 +10,7 @@
 
 tv_grab_fi --help
 
-tv_grab_fi [--config-file FILE] --configure
+tv_grab_fi [--config-file FILE] --configure [--gui OPTION]
 
 tv_grab_fi [--config-file FILE] [--output FILE] [--days N]
            [--offset N] [--quiet]
@@ -35,6 +35,11 @@
 default is B<~/.xmltv/tv_grab_fi.conf>.  This is the file written by
 B<--configure> and read when grabbing.
 
+B<--gui OPTION> Use this option to enable a graphical interface to be used.
+OPTION may be 'Tk' or 'GDialog', or left blank for the best available choice.
+Additional allowed values of OPTION are 'Term' for normal terminal output
+(default) and 'TermNoProgressBar' to disable the use of XMLTV::ProgressBar.
+
 B<--output FILE> write to FILE rather than standard output.
 
 B<--days N> grab N days.  The default is ten.
@@ -77,6 +82,7 @@
 
 use XMLTV;
 use XMLTV::Memoize;
+use XMLTV::ProgressBar;
 use XMLTV::Ask;
 use XMLTV::Config_file;
 use XMLTV::DST;
@@ -93,9 +99,6 @@
 END
   ;
 
-# Use Term::ProgressBar if installed.
-use constant Have_bar => eval { require Term::ProgressBar; 1 };
-
 # Attributes of the root element in output.
 my $HEAD = { 'source-info-url'     => 'http://www.katso.fi/',
 	     'source-data-url'     => "http://www.katso.fi/tvopas",
@@ -122,8 +125,8 @@
 # Get options, including undocumented --cache option.
 XMLTV::Memoize::check_argv('XMLTV::Get_nice::get_nice_aux');
 my ($opt_days, $opt_offset, $opt_help, $opt_output,
-    $opt_configure, $opt_config_file, $opt_quiet,
-    $opt_list_channels);
+    $opt_configure, $opt_config_file, $opt_gui,
+    $opt_quiet, $opt_list_channels);
 $opt_days  = 10; # default
 $opt_offset = 0; # default
 $opt_quiet  = 0; # default
@@ -132,6 +135,7 @@
 	   'help'          => \$opt_help,
 	   'configure'     => \$opt_configure,
 	   'config-file=s' => \$opt_config_file,
+           'gui:s'         => \$opt_gui,
 	   'output=s'      => \$opt_output,
 	   'quiet'         => \$opt_quiet,
 	   'list-channels' => \$opt_list_channels,
@@ -141,6 +145,9 @@
   if (defined $opt_days && $opt_days < 0);
 usage(1) if $opt_help;
 
+XMLTV::Ask::init($opt_gui);
+XMLTV::ProgressBar::init($opt_gui);
+
 my $mode = XMLTV::Mode::mode('grab', # default
 			     $opt_configure => 'configure',
 			     $opt_list_channels => 'list-channels',
@@ -176,7 +183,7 @@
     my @chs = sort keys %channels;
     my @names = map { $channels{$_} } @chs;
     my @qs = map { "add channel $_?" } @names;
-    my @want = askManyBooleanQuestions(1, @qs);
+    my @want = ask_many_boolean(1, @qs);
     foreach (@chs) {
 	my $w = shift @want;
 	warn("cannot read input, stopping channel questions"), last
@@ -266,14 +273,15 @@
 # This progress bar is for both downloading and parsing.  Maybe
 # they could be separate stages.
 #
-my $bar = new Term::ProgressBar('getting listings', scalar @to_get)
-  if Have_bar && not $opt_quiet;
+my $bar = new XMLTV::ProgressBar('getting listings', scalar @to_get)
+  if not $opt_quiet;
 foreach (@to_get) {
     foreach (process_table($_->[0], $_->[1], $_->[2])) {
 	$writer->write_programme($_);
     }
-    update $bar if Have_bar && not $opt_quiet;
+    update $bar if not $opt_quiet;
 }
+$bar->finish() if not $opt_quiet;
 $writer->end();
 
 ######################################################################
@@ -488,8 +496,8 @@
 
 # get channel listing
 sub get_channels {
-    my $bar = new Term::ProgressBar('getting list of channels', 1)
-      if Have_bar && not $opt_quiet;
+    my $bar = new XMLTV::ProgressBar('getting list of channels', 1)
+      if not $opt_quiet;
     my %channels;
     my $url="http://www.katso.fi/tvopas";
     my $local_data=get_nice($url);
@@ -531,7 +539,8 @@
 	}
     }
     die "no channels could be found" if not keys %channels;
-    update $bar if Have_bar && not $opt_quiet;
+    update $bar if not $opt_quiet;
+    $bar->finish() if not $opt_quiet;
     return %channels;
 }
 
Index: grab/fr/tv_grab_fr
===================================================================
RCS file: /cvsroot/xmltv/xmltv/grab/fr/tv_grab_fr,v
retrieving revision 1.4
diff -u -B -b -w -r1.4 tv_grab_fr
--- grab/fr/tv_grab_fr	10 Apr 2004 22:00:31 -0000	1.4
+++ grab/fr/tv_grab_fr	2 Jun 2004 09:37:04 -0000
@@ -6,7 +6,7 @@
 
 =head1 SYNOPSIS
 
-To configure: tv_grab_fr --configure [--config-file FILE]
+To configure: tv_grab_fr --configure [--config-file FILE] [--gui OPTION]
 To grab listings: tv_grab_fr [--output FILE] [--quiet]
 Slower, detailed grab: tv_grab_fr --slow [--output FILE] [--days N] [--offset N] [--quiet]
 Help: tv_grab_fr --help
@@ -23,6 +23,11 @@
 B<--configure> Grab channels informations from the website and ask for
 channel type and names.
 
+B<--gui OPTION> Use this option to enable a graphical interface to be used.
+OPTION may be 'Tk' or 'GDialog', or left blank for the best available choice.
+Additional allowed values of OPTION are 'Term' for normal terminal output
+(default) and 'TermNoProgressBar' to disable the use of Term::ProgressBar.
+
 B<--output FILE> write to FILE rather than standard output.
 
 B<--days N> grab N days starting from today, rather than as many as
@@ -60,9 +65,6 @@
 END
   ;
 
-# Use Term::ProgressBar if installed.
-use constant Have_bar => eval { require Term::ProgressBar; 1 };
-
 use warnings;
 use strict;
 use XMLTV::Version '$Id$ ';
@@ -75,6 +77,7 @@
 use XMLTV;
 use XMLTV::Memoize;
 use XMLTV::Ask;
+use XMLTV::ProgressBar;
 use XMLTV::Mode;
 use XMLTV::Config_file;
 use XMLTV::DST;
@@ -106,7 +109,7 @@
 # Global variables allocation according to options
 #***************************************************************************
 XMLTV::Memoize::check_argv('get_page_aux');
-my ($opt_days,  $opt_help,  $opt_output,  $opt_offset,  $opt_quiet,  $opt_list_channels, $opt_config_file, $opt_configure, $opt_slow, $opt_licons);
+my ($opt_days,  $opt_help,  $opt_output,  $opt_offset,  $opt_gui, $opt_quiet,  $opt_list_channels, $opt_config_file, $opt_configure, $opt_slow, $opt_licons);
 $opt_quiet  = 0;
 # The website is able to store up to nine days from now
 my $default_opt_days = 9;
@@ -118,6 +121,7 @@
      'quiet'     => \$opt_quiet,
      'configure' => \$opt_configure,
      'config-file=s' => \$opt_config_file,
+     'gui:s'     => \$opt_gui,
      'list-channels' => \$opt_list_channels,
      'slow' => \$opt_slow
     )
@@ -130,6 +134,9 @@
 die 'Cannot get more than one day before current day' if (defined $opt_offset && $opt_offset < -1);
 usage(1) if $opt_help;
 
+XMLTV::Ask::init($opt_gui);
+XMLTV::ProgressBar::init($opt_gui);
+
 if (not $opt_slow) {
     # Certain options are ignored in fast mode.
     my %slow_options = (days => $opt_days,
@@ -244,11 +251,11 @@
     my @gts = sort keys %GridType;
     my @gtnames = map { $GridType{$_} } @gts;
     my @gtqs = map { "Get channels type : $_?" } @gts;
-    my @gtwant = askManyBooleanQuestions(1, @gtqs);
+    my @gtwant = ask_many_boolean(1, @gtqs);
 
-    my $bar = new Term::ProgressBar('getting channel lists',
+    my $bar = new XMLTV::ProgressBar('getting channel lists',
                                     scalar grep { $_ } @gtwant)
-                    if Have_bar && not $opt_quiet;
+                    if not $opt_quiet;
     my %channels_for;
     foreach my $i (0 .. $#gts) {
         my ($gt, $gtw, $gtname) = ($gts[$i], $gtwant[$i], $gtnames[$i]);
@@ -256,8 +263,9 @@
         my %channels = get_channels( $gtname );
         die 'No channels could be found' if not %channels;
         $channels_for{$gt} = \%channels;
-        update $bar if Have_bar && not $opt_quiet;
+        update $bar if not $opt_quiet;
     }
+    $bar->finish();
 
     my %asked;
     foreach (@gts) {
@@ -271,7 +279,7 @@
             my @chs = grep { not $asked{$_}++ } sort keys %channels;
             my @names = map { $channels{$_}{name} } @chs;
             my @qs = map { "add channel $_?" } @names;
-            my @want = askManyBooleanQuestions(1, @qs);
+            my @want = ask_many_boolean(1, @qs);
             foreach (@chs) {
                 my $w = shift @want;
                 warn("cannot read input, stopping channel questions"), last if not defined $w;
@@ -323,7 +331,7 @@
     my @gts = sort keys %GridType;
     my @gtnames = map { $GridType{$_} } @gts;
     my @gtqs = map { "List channels for grid : $_?" } @gts;
-    my @gtwant = askManyBooleanQuestions(1, @gtqs);
+    my @gtwant = ask_many_boolean(1, @gtqs);
 
     foreach (@gts) {
         my $gtw = shift @gtwant;
@@ -384,13 +392,14 @@
         push @to_get, [ $url, $chid, $_ ];
     }
 }
-my $bar = new Term::ProgressBar('getting listings', scalar @to_get)  if Have_bar && not $opt_quiet;
+my $bar = new XMLTV::ProgressBar('getting listings', scalar @to_get)  if not $opt_quiet;
 foreach (@to_get) {
     my ($url, $chid, $slot ) = @$_;
     process_channel_grid_page($writer, $chid, $url, $slot);
-    update $bar if Have_bar && not $opt_quiet;
+    update $bar if not $opt_quiet;
 }
 $writer->end();
+$bar->finish();
 
 #***************************************************************************
 # Specific functions for grabbing information
Index: grab/hu/tv_grab_hu
===================================================================
RCS file: /cvsroot/xmltv/xmltv/grab/hu/Attic/tv_grab_hu,v
retrieving revision 1.13
diff -u -B -b -w -r1.13 tv_grab_hu
--- grab/hu/tv_grab_hu	10 Apr 2004 22:00:31 -0000	1.13
+++ grab/hu/tv_grab_hu	2 Jun 2004 09:37:05 -0000
@@ -10,7 +10,7 @@
 
 tv_grab_hu --help
 
-tv_grab_hu [--config-file FILE] --configure
+tv_grab_hu [--config-file FILE] --configure [--gui OPTION]
 
 tv_grab_hu [--config-file FILE] [--output FILE] [--days N]
            [--offset N] [--quiet]
@@ -33,6 +33,11 @@
 default is B<~/.xmltv/tv_grab_hu.conf>.  This is the file written by
 B<--configure> and read when grabbing.
 
+B<--gui OPTION> Use this option to enable a graphical interface to be used.
+OPTION may be 'Tk' or 'GDialog', or left blank for the best available choice.
+Additional allowed values of OPTION are 'Term' for normal terminal output
+(default) and 'TermNoProgressBar' to disable the use of XMLTV::ProgressBar.
+
 B<--output FILE> write to FILE rather than standard output.
 
 B<--days N> grab N days.  The default is eight.
@@ -76,6 +81,7 @@
 
 use XMLTV;
 use XMLTV::Memoize;
+use XMLTV::ProgressBar;
 use XMLTV::Ask;
 use XMLTV::DST;
 use XMLTV::Get_nice;
@@ -85,16 +91,13 @@
 # Todo: perhaps we should internationalize messages and docs?
 use XMLTV::Usage <<END
 $0: get Hungarian television listings in XMLTV format
-To configure: $0 --configure [--config-file FILE]
+To configure: $0 --configure [--config-file FILE] [--gui]
 To grab listings: $0 [--config-file FILE] [--output FILE] [--days N]
         [--offset N] [--quiet]
 To list channels: $0 --list-channels
 END
   ;
 
-# Use Term::ProgressBar if installed.
-use constant Have_bar => eval { require Term::ProgressBar; 1 };
-
 # Attributes of the root element in output.
 my $HEAD = { 'source-info-url'     => 'http://www.port.hu/',
 	     'source-data-url'     => "http://www.port.hu/tv/",
@@ -120,8 +123,8 @@
 # Get options, including undocumented --cache option.
 XMLTV::Memoize::check_argv('XMLTV::Get_nice::get_nice_aux');
 my ($opt_days, $opt_offset, $opt_help, $opt_output,
-    $opt_configure, $opt_config_file, $opt_quiet,
-    $opt_list_channels);
+    $opt_configure, $opt_config_file, $opt_gui,,
+    $opt_quiet, $opt_list_channels);
 $opt_days = 8; # default
 $opt_offset = 0; # default
 $opt_quiet = 0; # default
@@ -130,6 +133,7 @@
 	   'help'          => \$opt_help,
 	   'configure'     => \$opt_configure,
 	   'config-file=s' => \$opt_config_file,
+           'gui:s'         => \$opt_gui,
 	   'output=s'      => \$opt_output,
 	   'quiet'         => \$opt_quiet,
 	   'list-channels' => \$opt_list_channels,
@@ -138,6 +142,10 @@
 die 'number of days must not be negative'
   if (defined $opt_days && $opt_days < 0);
 usage(1) if $opt_help;
+
+XMLTV::Ask::init($opt_gui);
+XMLTV::ProgressBar::init($opt_gui);
+
 my $mode = XMLTV::Mode::mode('grab', # default
 			     $opt_configure => 'configure',
 			     $opt_list_channels => 'list-channels',
@@ -173,7 +181,7 @@
     my @chs = sort keys %channels;
     my @names = map { $channels{$_} } @chs;
     my @qs = map { "add channel $_?" } @names;
-    my @want = askManyBooleanQuestions(1, @qs);
+    my @want = ask_many_boolean(1, @qs);
     foreach (@chs) {
 	my $w = shift @want;
 	warn("cannot read input, stopping channel questions"), last
@@ -265,8 +273,8 @@
 # This progress bar is for both downloading and parsing.  Maybe
 # they could be separate stages.
 #
-my $bar = new Term::ProgressBar('getting listings', @days * @channels)
-  if Have_bar && not $opt_quiet;
+my $bar = new XMLTV::ProgressBar('getting listings', @days * @channels)
+  if not $opt_quiet;
 foreach my $d (@days) {
     my ($day, $i) = @$d;
     my $some_success = 0;
@@ -274,13 +282,15 @@
 	my @ps = process_table($day, xid($ch_did), $ch_did, $i);
 	$some_success = 1 if @ps;
 	$writer->write_programme($_) foreach @ps;
-	update $bar if Have_bar && not $opt_quiet;
+	update $bar if not $opt_quiet;
     }
     if (@channels and not $some_success) {
 	warn "failed to get any listings for day $i, stopping\n";
 	last;
     }
+    
 }
+$bar->finish() if not $opt_quiet;
 $writer->end();
 
 ######################################################################
@@ -465,8 +475,8 @@
 
 # get channel listing
 sub get_channels {
-    my $bar = new Term::ProgressBar('getting list of channels', 1)
-	if Have_bar && not $opt_quiet;
+    my $bar = new XMLTV::ProgressBar('getting list of channels', 1)
+	if not $opt_quiet;
     my %channels;
     my $url="http://www.port.hu/pls/tv/tv.prog";
     my $local_data=get_nice($url);
@@ -502,7 +512,8 @@
 	}
     }
     die "no channels could be found" if not keys %channels;
-    update $bar if Have_bar && not $opt_quiet;
+    update $bar if not $opt_quiet;
+    $bar->finish() if not $opt_quiet;
     return %channels;
 }
 
Index: grab/huro/tv_grab_huro
===================================================================
RCS file: /cvsroot/xmltv/xmltv/grab/huro/tv_grab_huro,v
retrieving revision 1.3
diff -u -B -b -w -r1.3 tv_grab_huro
--- grab/huro/tv_grab_huro	9 May 2004 17:49:12 -0000	1.3
+++ grab/huro/tv_grab_huro	2 Jun 2004 09:37:05 -0000
@@ -185,7 +185,7 @@
     open(CONF, ">$config_file") or die "cannot write to $config_file: $!";
 
     my $default_cn = 'Hungary';
-    my $cn = askQuestion('Grab listings for which country?',
+    my $cn = ask_choice('Grab listings for which country?',
 			 $default_cn,
 			 sort keys %COUNTRIES);
     my $c = $COUNTRIES{$cn}[0];
@@ -196,7 +196,7 @@
     my @chs = sort keys %channels;
     my @names = map { $channels{$_} } @chs;
     my @qs = map { "add channel $_?" } @names;
-    my @want = askManyBooleanQuestions(1, @qs);
+    my @want = ask_many_boolean(1, @qs);
     foreach (@chs) {
 	my $w = shift @want;
 	warn("cannot read input, stopping channel questions"), last
Index: grab/it/tv_grab_it.in
===================================================================
RCS file: /cvsroot/xmltv/xmltv/grab/it/tv_grab_it.in,v
retrieving revision 1.24
diff -u -B -b -w -r1.24 tv_grab_it.in
--- grab/it/tv_grab_it.in	10 Apr 2004 22:00:31 -0000	1.24
+++ grab/it/tv_grab_it.in	2 Jun 2004 09:37:06 -0000
@@ -11,7 +11,7 @@
 
 tv_grab_it --help
 
-tv_grab_it [--config-file FILE] --configure
+tv_grab_it [--config-file FILE] --configure [--gui OPTION]
 
 tv_grab_it [--config-file FILE] [--output FILE] [--days N]
            [--offset N] [--quiet] [--slow]
@@ -33,6 +33,11 @@
 default is B<~/.xmltv/tv_grab_it.conf>.  This is the file written
 by B<--configure> and read when grabbing.
 
+B<--gui OPTION> Use this option to enable a graphical interface to be used.
+OPTION may be 'Tk' or 'GDialog', or left blank for the best available choice.
+Additional allowed values of OPTION are 'Term' for normal terminal output
+(default) and 'TermNoProgressBar' to disable the use of XMLTV::ProgressBar.
+
 B<--output FILE> write to FILE rather than standard output.
 
 B<--days N> grab N days.  The default is 7.
@@ -79,6 +84,7 @@
 use Memoize;
 use XMLTV;
 use XMLTV::Memoize;
+use XMLTV::ProgressBar;
 use XMLTV::Ask;
 use XMLTV::Config_file;
 use XMLTV::DST;
@@ -86,7 +92,7 @@
 # Todo: perhaps we should internationalize messages and docs?
 use XMLTV::Usage <<END
 $0: get Italian television listings in XMLTV format
-To configure: $0 --configure [--config-file FILE]
+To configure: $0 --configure [--config-file FILE] [--gui OPTION]
 To grab listings: $0 [--config-file FILE] [--output FILE] [--days N]
         [--offset N] [--quiet] [--slow]
 END
@@ -115,11 +121,6 @@
 my $base="http://$domain/canali.phtml";
 my $rturl="http://$domain/";
 
-######################################################################
-# Use Term::ProgressBar if installed.
-use constant Have_bar => eval { require Term::ProgressBar; 1 };
-
-######################################################################
 # get options
 # Get options, including undocumented --cache option.
 
@@ -135,6 +136,7 @@
     $opt_slow,
     $opt_configure,
     $opt_config_file,
+    $opt_gui,
     $opt_quiet,
     $opt_share,
    );
@@ -150,6 +152,7 @@
 	   'help'          => \$opt_help,
 	   'configure'     => \$opt_configure,
 	   'config-file=s' => \$opt_config_file,
+           'gui:s'         => \$opt_gui,
 	   'output=s'      => \$opt_output,
 	   'quiet'         => \$opt_quiet,
 	   'slow'	   => \$opt_slow,
@@ -160,6 +163,9 @@
   if (defined $opt_days && $opt_days < 0);
 usage(1) if $opt_help;
 
+XMLTV::Ask::init($opt_gui);
+XMLTV::ProgressBar::init($opt_gui);
+
 # share/ directory for storing the list of dud channels.  This next line
 # is altered by processing through tv_grab_it.PL.  But we can use the
 # current directory instead of share/tv_grab_it for development.
@@ -206,17 +212,17 @@
 		}
     open(CONF, ">$config_file") or die "cannot write to $config_file: $!";
     # find list of available channels
-    my $bar = new Term::ProgressBar('getting list of channels', 1)
-      if Have_bar && not $opt_quiet;
+    my $bar = new XMLTV::ProgressBar('getting list of channels', 1)
+      if not $opt_quiet;
 	my %channels=get_channels_list($content);
     die "no channels could be found" if (scalar(keys(%channels))==0);
-    update $bar if Have_bar && not $opt_quiet;
+    update $bar if not $opt_quiet;
 
     # Ask about each channel.
     my @chs = sort keys %channels;
     my @names = map { $channels{$_} } @chs;
     my @qs = map { "add channel $_?" } @names;
-    my @want = askManyBooleanQuestions(1, @qs);
+    my @want = ask_many_boolean(1, @qs);
     foreach (@chs) {
 	my $w = shift @want;
 	warn("cannot read input, stopping channel questions"), last
@@ -230,6 +236,7 @@
 	# TODO don't store display-name in config file.
     }
 
+    $bar->finish() if not $opt_quiet;
     close CONF or warn "cannot close $config_file: $!";
     say("Finished configuration.");
 
@@ -297,8 +304,8 @@
 }
 t "will get $days2get days from $opt_offset onwards";
 
-my $bar2 = new Term::ProgressBar('getting icons', scalar @channels)
-  if Have_bar && not $opt_quiet;
+my $bar2 = new XMLTV::ProgressBar('getting icons', scalar @channels)
+  if not $opt_quiet;
 foreach my $ch_id (@channels) {
     my $ch_xid=id_to_xid($ch_id);
 
@@ -316,11 +323,12 @@
 		       'display-name' => [ [ $channels{$ch_id} ] ],
 		       icon => [{src => get_icon($url)}]
 		      });
-    update $bar2 if Have_bar && not $opt_quiet;
+    update $bar2 if not $opt_quiet;
 }
+$bar2->finish() if not $opt_quiet;
 
-my $bar = new Term::ProgressBar('getting listings', scalar @to_get)
-  if Have_bar && not $opt_quiet;
+my $bar = new XMLTV::ProgressBar('getting listings', scalar @to_get)
+  if not $opt_quiet;
 foreach (@to_get) {
     my $canale= $_->[1];
     $url   = $_->[0];
@@ -347,9 +355,10 @@
 	$w->write_programme($_) foreach @dati;
     }
     #			else {warn "skipping $canale \n";}
-    update $bar if Have_bar && not $opt_quiet;
+    update $bar if not $opt_quiet;
 }
 $w->end;
+$bar->finish() if not $opt_quiet;
 
 ######################################################################
 # subroutines
Index: grab/jp/tv_grab_jp
===================================================================
RCS file: /cvsroot/xmltv/xmltv/grab/jp/tv_grab_jp,v
retrieving revision 1.2
diff -u -B -b -w -r1.2 tv_grab_jp
--- grab/jp/tv_grab_jp	26 Feb 2004 04:28:26 -0000	1.2
+++ grab/jp/tv_grab_jp	2 Jun 2004 09:37:07 -0000
@@ -10,7 +10,7 @@
 
 tv_grab_jp --help
 
-tv_grab_jp [--config-file FILE] --configure
+tv_grab_jp [--config-file FILE] --configure [--gui OPTION]
 
 tv_grab_jp [--config-file FILE] [--output FILE] [--days N]
            [--offset N] [--enable-readstr] [--no-category-code]
@@ -37,6 +37,11 @@
 default is B<~/.xmltv/tv_grab_jp.conf>.  This is the file written by
 B<--configure> and read when grabbing.
 
+B<--gui OPTION> Use this option to enable a graphical interface to be used.
+OPTION may be 'Tk' or 'GDialog', or left blank for the best available choice.
+Additional allowed values of OPTION are 'Term' for normal terminal output
+(default) and 'TermNoProgressBar' to disable the use of Term::ProgressBar.
+
 B<--output FILE> write to FILE rather than standard output.
 
 B<--days N> grab N days.  The default and the maximum is 7.
@@ -84,6 +89,7 @@
 use XMLTV;
 use XMLTV::Memoize;
 use XMLTV::Ask;
+use XMLTV::ProgressBar;
 use XMLTV::Config_file;
 use XMLTV::Mode;
 use XMLTV::Get_nice;
@@ -103,9 +109,6 @@
 END
   ;
 
-# Use Term::ProgressBar if installed.
-use constant Have_bar => eval { require Term::ProgressBar; 1 };
-
 # Attributes of the root element in output.
 my $HEAD = { 'source-info-url'     => 'http://www.ontvjapan.com/',
 	     'source-data-url'     => "http://www.ontvjapan.com/program/",
@@ -152,7 +155,7 @@
 
 XMLTV::Memoize::check_argv('XMLTV::Get_nice::get_nice_aux');
 my ($opt_days, $opt_offset, $opt_help, $opt_output,
-    $opt_configure, $opt_config_file, $opt_quiet,
+    $opt_configure, $opt_config_file, $opt_quiet, $opt_gui,
     $opt_readstr, $opt_no_cat_code, $opt_list_channels);
 $opt_days        = 7; # default
 $opt_offset      = 0; # default
@@ -162,6 +165,7 @@
 GetOptions('help'             => \$opt_help,
 	   'configure'        => \$opt_configure,
 	   'config-file=s'    => \$opt_config_file,
+       'gui:s'            => \$opt_gui,
 	   'output=s'         => \$opt_output,
 	   'quiet'            => \$opt_quiet,
 	   'enable-readstr'   => \$opt_readstr,
@@ -177,6 +181,9 @@
     if (defined $opt_offset && $opt_offset < 0);
 usage(1) if $opt_help;
 
+XMLTV::Ask::init($opt_gui);
+XMLTV::ProgressBar::init($opt_gui);
+
 my $mode = XMLTV::Mode::mode('grab', # default
 			     $opt_configure => 'configure',
 			     $opt_list_channels => 'list-channels',
@@ -204,9 +211,10 @@
     # Get region information
     my $barmsg = select_lang('地域情報を取得中',
 			     'getting regions');
-    my $bar = new Term::ProgressBar($barmsg, 1) if Have_bar;
+    my $bar = new XMLTV::ProgressBar($barmsg, 1);
     my %regions = get_regions();
-    update $bar if Have_bar;
+    update $bar;
+    $bar->finish();
 
     # Ask region
     my %ask_regions;
@@ -233,7 +241,7 @@
 
     $msg = select_lang("\n地域を指定してください: ",
 		       "\nSelect your region: ");
-    $req_region = askQuestion($msg, $req_region, @r);
+    $req_region = ask_choice($msg, $req_region, @r);
     $regionid = $ask_regions{$req_region};
 
     if ($regionid eq 'CATV') {
@@ -250,9 +258,10 @@
 
     $barmsg = select_lang('チャンネル情報を取得中',
 			     'getting channels');
-    $bar = new Term::ProgressBar($barmsg, 1) if Have_bar;
+    $bar = new XMLTV::ProgressBar($barmsg, 1);
     my @channels = get_channels($regionid);
-    update $bar if Have_bar;
+    update $bar;
+    $bar->finish();
 
     # Ask about each channel.
     Text::Kakasi::getopt_argv('kakasi', '-ieuc', '-Ja', '-Ha', '-Ka', '-Ea');
@@ -271,7 +280,7 @@
     }
     Text::Kakasi::close_kanwadict();
 
-    my @want = askManyBooleanQuestions(1, @qs);
+    my @want = ask_many_boolean(1, @qs);
     foreach (@channels) {
 	my $w = shift @want;
 	my $id = $_->{id};
@@ -287,7 +296,7 @@
 
     close CONF or warn "cannot close $config_file: $!";
     $msg = select_lang("設定完了.\n", "Finished\n");
-    print $msg;
+    say( $msg );
 
     exit();
 }
@@ -314,16 +323,16 @@
     # Get region information
     my $barmsg = select_lang('地域情報を取得中',
 			     'getting regions');
-    my $bar = new Term::ProgressBar($barmsg, 1) if Have_bar;
+    my $bar = new XMLTV::ProgressBar($barmsg, 1);
     my %regions = get_regions();
-    update $bar if Have_bar;
+    update $bar;
+    $bar->finish();
 
     # get channels for all region
     my @ch_regions = grep { $_ ne 'CATV' } sort keys %regions;
     $barmsg = select_lang('チャンネル情報を取得中',
 			  'getting channels');
-    $bar = new Term::ProgressBar($barmsg, scalar keys %regions)
-	if Have_bar;
+    $bar = new XMLTV::ProgressBar($barmsg, scalar keys %regions);
     foreach (@ch_regions) {
 	my $regionid = $_;
 	my @channels = get_channels($regionid);
@@ -342,8 +351,9 @@
 				'callsign' => $callsign };
 	    }
 	}
-	update $bar if Have_bar;
+	update $bar;
     }
+    $bar->finish();
 
     # Write channels mode.
     foreach (@ch_all) {
@@ -385,9 +395,10 @@
 
 my $barmsg = select_lang('チャンネル情報を取得中',
 			 'getting channels');
-my $bar = new Term::ProgressBar($barmsg, 1) if Have_bar;
+my $bar = new XMLTV::ProgressBar($barmsg, 1);
 my @channels = get_channels($regionid);
-update $bar if Have_bar;
+update $bar;
+$bar->finish();
 
 foreach (@channels) {
     my $id = $_->{id};
@@ -409,13 +420,14 @@
 # This progress bar is for both downloading and parsing.
 $barmsg = select_lang('番組表を取得中',
 		      'getting program info');
-$bar = new Term::ProgressBar($barmsg, scalar @ch_all)
-    if Have_bar && not $opt_quiet;
+$bar = new XMLTV::ProgressBar($barmsg, scalar @ch_all)
+    if not $opt_quiet;
 foreach (@ch_all) {
     process_table($_->{'id'});
-    update $bar if Have_bar && not $opt_quiet;
+    update $bar if not $opt_quiet;
 }
 $writer->end();
+$bar->finish();
 
 ######################################################################
 # subroutine definitions
Index: grab/na/tv_grab_na
===================================================================
RCS file: /cvsroot/xmltv/xmltv/grab/na/Attic/tv_grab_na,v
retrieving revision 1.98
diff -u -B -b -w -r1.98 tv_grab_na
--- grab/na/tv_grab_na	2 Mar 2004 21:00:03 -0000	1.98
+++ grab/na/tv_grab_na	2 Jun 2004 09:37:10 -0000
@@ -8,7 +8,7 @@
 
 =head1 SYNOPSIS
 
-tv_grab_na [--help] [--debug] [--config-file FILE] [--output FILE] --configure CONFIG_OPTIONS
+tv_grab_na [--help] [--debug] [--config-file FILE] [--output FILE] [--gui OPTION] --configure CONFIG_OPTIONS
 
 tv_grab_na [--help] [--debug] [--config-file FILE] [--output FILE] GRAB_OPTIONS
 
@@ -45,6 +45,11 @@
 written in configure mode and read in grab mode.  The default location
 is B<~/.xmltv/tv_grab_na.conf>.
 
+B<--gui OPTION> Use this option to enable a graphical interface to be used.
+OPTION may be 'Tk' or 'GDialog', or left blank for the best available choice.
+Additional allowed values of OPTION are 'Term' for normal terminal output
+(default) and 'TermNoProgressBar' to disable the use of Term::ProgressBar.
+
 B<--output FILE> Write output to FILE rather than standard output.
 
 Configuration is normally interactive, but it can be run
@@ -510,6 +515,7 @@
 use Date::Manip;
 
 use XMLTV::ZapListings;
+use XMLTV::ProgressBar;
 use XMLTV::Ask;
 use XMLTV::Config_file;
 use XMLTV;
@@ -522,6 +528,15 @@
 
 our ($opt_debug, $opt_quiet, $opt_stats);
 
+sub termStatusMessage($)
+{
+    my $msg = shift;
+    die 'expected a message' if not defined $msg;
+    if ( !$opt_quiet ) {
+	print STDERR $msg;
+    }
+}
+
 sub statusMessage($)
 {
     my $msg = shift;
@@ -690,6 +705,9 @@
     print $fp "   --config-file <file>\n";
     print $fp "     use <file> as config file\n";
     print $fp "\n";
+    print $fp "   --gui <option>\n";
+    print $fp "     Use a GUI to ask questions\n";
+    print $fp "\n";
     print $fp "   --output <file>\n";
     print $fp "     redirect output to <file> rather than stdout. Currently only effects\n";
     print $fp "      --list-providers or --list-channels\n";
@@ -853,6 +871,7 @@
 	 $opt_zipcode,
 	 $opt_provider,
 	 $opt_config_file,
+         $opt_gui,
 	 $opt_output,
 	 $opt_retry_limit,
 	 $opt_retry_delay,
@@ -874,6 +893,7 @@
 			 quiet
 			 release-check=s
 			 config-file=s
+                         gui:s
 			 output=s
 			 postalcode=s
 			 zipcode=s
@@ -895,6 +915,9 @@
 
     checkCache(); # make sure $opt_cache is a dir or undef
 
+    XMLTV::Ask::init($opt_gui);
+    XMLTV::ProgressBar::init($opt_gui);
+    
     $opt_release_check=($opt_release_check=~m;^t;i);
 
     $opt_quiet=(defined($opt_quiet));
@@ -1098,9 +1121,13 @@
 );
 
     if ( $opt_release_check ) {
-	statusMessage("checking XMLTV release information..\n");
+            
+        my $bar = new XMLTV::ProgressBar('checking XMLTV release information..', 1);
 
 	my $retval=XMLTV::ZapListings::getCurrentReleaseInfo($xmltvFreshmeatRecord, $opt_debug);
+        $bar->update();
+        $bar->finish();
+
 	if ( !defined($retval) ) {
 	    statusMessage( <<END
 Warning: failed to get current release information from:
@@ -1121,7 +1148,7 @@
 	}
     }
 
-    statusMessage("\nstarting manual configuration process..\n\n");
+    termStatusMessage("\nstarting manual configuration process..\n\n");
 
     while ( !defined($config->{option_retry_limit}) ) {
 	my $res=ask('how many times do you want to retry on www site failures ? (default=2)');
@@ -1185,7 +1212,7 @@
 	$code=$config->{option_postalcode} if ( defined($config->{option_postalcode}) );
 	$code=$config->{option_zipcode} if ( defined($config->{option_zipcode}) );
 
-	statusMessage("\ngetting list of providers for postal/zip code $code, be patient..\n");
+        my $bar = new XMLTV::ProgressBar("getting list of providers for postal/zip code $code, be patient..", 1);
 
 	my @providers;
 	my $failed=0;
@@ -1224,11 +1251,11 @@
 		if ( $p->{id} eq $config->{option_provider} ) {
 		    if ( $config->{option_provider_desc} ne $p->{description} ) {
 			if ( $opt_auto_fail_on_provider_changes ) {
-			    statusMessage("provider:\n\t".$config->{option_provider}." \# ".$config->{option_provider_desc}.
+			    termStatusMessage("provider:\n\t".$config->{option_provider}." \# ".$config->{option_provider_desc}.
 					  "\nhas new description (".$p->{description}."), exiting\n");
 			    exit(1);
 			}
-			statusMessage("updating provider description to: $p->{description}\n");
+			termStatusMessage("updating provider description to: $p->{description}\n");
 			$config->{option_provider_desc}=$p->{description};
 		    }
 		    $still_valid=1;
@@ -1248,7 +1275,7 @@
 				exit(1);
 			    }
 			}
-			statusMessage("updating provider id to: $p->{id}\n");
+			termStatusMessage("updating provider id to: $p->{id}\n");
 			$config->{option_provider}=$p->{id};
 			$still_valid=1;
 		    }
@@ -1275,6 +1302,9 @@
 	    }
 	}
 
+        $bar->update();
+        $bar->finish();
+        
 	while ( !$config->{option_provider} ) {
 	    my @providersWithIds;
 
@@ -1308,7 +1338,7 @@
 		$defaultProviderId =$1;
 	    }
 
-	    my $res = askQuestion("Choose a service provider: ",
+	    my $res = ask_choice("Choose a service provider: ",
 				  $idToDesc{$defaultProviderId},
 				  @providersWithIds);
 	    die 'failed to choose service provider' if not defined $res;
@@ -1316,18 +1346,19 @@
 	    my $id = $1; die if not defined $id;
 	    $config->{option_provider}=$id;
 	    $config->{option_provider_desc}=$res;
-	    statusMessage("\nyou chose $id \# $res\n");
+	    termStatusMessage("\nyou chose $id \# $res\n");
 	}
     }
 
     # if we're in the configure step, lets refresh the list of channels
     # being careful to warn about additions and deletions
     
+    my $bar;
     if ( $config->haveAnyChannels() ) {
-	statusMessage("\nchecking for changes to channel list, be patient..\n");
+        $bar = new XMLTV::ProgressBar('checking for changes to channel list, be patient..', 1);
     }
     else {
-	statusMessage("\ngetting channel list, be patient..\n");
+        $bar = new XMLTV::ProgressBar('getting channel list, be patient..', 1);
     }
 
     my @channels;
@@ -1339,7 +1370,7 @@
 	}
 
 	@channels=$zl->getChannelList($config->{option_provider});
-	statusMessage("got channel list\n");
+	termStatusMessage("got channel list\n");
 	if ( ! @channels || !defined($channels[0]) ) {
 	    $failed=1;
 	}
@@ -1348,6 +1379,8 @@
 	    last;
 	}
     }
+    $bar->update();
+    $bar->finish();
 
     if ( $failed ) {
 	#errorMessage("$0: failed to get list of channels for postal/zip code $config->{option_postalzipcode}\n");
@@ -1375,18 +1408,18 @@
     if ( defined($opt_auto_new_channels) ) {
 	if ( $opt_auto_new_channels eq "ignore" ) {
 	    @r = map { 0 } @toAsk;
-	    statusMessage("ignoring new channel $_\n") foreach @toAsk;
+	    termStatusMessage("ignoring new channel $_\n") foreach @toAsk;
 	}
 	elsif ( $opt_auto_new_channels eq "add" ) {
 	    @r = map { 1 } @toAsk;
-	    statusMessage("adding new channel $_\n") foreach @toAsk;
+	    termStatusMessage("adding new channel $_\n") foreach @toAsk;
 	}
 	else {
 	    die "invalid auto_new_channels $opt_auto_new_channels\n";
 	}
     }
     else {
-	@r = askManyBooleanQuestions(1, map { "add channel $_ ?" } @toAsk);
+	@r = ask_many_boolean(1, map { "add channel $_ ?" } @toAsk);
     }
 
     # now use the Boolean answers in @r
@@ -1415,18 +1448,18 @@
     if ( defined($opt_auto_missing_channels) ) {
 	if ( $opt_auto_missing_channels eq "ignore" ) {
 	    @r = map { 0 } @toAsk;
-	    statusMessage("ignoring missing channel $_...\n") foreach @toAsk;
+	    termStatusMessage("ignoring missing channel $_...\n") foreach @toAsk;
 	}
 	elsif ( $opt_auto_missing_channels eq "remove" ) {
 	    @r = map { 1 } @toAsk;
-	    statusMessage("removing channel $_..\n") foreach @toAsk;
+	    termStatusMessage("removing channel $_..\n") foreach @toAsk;
 	}
 	else {
 	    die "invalid auto_missing_channels $opt_auto_missing_channels\n";
 	}
     }
     else {
-	@r = askManyBooleanQuestions(1, map { "remove no-longer available channel $_ ?" } @toAsk);
+	@r = ask_many_boolean(1, map { "remove no-longer available channel $_ ?" } @toAsk);
     }
 
     foreach my $station (@toAsk) {
@@ -1442,15 +1475,15 @@
 
     if ( $channelsUpdated == 0 ) {
 	if ( $config->haveAnyChannels() ) {
-	    statusMessage("\nchannel line-up hasn't changed\n");
+	    termStatusMessage("\nchannel line-up hasn't changed\n");
 	}
 	else {
-	    statusMessage("\nno channels added\n");
+	    termStatusMessage("\nno channels added\n");
 	}
     } 
 
     # write out config file
-    statusMessage("\nupdating $configfile..\n");
+    termStatusMessage("\nupdating $configfile..\n");
     if ( $config->save($configfile) != 0 ) {
 	errorMessage("$0: $configfile save failed\n");
 	exit(1);
Index: grab/na_dd/tv_grab_na_dd
===================================================================
RCS file: /cvsroot/xmltv/xmltv/grab/na_dd/Attic/tv_grab_na_dd,v
retrieving revision 1.39
diff -u -B -b -w -r1.39 tv_grab_na_dd
--- grab/na_dd/tv_grab_na_dd	11 Apr 2004 04:35:48 -0000	1.39
+++ grab/na_dd/tv_grab_na_dd	2 Jun 2004 09:37:10 -0000
@@ -14,6 +14,7 @@
     
     tv_grab_na_dd --configure [--config-file FILE] [--dd-data FILE]
                               [--reprocess] [--auto-config add|ignore]
+                              [--gui OPTION]
 
     tv_grab_na_dd --list-lineups [--config-file FILE] [--dd-data FILE]
                               [--reprocess]
@@ -55,6 +56,12 @@
 Lists available lineups.  Only requires username in the config file. Used by programs that
 automate the L<--configure> process.
 
+=item --gui OPTION
+
+Use this option to enable a graphical interface to be used.
+OPTION may be 'Tk' or 'GDialog', or left blank for the best available choice.
+Additional allowed values of OPTION are 'Term' for normal terminal output
+(default) and 'TermNoProgressBar' to disable the use of Term::ProgressBar.
 
 =back
 
@@ -236,10 +243,10 @@
 use File::Temp qw(tempfile);
 use Getopt::Long;
 use XML::Twig 3.10;
-use constant Have_bar => eval { require Term::ProgressBar; 1 };
 
 use XMLTV;
 use XMLTV::Ask;
+use XMLTV::ProgressBar;
 use XMLTV::Config_file;
 use XMLTV::Version '$Id$';
 use XMLTV::Usage <<END
@@ -250,6 +257,7 @@
 To Configure
     $0 --configure [--config-file FILE] [--dd-data FILE]
                               [--reprocess] [--auto-config add|ignore]
+                   [--gui OPTION]
 To get listings
     $0 [--config-file FILE] [--dd-data FILE]
                   [--reprocess] [--auto-config add|ignore]
@@ -298,6 +306,7 @@
 my $opt_help;           # ask for help
 my $opt_configure;      # configure mode
 my $opt_config_file ;   # config_file_name
+my $opt_gui;            # what type of gui to use
 my $opt_output;         # output name
 my $opt_days       =10; # days to fetch
 my $opt_offset     =0;  # day to start
@@ -325,6 +334,7 @@
     	   'configure'     => \$opt_configure,
     	   'config-file=s' => \$opt_config_file,
     	   'config_file=s' => \$opt_config_file,
+           'gui:s'         => \$opt_gui,
     	   'output=s'      => \$opt_output,
            'days=i'        => \$opt_days,
     	   'offset=i'      => \$opt_offset,
@@ -347,6 +357,10 @@
 	  )
   or usage(0);
 usage(1) if $opt_help;
+
+XMLTV::Ask::init($opt_gui);
+XMLTV::ProgressBar::init($opt_gui);
+
 die 'number of days must not be negative' if ($opt_days < 0);
 die "must specify --dd_data during reprocess\n" if $opt_reprocess and not $opt_dd_data;
 die "--auto-config must be 'add' or 'ignore'\n" if $opt_auto_config && $opt_auto_config !~ /^(add|ignore)$/;
@@ -414,7 +428,13 @@
     open(CONF,">$config_file") or die "can't open config file: $config_file\n";
     if ( ! $opt_auto_config )
     {
-       say("
+       while (1)
+       {
+#        $opt_tz_offset='local' unless length($opt_tz_offset);
+#        $opt_tz_offset=ask("Timezone offset (+/-####), or 'local' ($opt_tz_offset)")||'local';
+         $opt_tz_offset=$ENV{TZ} unless defined $opt_tz_offset || ( defined($ENV{TZ}) && $ENV{TZ} !~ /[+-]\d\d\d\d/ );
+         $opt_tz_offset='+0000'  unless defined $opt_tz_offset;
+         $opt_tz_offset=ask("
 DD data is in UTC by default
 Please specify the appropriate offset
 (We're working on improving this)
@@ -424,14 +444,9 @@
 -0600 Central  Standard or Mountain Daylight
 -0700 Mountain Standard or Pacific Daylight
 -0800 Pacific  Standard
-");
-       while (1)
-       {
-#        $opt_tz_offset='local' unless length($opt_tz_offset);
-#        $opt_tz_offset=ask("Timezone offset (+/-####), or 'local' ($opt_tz_offset)")||'local';
-         $opt_tz_offset=$ENV{TZ} unless defined $opt_tz_offset || $ENV{TZ} !~ /[+-]\d\d\d\d/;
-         $opt_tz_offset='+0000'  unless defined $opt_tz_offset;
-         $opt_tz_offset=ask("Timezone offset (+/-####) ($opt_tz_offset)") || $opt_tz_offset;
+
+Timezone offset (+/-####) ($opt_tz_offset)
+") || $opt_tz_offset;
          last if $opt_tz_offset =~ /[+-]\d\d\d\d/;
        }
 
@@ -442,7 +457,7 @@
 
     ");
         $dd_user=ask("Username ($dd_user):")|| $dd_user || die "DataDirect Username Required\n";
-        $dd_pass=ask("
+        $dd_pass=ask_password("
 WARNING: Storing the password in the config file is not secure
 If password is blank, it will be prompted as needed(more secure)
 Unsecured password ('x':delete,default:<keep>,):")|| $dd_pass;
@@ -524,17 +539,15 @@
     {
         require Term::ReadKey;
         $| = 1;
-        print STDERR "Password for $dd_user: ";
-        Term::ReadKey::ReadMode('noecho');
-        chomp($dd_pass = <STDIN>);
-        Term::ReadKey::ReadMode('restore');
-        print STDERR "\n";
+        $dd_pass = ask_password( "Password for $dd_user: " );
     }
 
 #
 # Fetch data
 # 
-    print STDERR "Fetching from DataDirect\n";
+    my $bar = new XMLTV::ProgressBar( "Fetching from DataDirect", 1 )
+        if not $opt_quiet;
+        
     print STDERR "    dd_data is in $dd_data_name\n" if $DEBUG || $opt_dd_data;   
 
     sub SOAP::Transport::HTTP::Client::get_basic_credentials
@@ -563,7 +576,14 @@
     $dd_data->flush;
     $dd_data_size= -s $dd_data;
     $time = int(time() - $time);
-    printf STDERR "    Fetched %d k/bytes in %d seconds\n",$dd_data_size/1024,$time;
+     
+    if($bar) {
+        $bar->update();
+        $bar->finish();
+    }
+    
+    printf STDERR "Fetched %d k/bytes in %d seconds\n", $dd_data_size/1024,
+        $time;
 } # get data
 
 #
@@ -673,12 +693,6 @@
                   		          },
 	         });
 
-
-unless ($opt_quiet)
-{
-    if (Have_bar) { $bar = new Term::ProgressBar('loading data',$dd_data_size+1) }
-    else          { print STDERR "loading data "; }
-}
 seek($dd_data,0,0);  #rewind
 eval { $twig->parse( $dd_data ) };
 if ($@) {
@@ -694,9 +708,6 @@
     die "\nBad XML from DD, cannot continue\n";
 }
 
-$bar->update($dd_data_size+1) if $bar;
-print STDERR "\n" unless ($opt_quiet || $bar );
-
 $twig=undef;  # destroy twig (just in case)
 
 
@@ -766,51 +777,36 @@
                                    $lineups{$dd_lineup}{name},
                                    $lineups{$dd_lineup}{type});
                                            
-        $val = askQuestion("\nWhich Lineup? ($dd_lineup)",$val,@choices);
+        $val = ask_choice("\nWhich Lineup?", $val, @choices);
         $dd_lineup = (split(/\|/,$val))[0];
     }
     print CONF "lineup: $dd_lineup\n";
 
+    my @channels;
     foreach (@{$lineups{$dd_lineup}{map}})
     {
-       my $res='yes';
-       my $key=sprintf("%d %s",$_->{channel},$station{$_->{station}}{callSign});
-       $chan_found{$key}=1;
-       if ($opt_auto_config)
-       {
-            if (exists $chan_config{$key})
-            {
+        push @channels, sprintf( "%d %s",
+            $_->{channel}, $station{$_->{station}}{callSign} );
             }
-            elsif ($opt_auto_config eq 'add')
-            {
-                print STDERR "Adding new channel: $key\n";
-                $chan_config{$key}=1;
+    
+    my @questions;
+        
+    foreach my $chan ( @channels ) {
+        push @questions, "Add channel $chan? ";
             }
-            else
-            {
-                print STDERR "Ignoring new channel: $key\n";
-                $chan_config{$key}=0;
+    my @answers = ask_many_boolean( 1, @questions );
+        
+    my $i = 0;
+    for( my $i=0; $i <= $#channels; $i++ ) {
+        if( $answers[$i] ) {
+            print CONF "channel: ".$channels[$i]."\n";
+        } else {
+            print CONF "not channel: ".$channels[$i]."\n";
             }
        }
-       else
-       {
-            $res='no' if exists $chan_config{$key} and ! $chan_config{$key};
-            $res = askQuestion("Add channel $key?",$res,qw(yes no all none ));
-            $chan_config{$key}=1       if $res=~/^[ya]/i;
-            $chan_config{$key}=0       if $res=~/^[n]/i;
-            $opt_auto_config='add'     if $res=~/^all/i;
-            $opt_auto_config='ign'     if $res=~/^none/i;
-       } # ask question (or --auto-new )
-
-       print CONF ( $chan_config{$key} ? '' : 'not ' ), "channel: $key\n";
-    } # channel loop
 
-    foreach (sort keys %chan_config)
-    {
-        next if $chan_found{$_};
-        print STDERR "Channel '$_' no longer exists\n";
-    }
     close CONF;
+    say( "Configuration complete." );
     exit 0;
 } # --configure channel list
 
@@ -1134,13 +1130,12 @@
 seek($dd_data,0,0);  #rewind
 unless ($opt_quiet)
 {
-    if (Have_bar) { $bar = new Term::ProgressBar('Writing schedule',$dd_data_size+1) }
-    else          { print STDERR "Writing schedule"; }
+    $bar = new XMLTV::ProgressBar('writing schedule',$dd_data_size+1);
 }
 my $time=time();
 $twig->parse( $dd_data );
-print STDERR "\n" unless ($opt_quiet || $bar );
 $bar->update($dd_data_size+1) if $bar;
+$bar->finish();
 
 $writer->end();
 if (@messages)
@@ -1148,7 +1143,7 @@
     print STDERR "Message: $_\n" foreach @messages;
 }
 
-printf STDERR "Downloaded %d programs in %d seconds\n",$prog_count,time()-$start_time
+say( sprintf("Downloaded %d programs in %d seconds\n",$prog_count,time()-$start_time) )
     unless $opt_quiet;
 
 exit(0);
Index: grab/na_dd/tv_grab_na_dd.in
===================================================================
RCS file: /cvsroot/xmltv/xmltv/grab/na_dd/tv_grab_na_dd.in,v
retrieving revision 1.5
diff -u -B -b -w -r1.5 tv_grab_na_dd.in
--- grab/na_dd/tv_grab_na_dd.in	22 May 2004 20:22:30 -0000	1.5
+++ grab/na_dd/tv_grab_na_dd.in	2 Jun 2004 09:37:12 -0000
@@ -823,12 +823,12 @@
                                    $lineups{$dd_lineup}{name},
                                    $lineups{$dd_lineup}{type});
                                            
-        $val = askQuestion("\nWhich Lineup? ($dd_lineup)",$val,@choices);
+        $val = ask_choice("\nWhich Lineup? ($dd_lineup)",$val,@choices);
         $dd_lineup = (split(/\|/,$val))[0];
     }
     print CONF "lineup: $dd_lineup\n";
 
-        $opt_auto_config='add' if !$opt_auto_config && !askBooleanQuestion("
+        $opt_auto_config='add' if !$opt_auto_config && !ask_boolean("
 The preferred method for controlling the channel lineup is through
 the DataDirect web site, but you can omit channels here as well.
 Do you want to skip some channels?",0);
@@ -859,7 +859,7 @@
        else
        {
             $res='no' if exists $chan_config{$key} and ! $chan_config{$key};
-            $res = askQuestion("Add channel $key?",$res,qw(yes no all none ));
+            $res = ask_choice("Add channel $key?",$res,qw(yes no all none ));
             $chan_config{$key}=1       if $res=~/^[ya]/i;
             $chan_config{$key}=0       if $res=~/^[n]/i;
             $opt_auto_config='add'     if $res=~/^all/i;
Index: grab/na_icons/tv_grab_na_icons.in
===================================================================
RCS file: /cvsroot/xmltv/xmltv/grab/na_icons/tv_grab_na_icons.in,v
retrieving revision 1.2
diff -u -B -b -w -r1.2 tv_grab_na_icons.in
--- grab/na_icons/tv_grab_na_icons.in	23 May 2004 17:38:53 -0000	1.2
+++ grab/na_icons/tv_grab_na_icons.in	2 Jun 2004 09:37:12 -0000
@@ -139,7 +139,7 @@
     map {s/\xa0//g} @names;  # remove non-breaking spaces
 
     my $name=$names[0];    
-    $name=askQuestion("\nLineup?",$name,@names);
+    $name=ask_choice("\nLineup?",$name,@names);
 
     foreach (0..$#names)
     {
Index: grab/nl/tv_grab_nl.in
===================================================================
RCS file: /cvsroot/xmltv/xmltv/grab/nl/tv_grab_nl.in,v
retrieving revision 1.7
diff -u -B -b -w -r1.7 tv_grab_nl.in
--- grab/nl/tv_grab_nl.in	10 Apr 2004 22:00:31 -0000	1.7
+++ grab/nl/tv_grab_nl.in	2 Jun 2004 09:37:14 -0000
@@ -82,6 +82,7 @@
 use Date::Manip;
 use XMLTV;
 use XMLTV::Memoize;
+use XMLTV::ProgressBar;
 use XMLTV::Ask;
 use XMLTV::Config_file;
 use XMLTV::DST;
@@ -111,9 +112,6 @@
     }
 }
 
-# Use Term::ProgressBar if installed.
-use constant Have_bar => eval { require Term::ProgressBar; 1 };
-
 # Function prototypes.
 sub time_tot_str( $ );
 sub time_van_str( $ );
@@ -139,7 +137,7 @@
 # Get options, including undocumented --cache option.
 XMLTV::Memoize::check_argv('XMLTV::Get_nice::get_nice_aux');
 my ($opt_days, $opt_offset, $opt_help, $opt_output,
-    $opt_configure, $opt_config_file, $opt_quiet,
+    $opt_configure, $opt_config_file, $opt_gui, $opt_quiet,
     $opt_list_channels, $opt_slow, $opt_share);
 $opt_days   = 7; # default
 $opt_offset = 0; # default
@@ -148,6 +146,7 @@
 	   'help'          => \$opt_help,
 	   'configure'     => \$opt_configure,
 	   'config-file=s' => \$opt_config_file,
+           'gui:s'         => \$opt_gui,
 	   'output=s'      => \$opt_output,
 	   'quiet'         => \$opt_quiet,
 	   'slow'	   => \$opt_slow,
@@ -172,6 +171,9 @@
 my $OUR_SHARE_DIR = (defined $SHARE_DIR) ? "$SHARE_DIR/tv_grab_nl" : '.';
 (my $CHANNELS_FILE = "$OUR_SHARE_DIR/channels") =~ tr!/!/!s;
 
+XMLTV::Ask::init($opt_gui);
+XMLTV::ProgressBar::init($opt_gui);
+
 my $mode = XMLTV::Mode::mode('grab', # default
 			     $opt_configure => 'configure',
 			     $opt_list_channels => 'list-channels',
@@ -189,17 +191,18 @@
     # configuring it's useful to download at least something as a
     # check that the site is reachable etc.
     #
-    my $bar = new Term::ProgressBar('getting list of channels', 1)
-      if Have_bar && not $opt_quiet;
+    my $bar = new XMLTV::ProgressBar('getting list of channels', 1)
+      if not $opt_quiet;
     my %channels = get_channels(1);
     die 'no channels could be found' if (scalar(keys(%channels)) == 0);
-    update $bar if Have_bar && not $opt_quiet;
+    update $bar if not $opt_quiet;
+    $bar->finish();
 
     # Ask about each channel.
     my @chs = sort { $a <=> $b } keys %channels;
     my @names = map { $channels{$_} } @chs;
     my @qs = map { "add channel $_?" } @names;
-    my @want = askManyBooleanQuestions(1, @qs);
+    my @want = ask_many_boolean(1, @qs);
     foreach (@chs) {
 	my $w = shift @want;
 	warn("cannot read input, stopping channel questions"), last
@@ -246,15 +249,15 @@
    });
 
 if ($mode eq 'list-channels') {
-    my $bar = new Term::ProgressBar('getting list of channels', 1)
-      if Have_bar && not $opt_quiet;
+    my $bar = new XMLTV::ProgressBar('getting list of channels', 1)
+      if not $opt_quiet;
 
     # Don't do any page fetches - assume the contents of the channels
     # file has already been checked enough.
     #
     my %channels = get_channels(0);
     die 'no channels could be found' if (scalar(keys(%channels)) == 0);
-    update $bar if Have_bar && not $opt_quiet;
+    update $bar if not $opt_quiet;
 
     foreach my $ch_did (sort(keys %channels)) {
 	my $ch_name = $channels{$ch_did};
@@ -263,6 +266,7 @@
 				 'display-name' => [ [ $ch_name ] ] });
     }
     $writer->end();
+    $bar->finish();
     exit();
 }
 
@@ -335,16 +339,17 @@
 my %warned_ch_name; # suppress duplicate warnings
 
 my @summary_page_data;
-my $bar = new Term::ProgressBar('downloading summary', scalar @to_get)
-  if Have_bar && not $opt_quiet;
+my $bar = new XMLTV::ProgressBar('downloading summary', scalar @to_get)
+  if not $opt_quiet;
 foreach (@to_get) {
     my ($url, $day, $ch_xmltv_id, $ch_tvgids_id) = @$_;
     die if ref $url;
     push @summary_page_data,
       [ $url, $ch_xmltv_id,
 	[ process_summary_page($url, $day, $day, \$channels{$ch_tvgids_id}) ] ];
-    update $bar if Have_bar and not $opt_quiet;
+    update $bar if not $opt_quiet;
 }
+$bar->finish();
 
 # Now we've fetched the descriptions, we know the channel names.
 foreach $ch_did (@channels) {
@@ -467,8 +472,8 @@
 
 my @to_write;
 if ($opt_slow) {
-    $bar = new Term::ProgressBar('getting details', scalar @summary_programmes)
-      if Have_bar && not $opt_quiet;
+    $bar = new XMLTV::ProgressBar('getting details', scalar @summary_programmes)
+      if not $opt_quiet;
     foreach my $s (@summary_programmes) {
 	my $urls = delete $s->{url};
 	if (not defined $urls) {
@@ -563,6 +568,7 @@
 	push @to_write, $detailed;
 	update $bar if $bar;
     }
+    $bar->finish();
 }
 else {
     @to_write = @summary_programmes;
Index: grab/nl_wolf/tv_grab_nl_wolf
===================================================================
RCS file: /cvsroot/xmltv/xmltv/grab/nl_wolf/tv_grab_nl_wolf,v
retrieving revision 1.11
diff -u -B -b -w -r1.11 tv_grab_nl_wolf
--- grab/nl_wolf/tv_grab_nl_wolf	14 Feb 2004 22:10:31 -0000	1.11
+++ grab/nl_wolf/tv_grab_nl_wolf	2 Jun 2004 09:37:14 -0000
@@ -81,7 +81,7 @@
 my $url_base = "$URL_HOST$URL_DIR";
 
 # Returns a hash mapping YYYMMDD to URL.
-sub urls_by_date( $ ) {
+sub urls_by_date( $$$ ) {
     my $pkg = shift;
     my $index = $pkg->get($url_base);
     die "could not get index page $url_base, aborting\n"
@@ -135,29 +135,9 @@
 my $warned_bad_stop_time = 0;
 sub xml_from_data( $$ ) {
     my $pkg = shift;
-    my @lines = split /\n/, shift;
-    foreach (@lines) {
-	if (/<programme/) {
-	    # First change to numeric timezones.
-	    s{(start|stop)="(\d+) ([A-Z]+)"}
-	    {qq'$1="$2 ' . tz_to_num($3) . '"'}eg;
+    my $xml = shift;
 	    
-	    # Now remove stop times before start.  Only worry about
-	    # cases where the timezone is the same - we hope the
-	    # upstream data will be fixed by the next TZ changeover.
-	    #
-	    /start="(\d+) (\S+)"/ or next;
-	    my ($start, $tz) = ($1, $2);
-	    /stop="(\d+) \Q$tz\E"/ or next;
-	    my $stop = $1;
-	    if ($stop lt $start) {
-		warn "removing stop time before start time: $_"
-		  unless $warned_bad_stop_time++;
-		s/stop="[^""]+"\s*// or die;
-	    }
-	}
-    }
-    return join('', @lines);
+    return pkg->remove_early_stop_times( $xml );
 }
 
 Grab_XML_nl_wolf->go();
Index: grab/no/tv_grab_no
===================================================================
RCS file: /cvsroot/xmltv/xmltv/grab/no/tv_grab_no,v
retrieving revision 1.3
diff -u -B -b -w -r1.3 tv_grab_no
--- grab/no/tv_grab_no	10 Apr 2004 22:00:32 -0000	1.3
+++ grab/no/tv_grab_no	2 Jun 2004 09:37:14 -0000
@@ -172,7 +172,7 @@
     my @chs = sort keys %channels;
     my @names = map { $channels{$_} } @chs;
     my @qs = map { "add channel $_?" } @names;
-    my @want = askManyBooleanQuestions(1, @qs);
+    my @want = ask_many_boolean(1, @qs);
     foreach (@chs) {
 	my $w = shift @want;
 	warn("cannot read input, stopping channel questions"), last
Index: grab/pt/tv_grab_pt
===================================================================
RCS file: /cvsroot/xmltv/xmltv/grab/pt/tv_grab_pt,v
retrieving revision 1.3
diff -u -B -b -w -r1.3 tv_grab_pt
--- grab/pt/tv_grab_pt	13 Apr 2004 22:06:09 -0000	1.3
+++ grab/pt/tv_grab_pt	2 Jun 2004 09:37:15 -0000
@@ -163,7 +163,7 @@
     my @chs = sort keys %channels;
     my @names = map { $channels{$_} } @chs;
     my @qs = map { "add channel $_?" } @names;
-    my @want = askManyBooleanQuestions(1, @qs);
+    my @want = ask_many_boolean(1, @qs);
     foreach (@chs) {
 	my $w = shift @want;
 	warn("cannot read input, stopping channel questions"), last
Index: grab/se/tv_grab_se
===================================================================
RCS file: /cvsroot/xmltv/xmltv/grab/se/tv_grab_se,v
retrieving revision 1.8
diff -u -B -b -w -r1.8 tv_grab_se
--- grab/se/tv_grab_se	13 Apr 2004 22:20:55 -0000	1.8
+++ grab/se/tv_grab_se	2 Jun 2004 09:37:15 -0000
@@ -10,7 +10,7 @@
 
 tv_grab_se --help
 
-tv_grab_se [--config-file FILE] --configure
+tv_grab_se [--config-file FILE] --configure [--gui OPTION]
 
 tv_grab_se [--config-file FILE] [--output FILE] [--days N]
            [--offset N] [--quiet] [--debug]
@@ -34,6 +34,11 @@
 default is B<~/.xmltv/tv_grab_se.conf>.  This is the file written by
 B<--configure> and read when grabbing.
 
+B<--gui OPTION> Use this option to enable a graphical interface to be used.
+OPTION may be 'Tk' or 'GDialog', or left blank for the best available choice.
+Additional allowed values of OPTION are 'Term' for normal terminal output
+(default) and 'TermNoProgressBar' to disable the use of Term::ProgressBar.
+
 B<--output FILE> When grabbing, write output to FILE rather than
 standard output.
 
@@ -84,6 +89,7 @@
 
 use XMLTV;
 use XMLTV::Ask;
+use XMLTV::ProgressBar;
 use XMLTV::Config_file;
 use XMLTV::Get_nice qw(get_nice);
 use XMLTV::Memoize;
@@ -96,9 +102,6 @@
 use Getopt::Long;
 use URI;
 
-# Use Term::ProgressBar if installed.
-use constant Have_bar => eval { require Term::ProgressBar; 1 };
-
 # The timezone for Sweden during wintertime.
 use constant LOCAL_TZ => "+0100";
 
@@ -150,6 +153,7 @@
 my $opt = { days => 5,
             offset => 0,
             "config-file" => undef,
+            gui => undef,
             configure => 0,
             help => 0,
             quiet => 0,
@@ -161,6 +165,7 @@
                       days=i
                       offset=i
                       config-file=s
+                      gui
                       configure
                       help|h
                       quiet
@@ -175,7 +180,7 @@
   print << 'EOH';
 tv_grab_se --help
 
-tv_grab_se [--config-file FILE] --configure
+tv_grab_se [--config-file FILE] --configure [--gui OPTION]
 
 tv_grab_se [--config-file FILE] [--output FILE] [--days N]
            [--offset N] [--quiet] [--debug]
@@ -185,6 +190,9 @@
   exit(1);
 }
 
+XMLTV::Ask::init($opt->{'gui'});
+XMLTV::ProgressBar::init($opt->{'gui'});
+
 # XMLTV::DST says that we should do this...
 Date_Init('TZ=UTC');
 
@@ -240,10 +248,10 @@
 my $date = increase_date( $today, $opt->{offset} );
 
 my $bar = undef;
-$bar = new Term::ProgressBar( {
+$bar = new XMLTV::ProgressBar( {
     name => 'downloading listings',
     count => $opt->{days} * @channel_list
-    }) if Have_bar && (not $opt->{quiet}) && (not $opt->{debug});
+    }) if (not $opt->{quiet}) && (not $opt->{debug});
 
 for( my $i=0; $i < $opt->{days}; $i++ )
 {
@@ -258,7 +266,7 @@
 
     $date = increase_date( $date, 1 );
 }
-
+$bar->finish();
 $w->end();
 
 # Signal that something went wrong if there were warnings.
@@ -365,7 +373,7 @@
     open(CONF, ">$config_file") or die "cannot write to $config_file: $!";
 
     my @all = sort keys %channels;
-    my @wanted = askManyBooleanQuestions(1,
+    my @wanted = ask_many_boolean(1,
                                          map { "get channel $_?" }
                                          @all );
     foreach (@all) {
Index: grab/uk_rt/tv_grab_uk_rt.in
===================================================================
RCS file: /cvsroot/xmltv/xmltv/grab/uk_rt/tv_grab_uk_rt.in,v
retrieving revision 1.59
diff -u -B -b -w -r1.59 tv_grab_uk_rt.in
--- grab/uk_rt/tv_grab_uk_rt.in	10 Apr 2004 22:00:32 -0000	1.59
+++ grab/uk_rt/tv_grab_uk_rt.in	2 Jun 2004 09:37:17 -0000
@@ -11,7 +11,7 @@
 
 tv_grab_uk_rt --help
 
-tv_grab_uk_rt [--config-file FILE] --configure
+tv_grab_uk_rt [--config-file FILE] --configure [--gui OPTION]
 
 tv_grab_uk_rt [--config-file FILE] [--output FILE] [--quiet]
               [--days N] [--offset N]
@@ -36,6 +36,11 @@
 default is B<~/.xmltv/tv_grab_uk_rt.conf>.  This is the file written by
 B<--configure> and read when grabbing.
 
+B<--gui OPTION> Use this option to enable a graphical interface to be used.
+OPTION may be 'Tk' or 'GDialog', or left blank for the best available choice.
+Additional allowed values of OPTION are 'Term' for normal terminal output
+(default) and 'TermNoProgressBar' to disable the use of Term::ProgressBar.
+
 B<--output FILE> When grabbing, write output to FILE rather than
 standard output.
 
@@ -89,6 +94,7 @@
 require HTML::Entities;
 use XMLTV;
 use XMLTV::Memoize;
+use XMLTV::ProgressBar;
 use XMLTV::Ask;
 use XMLTV::DST;
 use XMLTV::Config_file;
@@ -179,6 +185,7 @@
     $opt_share,
     $opt_configure,
     $opt_config_file,
+    $opt_gui,
     $opt_offset,
     $opt_quiet,
     $opt_slow,
@@ -194,6 +201,7 @@
 	   'help'          => \$opt_help,
 	   'configure'     => \$opt_configure,
 	   'config-file=s' => \$opt_config_file,
+           'gui:s'         => \$opt_gui,
 	   'output=s'      => \$opt_output,
 	   'share=s'       => \$opt_share, # also undocumented
            'offset=i'      => \$opt_offset,
@@ -210,6 +218,9 @@
 die "--limit-details makes no sense without --slow\n"
   if defined $opt_detailtimerange and not $opt_slow;
 
+XMLTV::Ask::init($opt_gui);
+XMLTV::ProgressBar::init($opt_gui);
+
 # Date::Manip has a bug where 'now' will be wrong if you change the
 # timezone.  It won't be correctly converted from the system timezone
 # to the new one.  So we call parse_date('today midnight') _before_
@@ -389,8 +400,11 @@
 	}
     }
     my $days = @dates_to_get > 1 ? 'days' : 'day';
-    say('getting ' . (scalar @dates_to_get) . " $days of listings\n")
+    
+    my $bar = new XMLTV::ProgressBar('getting ' . (scalar @dates_to_get)
+        . " $days of listings", 1)
       unless $opt_quiet;
+    
     t 'getting dates:' . d \@dates_to_get;
 
     $writer->start({ 'source-info-url'     => "$BASE_URL/",
@@ -647,6 +661,9 @@
 	$writer->write_programme($_);
     }
     $writer->end();
+    
+    $bar->finish();
+    
 }
 
 
@@ -1767,10 +1784,12 @@
 
     XMLTV::Config_file::check_no_overwrite($config_file);
 
-    # FIXME turn into progress bar.
-    print STDERR "finding channels:\t";
+    my $bar = new XMLTV::ProgressBar('finding channels', 1)
+	  if not $opt_quiet;
+    
     my %channels = get_channels();
-    print STDERR "got ". (scalar keys %channels) . " done.\n" unless $opt_quiet;
+    
+    $bar->finish();
 
     # FIXME need to make directory
     open(CONF, ">$config_file") or die "cannot write to $config_file: $!";
@@ -1813,7 +1832,7 @@
 	}
 	elsif (@poss == 1) {
 	    my $ch = $poss[0];
-	    if (askBooleanQuestion('Add channel ' .
+	    if (ask_boolean('Add channel ' .
 				   $ch->{'display-name'}->[0]->[0] . '?', 1)) {
 		my $xmltv_id = $ch->{id};
 		unless ($chose_ch{$xmltv_id}++) {
@@ -1831,7 +1850,7 @@
 	    }
 	    my $none_option = 'None of the above are what I wanted';
 	    die 'silly channel name' if exists $dn_to_ch{$none_option};
-	    my $r = askQuestion('Which channel to add?',
+	    my $r = ask_choice('Which channel to add?',
 				$poss[0]->{'display-name'}->[0]->[0],
 				(sort keys %dn_to_ch), $none_option);
 	    die "cannot read input, aborting configuration\n"
Index: lib/Ask.pm
===================================================================
RCS file: /cvsroot/xmltv/xmltv/lib/Ask.pm,v
retrieving revision 1.12
diff -u -B -b -w -r1.12 Ask.pm
--- lib/Ask.pm	3 Jan 2004 14:52:53 -0000	1.12
+++ lib/Ask.pm	2 Jun 2004 09:37:17 -0000
@@ -7,9 +7,20 @@
 
 package XMLTV::Ask;
 use strict;
-use Carp qw(croak carp);
+use XMLTV::GUI;
 
-# Use Log::TraceMessages if installed, and choose graphical or not.
+use vars qw(@ISA @EXPORT);
+use Exporter;
+@ISA = qw(Exporter);
+@EXPORT = qw(ask
+             ask_password
+             ask_choice
+             ask_boolean
+             ask_many_boolean
+             say
+             );
+
+# Use Log::TraceMessages if installed.
 BEGIN {
     eval { require Log::TraceMessages };
     if ($@) {
@@ -20,22 +31,38 @@
 	*t = \&Log::TraceMessages::t;
 	*d = \&Log::TraceMessages::d;
     }
-
-    # For now we do graphical configuration only if the undocumented
-    # XMLTV_TK environment variable is set to a true value.
-    #
-    if ($ENV{XMLTV_TK}
-	and (defined($ENV{DISPLAY}) || $^O eq 'MSWin32')
-	and eval { require Tk }) {
-	require XMLTV::AskTk; XMLTV::AskTk->import;
-	*XMLTV::Ask:: = *XMLTV::AskTk::;
-    }
-    else {
-	require XMLTV::AskTerm; XMLTV::AskTerm->import;
-	*XMLTV::Ask:: = *XMLTV::AskTerm::;
     }
+
+my $real_class = 'XMLTV::Ask::Term';
+
+sub AUTOLOAD {
+        use vars qw($AUTOLOAD);
+        (my $method_name = $AUTOLOAD) =~ s/.*::(.*?)/$1/;
+        (my $real_class_path = $real_class.".pm") =~ s/::/\//g;
+        
+        require $real_class_path;
+        import $real_class_path;
+        
+        $real_class->$method_name(@_);       
 }
 
 
+# Must be called before we use this module if we want to use a gui.
+sub init( $ ) {
+        my $opt_gui = shift;
+        
+        # Ask the XMLTV::GUI module for the graphics type we will use  
+        my $gui_type = XMLTV::GUI::get_gui_type($opt_gui);
+        
+        if($gui_type =~ /^term/) {        
+                $real_class = 'XMLTV::Ask::Term';
+        } elsif($gui_type eq 'tk') {
+                $real_class = 'XMLTV::Ask::Tk';
+        } elsif($gui_type eq 'gdialog') {
+                $real_class = 'XMLTV::Ask::GDialog';
+        } else {
+                die "Unknown gui type: '$gui_type'.";
+        }
+}
 
 1;
Index: lib/exe_opt.pl
===================================================================
RCS file: /cvsroot/xmltv/xmltv/lib/exe_opt.pl,v
retrieving revision 1.7
diff -u -B -b -w -r1.7 exe_opt.pl
--- lib/exe_opt.pl	16 Mar 2004 22:43:08 -0000	1.7
+++ lib/exe_opt.pl	2 Jun 2004 09:37:17 -0000
@@ -14,7 +14,10 @@
 #
 print '-nologo
 -force
--add="XMLTV::Ask::Term;XMLTV::ProgressBar::Term"
+-add="XMLTV::Ask::Term;XMLTV::Ask::Tk"
+-add="XMLTV::ProgressBar::Term;XMLTV::ProgressBar::Tk;XMLTV::ProgressBar::None;"
+-add="XMLTV::GUI"
+-add="Tk::ProgressBar"
 -bind=libexpat.dll[file=\perl\site\lib\auto\XML\Parser\Expat\libexpat.dll,extract]
 -trim="Convert::EBCDIC;DB_File;Encode;HASH;HTML::FromText;Text::Iconv;Unicode::Map8;v5;URI/urn::isbn.pm;URI/urn::oid.pm;PerlIO/gzip.pm;HTML::FormatText"
 -info CompanyName="XMLTV Project http://membled.com/work/apps/xmltv/"
