Index: MANIFEST
===================================================================
RCS file: /cvsroot/xmltv/xmltv/MANIFEST,v
retrieving revision 1.94
diff -u -B -b -w -r1.94 MANIFEST
--- MANIFEST	17 Mar 2004 18:36:10 -0000	1.94
+++ MANIFEST	30 Mar 2004 17:21:39 -0000
@@ -4,8 +4,14 @@
 Uninstall.pm
 README
 lib/Ask.pm
-lib/AskTerm.pm
-lib/AskTk.pm
+lib/Ask/GDialog.pm
+lib/Ask/Term.pm
+lib/Ask/Tk.pm
+lib/ProgressBar.pm
+lib/ProgressBar/GDialog.pm
+lib/ProgressBar/None.pm
+lib/ProgressBar/Term.pm
+lib/ProgressBar/Tk.pm
 lib/XMLTV.pm.PL
 lib/XMLTV.pm.in
 lib/TZ.pm
Index: Makefile.PL
===================================================================
RCS file: /cvsroot/xmltv/xmltv/Makefile.PL,v
retrieving revision 1.193
diff -u -B -b -w -r1.193 Makefile.PL
--- Makefile.PL	21 Mar 2004 14:51:33 -0000	1.193
+++ Makefile.PL	30 Mar 2004 17:21:40 -0000
@@ -107,9 +107,16 @@
      'lib/Usage.pm'               => '$(INST_LIBDIR)/XMLTV/Usage.pm',
      'lib/Date.pm'                => '$(INST_LIBDIR)/XMLTV/Date.pm',
      'lib/Version.pm'             => '$(INST_LIBDIR)/XMLTV/Version.pm',
+     'lib/GUI.pm'                 => '$(INST_LIBDIR)/XMLTV/GUI.pm',
      'lib/Ask.pm'                 => '$(INST_LIBDIR)/XMLTV/Ask.pm',
-     'lib/AskTk.pm'               => '$(INST_LIBDIR)/XMLTV/AskTk.pm',
-     'lib/AskTerm.pm'             => '$(INST_LIBDIR)/XMLTV/AskTerm.pm',
+     'lib/Ask/GDialog.pm'         => '$(INST_LIBDIR)/XMLTV/Ask/GDialog.pm',
+     'lib/Ask/Term.pm'            => '$(INST_LIBDIR)/XMLTV/Ask/Term.pm',
+     'lib/Ask/Tk.pm'              => '$(INST_LIBDIR)/XMLTV/Ask/Tk.pm',
+     'lib/ProgressBar.pm'         => '$(INST_LIBDIR)/XMLTV/ProgressBar.pm',
+     'lib/ProgressBar/GDialog.pm' => '$(INST_LIBDIR)/XMLTV/ProgressBar/GDialog.pm',
+     'lib/ProgressBar/None.pm'    => '$(INST_LIBDIR)/XMLTV/ProgressBar/None.pm',
+     'lib/ProgressBar/Term.pm'    => '$(INST_LIBDIR)/XMLTV/ProgressBar/Term.pm',
+     'lib/ProgressBar/Tk.pm'      => '$(INST_LIBDIR)/XMLTV/ProgressBar/Tk.pm',
      'lib/Summarize.pm'           => '$(INST_LIBDIR)/XMLTV/Summarize.pm',
      'lib/IMDB.pm'                => '$(INST_LIBDIR)/XMLTV/IMDB.pm',
      'lib/Gunzip.pm'              => '$(INST_LIBDIR)/XMLTV/Gunzip.pm',
@@ -179,8 +186,8 @@
     *ask = sub { print "$_[0] yes\n"; 1 };
 }
 else {
-    do 'lib/AskTerm.pm' or die 'could not load lib/AskTerm.pm, aborting';
-    *ask = \&XMLTV::AskTerm::askBooleanQuestion;
+    do 'lib/Ask/Term.pm' or die 'could not load lib/Ask/Term.pm, aborting';
+    *ask = \&XMLTV::Ask::Term::askBooleanQuestion;
 }
 
 # Weird shit happens when you change things like PREFIX without
@@ -194,7 +201,7 @@
 
 END
   ;
-    if (! ask("Do you wish to continue anyway?", 0)) {
+    if (! ask(0, "Do you wish to continue anyway?", 0)) {
 	exit(1);
     }
 }
@@ -430,7 +437,7 @@
 		  "\n");
 }
 print STDERR "\n";
-if (not ask('Do you want to proceed with this configuration?', 1)) {
+if (not ask(0,'Do you want to proceed with this configuration?', 1)) {
     # Need to set {install} for each component by prompting.
     foreach my $info (@opt_components) {
 	my $missing = $info->{missing};
@@ -460,7 +467,7 @@
 	else { die }
 	
 	$info->{install} =
-	  ask($msg, not $missing);
+	  ask(0, $msg, not $missing);
     }
 }
 
Index: grab/de_tvtoday/tv_grab_de_tvtoday.in
===================================================================
RCS file: /cvsroot/xmltv/xmltv/grab/de_tvtoday/tv_grab_de_tvtoday.in,v
retrieving revision 1.3
diff -u -B -b -w -r1.3 tv_grab_de_tvtoday.in
--- grab/de_tvtoday/tv_grab_de_tvtoday.in	12 Mar 2004 17:32:11 -0000	1.3
+++ grab/de_tvtoday/tv_grab_de_tvtoday.in	30 Mar 2004 17:21:41 -0000
@@ -11,7 +11,7 @@
 
 tv_grab_de_tvtoday --help
 
-tv_grab_de_tvtoday [--config-file FILE] --configure
+tv_grab_de_tvtoday [--config-file FILE] --configure [--gui OPTION]
 
 tv_grab_de_tvtoday [--config-file FILE] [--output FILE] 
                    [--days N] [--offset N] [--force]
@@ -37,6 +37,11 @@
 default is B<~/.xmltv/tv_grab_de_tvtoday.conf>.  This is the file 
 written by B<--configure> and read when grabbing.
 
+B<--gui OPTION> Use this option to enable a graphical interface to be used.
+OPTION may be 'Tk' or 'GDialog', or left blank for the best available choice.
+Additional allowed values of OPTION are 'Term' for normal terminal output
+(default) and 'TermNoProgressBar' to disable the use of Term::ProgressBar.
+
 B<--output FILE> write to FILE rather than standard output.
 
 B<--days N> grab N days.  The default is seven.
@@ -100,6 +105,8 @@
 use URI::Escape;
 use XMLTV;
 use XMLTV::Ask;
+use XMLTV::GUI;
+use XMLTV::ProgressBar;
 use XMLTV::Europe_TZ;
 use XMLTV::Config_file;
 use XMLTV::Mode;
@@ -107,7 +114,7 @@
 use XMLTV::Memoize;
 use XMLTV::Usage <<END
 $0: get German television listings from www.tvtoday.de in XMLTV format
-To configure: $0 --configure [--config-file FILE]
+To configure: $0 --configure [--config-file FILE] [--gui OPTION]
 To grab data: $0 [--config-file FILE] [--output FILE] 
                  [--days N] [--offset N] [--force]
                  [--quiet] [--slow] [--nosqueezeout]
@@ -137,9 +144,6 @@
 my $debug = 0;
 $XMLTV::Get_nice::Delay = 0 if($debug);
 
-#-- Use Term::ProgressBar if installed.
-use constant Have_bar => eval { require Term::ProgressBar; 1; };
-
 #-- attributes of xmltv root element
 my $head = { 
     'source-data-url'      => 'http://www.tvtoday.de/tv/programm/programm.php',
@@ -156,6 +160,7 @@
 
 my $opt_configure;
 my $opt_config_file;
+my $opt_gui;
 my $opt_output;
 my $opt_days;
 my $opt_offset = 0;
@@ -170,6 +175,7 @@
 GetOptions(
     'configure'      => \$opt_configure,
     'config-file=s'  => \$opt_config_file,
+    'gui:s'          => \$opt_gui,
     'output=s'       => \$opt_output,
     'days=i'         => \$opt_days,
     'offset=i'       => \$opt_offset,
@@ -184,6 +190,12 @@
 
 usage(1) if $opt_help;
 
+# Ask the XMLTV::GUI module for the graphics type we will use  
+my $gui_type = XMLTV::GUI::get_gui_type($opt_gui);
+
+XMLTV::Ask::init($gui_type);
+XMLTV::ProgressBar::init($gui_type);
+
 #-- make sure offset+days arguments are within range
 die "you have to specify how many days to fetch, when trying to force"
   if($opt_force && !defined($opt_days));
@@ -373,16 +385,18 @@
 
 #-- write out <programme> tags
 my $numdays = $opt_days + $opt_offset - 1;
-my $bar = new Term::ProgressBar('grabbing', scalar(@requests) * $opt_days)
-  if Have_bar && not $opt_quiet;
+  
+my $bar = new XMLTV::ProgressBar('grabbing', scalar(@requests) * $opt_days)
+  if not $opt_quiet;
 
 foreach(@requests) {
     for (my $day = $opt_offset; $day <= $numdays; $day ++) {
 	grab_data($_, $day, $day == $numdays);
-	update $bar if Have_bar && not $opt_quiet;
+	update $bar if not $opt_quiet;
     }
 }
 
+$bar->finish();
 
 #-- hey, looks like we've finished ...
 $writer->end();
@@ -414,8 +428,8 @@
 	'lastday'   => $lday,
     };
 
-    print "grabbing channel $channels{$ch} (", UnixDate($grab->{day}, "%a, %Y-%m-%d"), ") ...\n"
-      unless(Have_bar || $opt_quiet);
+    say( "grabbing channel $channels{$ch} (", UnixDate($grab->{day}, "%a, %Y-%m-%d"), ") ..." )
+      unless($opt_quiet);
 	    
     while (defined($grab->{url})) {
 	my $tb = HTML::TreeBuilder->new();
Index: grab/dk/tv_grab_dk
===================================================================
RCS file: /cvsroot/xmltv/xmltv/grab/dk/tv_grab_dk,v
retrieving revision 1.12
diff -u -B -b -w -r1.12 tv_grab_dk
--- grab/dk/tv_grab_dk	10 Jan 2004 10:47:08 -0000	1.12
+++ grab/dk/tv_grab_dk	30 Mar 2004 17:21:41 -0000
@@ -10,7 +10,7 @@
 
 tv_grab_dk --help
 
-tv_grab_dk [--config-file FILE] --configure
+tv_grab_dk [--config-file FILE] --configure [--gui OPTION]
 
 tv_grab_dk [--config-file FILE] [--output FILE] [--days N]
            [--offset N] [--quiet]
@@ -32,6 +32,11 @@
 default is B<~/.xmltv/tv_grab_dk.conf>.  This is the file written by
 B<--configure> and read when grabbing.
 
+B<--gui OPTION> Use this option to enable a graphical interface to be used.
+OPTION may be 'Tk' or 'GDialog', or left blank for the best available choice.
+Additional allowed values of OPTION are 'Term' for normal terminal output
+(default) and 'TermNoProgressBar' to disable the use of Term::ProgressBar.
+
 B<--output FILE> write to FILE rather than standard output.
 
 B<--days N> grab N days.  The default is one week.
@@ -79,6 +84,8 @@
 
 use XMLTV;
 use XMLTV::Memoize;
+use XMLTV::GUI;
+use XMLTV::ProgressBar;
 use XMLTV::Ask;
 use XMLTV::Mode;
 use XMLTV::Config_file;
@@ -109,8 +116,7 @@
     }
 }
 
-# Use Term::ProgressBar if installed.
-use constant Have_bar => eval { require Term::ProgressBar; 1 };
+
 
 # Whether zero-length programmes should be included in the output.
 my $WRITE_ZERO_LENGTH = 0;
@@ -130,8 +136,8 @@
 # Get options
 XMLTV::Memoize::check_argv('XMLTV::Get_nice::get_nice_aux');
 my ($opt_days, $opt_offset, $opt_help, $opt_output,
-    $opt_configure, $opt_config_file, $opt_quiet,
-    $opt_list_channels);
+    $opt_configure, $opt_config_file, $opt_gui,
+    $opt_quiet, $opt_list_channels);
 $opt_days   = 7; # default
 $opt_offset = 0; # default
 GetOptions('days=i'        => \$opt_days,
@@ -139,6 +145,7 @@
 	   'help'          => \$opt_help,
 	   'configure'     => \$opt_configure,
 	   'config-file=s' => \$opt_config_file,
+           'gui:s'         => \$opt_gui,
 	   'output=s'      => \$opt_output,
 	   'quiet'         => \$opt_quiet,
 	   'list-channels' => \$opt_list_channels,
@@ -148,6 +155,12 @@
   if (defined $opt_days && $opt_days < 0);
 usage(1) if $opt_help;
 
+# Ask the XMLTV::GUI module for the graphics type we will use  
+my $gui_type = XMLTV::GUI::get_gui_type($opt_gui);
+
+XMLTV::Ask::init($gui_type);
+XMLTV::ProgressBar::init($gui_type);
+
 my $mode = XMLTV::Mode::mode('grab', # default
 			     $opt_configure => 'configure',
 			     $opt_list_channels => 'list-channels',
@@ -161,11 +174,11 @@
     XMLTV::Config_file::check_no_overwrite($config_file);
     open(CONF, ">$config_file") or die "cannot write to $config_file: $!";
     # find list of available channels
-    my $bar = new Term::ProgressBar('getting list of channels', 1)
-      if Have_bar && not $opt_quiet;
+    my $bar = new XMLTV::ProgressBar('getting list of channels', 1)
+      if not $opt_quiet;
     my %channels = get_channels();
     die 'no channels could be found' if (scalar(keys(%channels)) == 0);
-    update $bar if Have_bar && not $opt_quiet;
+    update $bar if not $opt_quiet;
 
     # Ask about each channel.
     my @chs = sort keys %channels;
@@ -185,6 +198,8 @@
 	# TODO don't store display-name in config file.
     }
 
+    $bar->finish() if not $opt_quiet;
+    
     close CONF or warn "cannot close $config_file: $!";
     say("Finished configuration.");
 
@@ -219,11 +234,11 @@
    });
 
 if ($opt_list_channels) {
-    my $bar = new Term::ProgressBar('getting list of channels', 1)
-      if Have_bar && not $opt_quiet;
+    my $bar = new XMLTV::ProgressBar('getting list of channels', 1)
+      if not $opt_quiet;
     my %channels = get_channels();
     die 'no channels could be found' if (scalar(keys(%channels)) == 0);
-    update $bar if Have_bar && not $opt_quiet;
+    update $bar if not $opt_quiet;
 
     foreach my $ch_did (sort(keys %channels)) {
 	my $ch_name = $channels{$ch_did};
@@ -231,6 +246,7 @@
 	$writer->write_channel({ id => $ch_xid,
 				 'display-name' => [ [ $ch_name ] ] });
     }
+    $bar->finish() if not $opt_quiet;
     $writer->end();
     exit();
 }
@@ -290,16 +306,17 @@
 
 my %warned_ch_name; # suppress duplicate warnings
 
-my $bar = new Term::ProgressBar('fetching data', scalar @to_get)
-  if Have_bar && not $opt_quiet;
+my $bar = new XMLTV::ProgressBar('fetching data', scalar @to_get)
+  if not $opt_quiet;
 my @to_get_detailed;
 my $num_detailed = 0;
 foreach (@to_get) {
     my ($url, $date, $ch_xmltv_id, $ch_tvgids_id) = @$_;
     t "going to get $ch_xmltv_id for $date";
     process_listings_page($writer, $ch_xmltv_id, $url, $date);
-    update $bar if Have_bar && not $opt_quiet;
+    update $bar if not $opt_quiet;
 }
+$bar->finish() if not $opt_quiet;
 $writer->end();
 
 ######################################################################
Index: grab/es/tv_grab_es
===================================================================
RCS file: /cvsroot/xmltv/xmltv/grab/es/tv_grab_es,v
retrieving revision 1.24
diff -u -B -b -w -r1.24 tv_grab_es
--- grab/es/tv_grab_es	14 Feb 2004 17:53:33 -0000	1.24
+++ grab/es/tv_grab_es	30 Mar 2004 17:21:43 -0000
@@ -10,7 +10,7 @@
 
 tv_grab_es --help
 
-tv_grab_es [--config-file FILE] --configure
+tv_grab_es [--config-file FILE] --configure [--gui OPTION]
 
 tv_grab_es [--config-file FILE] [--output FILE] [--days N]
            [--offset N] [--quiet]
@@ -38,6 +38,11 @@
 default is B<~/.xmltv/tv_grab_es.conf>.  This is the file written by
 B<--configure> and read when grabbing.
 
+B<--gui OPTION> Use this option to enable a graphical interface to be used.
+OPTION may be 'Tk' or 'GDialog', or left blank for the best available choice.
+Additional allowed values of OPTION are 'Term' for normal terminal output
+(default) and 'TermNoProgressBar' to disable the use of XMLTV::ProgressBar.
+
 B<--output FILE> write to FILE rather than standard output.
 
 B<--days N> grab N days.  The default is 3.
@@ -97,6 +102,8 @@
 
 use XMLTV;
 use XMLTV::Memoize;
+use XMLTV::GUI;
+use XMLTV::ProgressBar;
 use XMLTV::Ask;
 use XMLTV::Config_file;
 use XMLTV::Europe_TZ;
@@ -113,9 +120,6 @@
 END
   ;
 
-# Use Term::ProgressBar if installed.
-use constant Have_bar => eval { require Term::ProgressBar; 1 };
-
 # Attributes of the root element in output.
 my $HEAD = { 'source-info-url'     => 'http://www.elpais.es/parrillatv/portada.html',
 	     'source-data-url'     => "http://www.elpais.es/parrillatv/resultados.html",
@@ -138,8 +142,8 @@
 # Get options, including undocumented --cache option.
 XMLTV::Memoize::check_argv('XMLTV::Get_nice::get_nice_aux');
 my ($opt_days, $opt_offset, $opt_help, $opt_output,
-    $opt_configure, $opt_config_file, $opt_quiet,
-    $opt_list_channels);
+    $opt_configure, $opt_config_file, $opt_gui,
+    $opt_quiet, $opt_list_channels);
 $opt_days  = 3; # default
 $opt_offset = 0; # default
 $opt_quiet  = 0; # default
@@ -148,6 +152,7 @@
 	   'help'          => \$opt_help,
 	   'configure'     => \$opt_configure,
 	   'config-file=s' => \$opt_config_file,
+           'gui:s'         => \$opt_gui,
 	   'output=s'      => \$opt_output,
 	   'quiet'         => \$opt_quiet,
 	   'list-channels' => \$opt_list_channels
@@ -157,6 +162,12 @@
   if (defined $opt_days && $opt_days < 0);
 usage(1) if $opt_help;
 
+# Ask the XMLTV::GUI module for the graphics type we will use  
+my $gui_type = XMLTV::GUI::get_gui_type($opt_gui);
+
+XMLTV::Ask::init($gui_type);
+XMLTV::ProgressBar::init($gui_type);
+
 my $mode = XMLTV::Mode::mode('grab', # default
 			     $opt_configure => 'configure',
 			     $opt_list_channels => 'list-channels',
@@ -285,14 +296,15 @@
 # This progress bar is for both downloading and parsing.  Maybe
 # they could be separate.
 #
-my $bar = new Term::ProgressBar('getting listings', scalar @to_get)
-  if Have_bar && not $opt_quiet;
+my $bar = new XMLTV::ProgressBar('getting listings', scalar @to_get)
+  if not $opt_quiet;
 foreach (@to_get) {
 	foreach (process_table($_->[0], $_->[1], $_->[2])) {
 		$writer->write_programme($_);
 	}
-	update $bar if Have_bar && not $opt_quiet;
+	update $bar if not $opt_quiet;
 }
+$bar->finish() if not $opt_quiet;
 $writer->end();
 
 ######################################################################
@@ -482,8 +494,8 @@
 
 # get channel listing
 sub get_channels {
-    my $bar = new Term::ProgressBar('getting list of channels', 1)
-	if Have_bar && not $opt_quiet;
+    my $bar = new XMLTV::ProgressBar('getting list of channels', 1)
+	if not $opt_quiet;
     my %channels;
     my $url="http://www.elpais.es/parrillatv/portada.html";
     t $url;
@@ -518,7 +530,8 @@
 	}
     }
     die "no channels could be found" if not keys %channels;
-    update $bar if Have_bar && not $opt_quiet;
+    update $bar if not $opt_quiet;
+    $bar->finish() if not $opt_quiet;
     return %channels;
 }
 
Index: grab/es_digital/tv_grab_es_digital
===================================================================
RCS file: /cvsroot/xmltv/xmltv/grab/es_digital/tv_grab_es_digital,v
retrieving revision 1.2
diff -u -B -b -w -r1.2 tv_grab_es_digital
--- grab/es_digital/tv_grab_es_digital	6 Mar 2004 11:35:18 -0000	1.2
+++ grab/es_digital/tv_grab_es_digital	30 Mar 2004 17:21:43 -0000
@@ -10,7 +10,7 @@
 
 tv_grab_es_digital --help
 
-tv_grab_es_digital [--config-file FILE] --configure
+tv_grab_es_digital [--config-file FILE] --configure [--gui OPTION]
 
 tv_grab_es_digital [--config-file FILE] [--output FILE] [--days N]
            [--offset N] [--quiet]
@@ -36,6 +36,11 @@
 default is B<~/.xmltv/tv_grab_es_digital.conf>.  This is the file written by
 B<--configure> and read when grabbing.
 
+B<--gui OPTION> Use this option to enable a graphical interface to be used.
+OPTION may be 'Tk' or 'GDialog', or left blank for the best available choice.
+Additional allowed values of OPTION are 'Term' for normal terminal output
+(default) and 'TermNoProgressBar' to disable the use of Term::ProgressBar.
+
 B<--output FILE> write to FILE rather than standard output.
 
 B<--days N> grab N days.  The default is 3.
@@ -93,6 +98,8 @@
 
 use XMLTV;
 use XMLTV::Memoize;
+use XMLTV::GUI;
+use XMLTV::ProgressBar;
 use XMLTV::Ask;
 use XMLTV::Config_file;
 use XMLTV::Europe_TZ;
@@ -108,9 +115,6 @@
 END
   ;
 
-# Use Term::ProgressBar if installed.
-use constant Have_bar => eval { require Term::ProgressBar; 1 };
-
 # Attributes of the root element in output.
 my $HEAD = { 'source-info-url'     => 'http://http://www.plus.es/codigo/television/rejilla_dia_dia.asp',
 	     'source-data-url'     => "http://http://www.plus.es/codigo/television/rejilla_dia_dia.asp",
@@ -133,8 +137,8 @@
 # Get options, including undocumented --cache option.
 XMLTV::Memoize::check_argv('XMLTV::Get_nice::get_nice_aux');
 my ($opt_days, $opt_offset, $opt_help, $opt_output,
-    $opt_configure, $opt_config_file, $opt_quiet,
-    $opt_list_channels);
+    $opt_configure, $opt_config_file, $opt_gui,
+    $opt_quiet, $opt_list_channels);
 $opt_days  = 3; # default
 $opt_offset = 0; # default
 $opt_quiet  = 0; # default
@@ -143,6 +147,7 @@
 	   'help'          => \$opt_help,
 	   'configure'     => \$opt_configure,
 	   'config-file=s' => \$opt_config_file,
+       'gui:s'         => \$opt_gui,
 	   'output=s'      => \$opt_output,
 	   'quiet'         => \$opt_quiet,
 	   'list-channels' => \$opt_list_channels
@@ -152,6 +157,12 @@
   if (defined $opt_days && $opt_days < 0);
 usage(1) if $opt_help;
 
+# Ask the XMLTV::GUI module for the graphics type we will use  
+my $gui_type = XMLTV::GUI::get_gui_type($opt_gui);
+
+XMLTV::Ask::init($gui_type);
+XMLTV::ProgressBar::init($gui_type);
+
 my $mode = XMLTV::Mode::mode('grab', # default
 			     $opt_configure => 'configure',
 			     $opt_list_channels => 'list-channels',
@@ -278,16 +289,18 @@
 # This progress bar is for both downloading and parsing.  Maybe
 # they could be separate.
 #
-my $bar = new Term::ProgressBar('getting listings', scalar @to_get)
-  if Have_bar && not $opt_quiet;
+my $bar = new XMLTV::ProgressBar('getting listings', scalar @to_get)
+  if not $opt_quiet;
 foreach (@to_get) {
 	foreach (process_table($_->[0], $_->[1], $_->[2])) {
 		$writer->write_programme($_);
 	}
-	update $bar if Have_bar && not $opt_quiet;
+	update $bar if not $opt_quiet;
 }
 $writer->end();
 
+$bar->finish();
+
 ######################################################################
 # subroutine definitions
 
@@ -480,8 +493,8 @@
 
 # get channel listing
 sub get_channels {
-    my $bar = new Term::ProgressBar('getting list of channels', 1)
-	if Have_bar && not $opt_quiet;
+    my $bar = new XMLTV::ProgressBar('getting list of channels', 1)
+	if not $opt_quiet;
     my %channels;
     my $url="http://www.plus.es/codigo/television/plataformas/digitalplus/portada2.asp";
     t $url;
@@ -517,7 +530,8 @@
 	}
     }
     die "no channels could be found" if not keys %channels;
-    update $bar if Have_bar && not $opt_quiet;
+    update $bar if not $opt_quiet;
+    $bar->finish();
     return %channels;
 }
 
Index: grab/fi/tv_grab_fi
===================================================================
RCS file: /cvsroot/xmltv/xmltv/grab/fi/tv_grab_fi,v
retrieving revision 1.30
diff -u -B -b -w -r1.30 tv_grab_fi
--- grab/fi/tv_grab_fi	1 Feb 2004 20:38:07 -0000	1.30
+++ grab/fi/tv_grab_fi	30 Mar 2004 17:21:43 -0000
@@ -10,7 +10,7 @@
 
 tv_grab_fi --help
 
-tv_grab_fi [--config-file FILE] --configure
+tv_grab_fi [--config-file FILE] --configure [--gui OPTION]
 
 tv_grab_fi [--config-file FILE] [--output FILE] [--days N]
            [--offset N] [--quiet]
@@ -35,6 +35,11 @@
 default is B<~/.xmltv/tv_grab_fi.conf>.  This is the file written by
 B<--configure> and read when grabbing.
 
+B<--gui OPTION> Use this option to enable a graphical interface to be used.
+OPTION may be 'Tk' or 'GDialog', or left blank for the best available choice.
+Additional allowed values of OPTION are 'Term' for normal terminal output
+(default) and 'TermNoProgressBar' to disable the use of XMLTV::ProgressBar.
+
 B<--output FILE> write to FILE rather than standard output.
 
 B<--days N> grab N days.  The default is ten.
@@ -77,6 +82,8 @@
 
 use XMLTV;
 use XMLTV::Memoize;
+use XMLTV::GUI;
+use XMLTV::ProgressBar;
 use XMLTV::Ask;
 use XMLTV::Config_file;
 use XMLTV::Europe_TZ;
@@ -93,9 +100,6 @@
 END
   ;
 
-# Use Term::ProgressBar if installed.
-use constant Have_bar => eval { require Term::ProgressBar; 1 };
-
 # Attributes of the root element in output.
 my $HEAD = { 'source-info-url'     => 'http://www.katso.fi/',
 	     'source-data-url'     => "http://www.katso.fi/tvopas",
@@ -122,8 +126,8 @@
 # Get options, including undocumented --cache option.
 XMLTV::Memoize::check_argv('XMLTV::Get_nice::get_nice_aux');
 my ($opt_days, $opt_offset, $opt_help, $opt_output,
-    $opt_configure, $opt_config_file, $opt_quiet,
-    $opt_list_channels);
+    $opt_configure, $opt_config_file, $opt_gui,
+    $opt_quiet, $opt_list_channels);
 $opt_days  = 10; # default
 $opt_offset = 0; # default
 $opt_quiet  = 0; # default
@@ -132,6 +136,7 @@
 	   'help'          => \$opt_help,
 	   'configure'     => \$opt_configure,
 	   'config-file=s' => \$opt_config_file,
+           'gui:s'         => \$opt_gui,
 	   'output=s'      => \$opt_output,
 	   'quiet'         => \$opt_quiet,
 	   'list-channels' => \$opt_list_channels,
@@ -141,6 +146,12 @@
   if (defined $opt_days && $opt_days < 0);
 usage(1) if $opt_help;
 
+# Ask the XMLTV::GUI module for the graphics type we will use  
+my $gui_type = XMLTV::GUI::get_gui_type($opt_gui);
+
+XMLTV::Ask::init($gui_type);
+XMLTV::ProgressBar::init($gui_type);
+
 my $mode = XMLTV::Mode::mode('grab', # default
 			     $opt_configure => 'configure',
 			     $opt_list_channels => 'list-channels',
@@ -266,14 +277,15 @@
 # This progress bar is for both downloading and parsing.  Maybe
 # they could be separate stages.
 #
-my $bar = new Term::ProgressBar('getting listings', scalar @to_get)
-  if Have_bar && not $opt_quiet;
+my $bar = new XMLTV::ProgressBar('getting listings', scalar @to_get)
+  if not $opt_quiet;
 foreach (@to_get) {
     foreach (process_table($_->[0], $_->[1], $_->[2])) {
 	$writer->write_programme($_);
     }
-    update $bar if Have_bar && not $opt_quiet;
+    update $bar if not $opt_quiet;
 }
+$bar->finish() if not $opt_quiet;
 $writer->end();
 
 ######################################################################
@@ -488,8 +500,8 @@
 
 # get channel listing
 sub get_channels {
-    my $bar = new Term::ProgressBar('getting list of channels', 1)
-      if Have_bar && not $opt_quiet;
+    my $bar = new XMLTV::ProgressBar('getting list of channels', 1)
+      if not $opt_quiet;
     my %channels;
     my $url="http://www.katso.fi/tvopas";
     my $local_data=get_nice($url);
@@ -531,7 +543,8 @@
 	}
     }
     die "no channels could be found" if not keys %channels;
-    update $bar if Have_bar && not $opt_quiet;
+    update $bar if not $opt_quiet;
+    $bar->finish() if not $opt_quiet;
     return %channels;
 }
 
Index: grab/fr/tv_grab_fr
===================================================================
RCS file: /cvsroot/xmltv/xmltv/grab/fr/Attic/tv_grab_fr,v
retrieving revision 1.1
diff -u -B -b -w -r1.1 tv_grab_fr
--- grab/fr/tv_grab_fr	23 Feb 2004 20:55:57 -0000	1.1
+++ grab/fr/tv_grab_fr	30 Mar 2004 17:21:44 -0000
@@ -6,7 +6,7 @@
 
 =head1 SYNOPSIS
 
-To configure: tv_grab_fr --configure [--config-file FILE]
+To configure: tv_grab_fr --configure [--config-file FILE] [--gui OPTION]
 To grab listings: tv_grab_fr [--output FILE] [--quiet]
 Slower, detailed grab: tv_grab_fr --slow [--output FILE] [--days N] [--offset N] [--quiet]
 Help: tv_grab_fr --help
@@ -23,6 +23,11 @@
 B<--configure> Grab channels informations from the website and ask for
 channel type and names.
 
+B<--gui OPTION> Use this option to enable a graphical interface to be used.
+OPTION may be 'Tk' or 'GDialog', or left blank for the best available choice.
+Additional allowed values of OPTION are 'Term' for normal terminal output
+(default) and 'TermNoProgressBar' to disable the use of Term::ProgressBar.
+
 B<--output FILE> write to FILE rather than standard output.
 
 B<--days N> grab N days starting from today, rather than as many as
@@ -60,9 +65,6 @@
 END
   ;
 
-# Use Term::ProgressBar if installed.
-use constant Have_bar => eval { require Term::ProgressBar; 1 };
-
 use warnings;
 use strict;
 use XMLTV::Version '$Id$ ';
@@ -75,6 +77,8 @@
 use XMLTV;
 use XMLTV::Memoize;
 use XMLTV::Ask;
+use XMLTV::GUI;
+use XMLTV::ProgressBar;
 use XMLTV::Mode;
 use XMLTV::Config_file;
 use XMLTV::Europe_TZ;
@@ -106,7 +110,7 @@
 # Global variables allocation according to options
 #***************************************************************************
 XMLTV::Memoize::check_argv('get_page_aux');
-my ($opt_days,  $opt_help,  $opt_output,  $opt_offset,  $opt_quiet,  $opt_list_channels, $opt_config_file, $opt_configure, $opt_slow, $opt_licons);
+my ($opt_days,  $opt_help,  $opt_output,  $opt_offset,  $opt_gui, $opt_quiet,  $opt_list_channels, $opt_config_file, $opt_configure, $opt_slow, $opt_licons);
 $opt_quiet  = 0;
 # The website is able to store up to nine days from now
 my $default_opt_days = 9;
@@ -118,6 +122,7 @@
      'quiet'     => \$opt_quiet,
      'configure' => \$opt_configure,
      'config-file=s' => \$opt_config_file,
+     'gui:s'     => \$opt_gui,
      'list-channels' => \$opt_list_channels,
      'slow' => \$opt_slow
     )
@@ -130,6 +135,12 @@
 die 'Cannot get more than one day before current day' if (defined $opt_offset && $opt_offset < -1);
 usage(1) if $opt_help;
 
+# Ask the XMLTV::GUI module for the graphics type we will use  
+my $gui_type = XMLTV::GUI::get_gui_type($opt_gui);
+
+XMLTV::Ask::init($gui_type);
+XMLTV::ProgressBar::init($gui_type);
+
 if (not $opt_slow) {
     # Certain options are ignored in fast mode.
     my %slow_options = (days => $opt_days,
@@ -246,9 +257,9 @@
     my @gtqs = map { "Get channels type : $_?" } @gts;
     my @gtwant = askManyBooleanQuestions(1, @gtqs);
 
-    my $bar = new Term::ProgressBar('getting channel lists',
+    my $bar = new XMLTV::ProgressBar('getting channel lists',
                                     scalar grep { $_ } @gtwant)
-                    if Have_bar && not $opt_quiet;
+                    if not $opt_quiet;
     my %channels_for;
     foreach my $i (0 .. $#gts) {
         my ($gt, $gtw, $gtname) = ($gts[$i], $gtwant[$i], $gtnames[$i]);
@@ -256,8 +267,9 @@
         my %channels = get_channels( $gtname );
         die 'No channels could be found' if not %channels;
         $channels_for{$gt} = \%channels;
-        update $bar if Have_bar && not $opt_quiet;
+        update $bar if not $opt_quiet;
     }
+    $bar->finish();
 
     my %asked;
     foreach (@gts) {
@@ -384,13 +396,14 @@
         push @to_get, [ $url, $chid, $_ ];
     }
 }
-my $bar = new Term::ProgressBar('getting listings', scalar @to_get)  if Have_bar && not $opt_quiet;
+my $bar = new XMLTV::ProgressBar('getting listings', scalar @to_get)  if not $opt_quiet;
 foreach (@to_get) {
     my ($url, $chid, $slot ) = @$_;
     process_channel_grid_page($writer, $chid, $url, $slot);
-    update $bar if Have_bar && not $opt_quiet;
+    update $bar if not $opt_quiet;
 }
 $writer->end();
+$bar->finish();
 
 #***************************************************************************
 # Specific functions for grabbing information
Index: grab/hu/tv_grab_hu
===================================================================
RCS file: /cvsroot/xmltv/xmltv/grab/hu/tv_grab_hu,v
retrieving revision 1.12
diff -u -B -b -w -r1.12 tv_grab_hu
--- grab/hu/tv_grab_hu	1 Feb 2004 20:38:07 -0000	1.12
+++ grab/hu/tv_grab_hu	30 Mar 2004 17:21:44 -0000
@@ -10,7 +10,7 @@
 
 tv_grab_hu --help
 
-tv_grab_hu [--config-file FILE] --configure
+tv_grab_hu [--config-file FILE] --configure [--gui OPTION]
 
 tv_grab_hu [--config-file FILE] [--output FILE] [--days N]
            [--offset N] [--quiet]
@@ -33,6 +33,11 @@
 default is B<~/.xmltv/tv_grab_hu.conf>.  This is the file written by
 B<--configure> and read when grabbing.
 
+B<--gui OPTION> Use this option to enable a graphical interface to be used.
+OPTION may be 'Tk' or 'GDialog', or left blank for the best available choice.
+Additional allowed values of OPTION are 'Term' for normal terminal output
+(default) and 'TermNoProgressBar' to disable the use of XMLTV::ProgressBar.
+
 B<--output FILE> write to FILE rather than standard output.
 
 B<--days N> grab N days.  The default is eight.
@@ -76,6 +81,8 @@
 
 use XMLTV;
 use XMLTV::Memoize;
+use XMLTV::GUI;
+use XMLTV::ProgressBar;
 use XMLTV::Ask;
 use XMLTV::Europe_TZ;
 use XMLTV::Get_nice;
@@ -85,16 +92,13 @@
 # Todo: perhaps we should internationalize messages and docs?
 use XMLTV::Usage <<END
 $0: get Hungarian television listings in XMLTV format
-To configure: $0 --configure [--config-file FILE]
+To configure: $0 --configure [--config-file FILE] [--gui]
 To grab listings: $0 [--config-file FILE] [--output FILE] [--days N]
         [--offset N] [--quiet]
 To list channels: $0 --list-channels
 END
   ;
 
-# Use Term::ProgressBar if installed.
-use constant Have_bar => eval { require Term::ProgressBar; 1 };
-
 # Attributes of the root element in output.
 my $HEAD = { 'source-info-url'     => 'http://www.port.hu/',
 	     'source-data-url'     => "http://www.port.hu/tv/",
@@ -120,8 +124,8 @@
 # Get options, including undocumented --cache option.
 XMLTV::Memoize::check_argv('XMLTV::Get_nice::get_nice_aux');
 my ($opt_days, $opt_offset, $opt_help, $opt_output,
-    $opt_configure, $opt_config_file, $opt_quiet,
-    $opt_list_channels);
+    $opt_configure, $opt_config_file, $opt_gui,,
+    $opt_quiet, $opt_list_channels);
 $opt_days = 8; # default
 $opt_offset = 0; # default
 $opt_quiet = 0; # default
@@ -130,6 +134,7 @@
 	   'help'          => \$opt_help,
 	   'configure'     => \$opt_configure,
 	   'config-file=s' => \$opt_config_file,
+           'gui:s'         => \$opt_gui,
 	   'output=s'      => \$opt_output,
 	   'quiet'         => \$opt_quiet,
 	   'list-channels' => \$opt_list_channels,
@@ -138,6 +143,13 @@
 die 'number of days must not be negative'
   if (defined $opt_days && $opt_days < 0);
 usage(1) if $opt_help;
+
+# Ask the XMLTV::GUI module for the graphics type we will use  
+my $gui_type = XMLTV::GUI::get_gui_type($opt_gui);
+
+XMLTV::Ask::init($gui_type);
+XMLTV::ProgressBar::init($gui_type);
+
 my $mode = XMLTV::Mode::mode('grab', # default
 			     $opt_configure => 'configure',
 			     $opt_list_channels => 'list-channels',
@@ -265,8 +277,8 @@
 # This progress bar is for both downloading and parsing.  Maybe
 # they could be separate stages.
 #
-my $bar = new Term::ProgressBar('getting listings', @days * @channels)
-  if Have_bar && not $opt_quiet;
+my $bar = new XMLTV::ProgressBar('getting listings', @days * @channels)
+  if not $opt_quiet;
 foreach my $d (@days) {
     my ($day, $i) = @$d;
     my $some_success = 0;
@@ -274,13 +286,15 @@
 	my @ps = process_table($day, xid($ch_did), $ch_did, $i);
 	$some_success = 1 if @ps;
 	$writer->write_programme($_) foreach @ps;
-	update $bar if Have_bar && not $opt_quiet;
+	update $bar if not $opt_quiet;
     }
     if (@channels and not $some_success) {
 	warn "failed to get any listings for day $i, stopping\n";
 	last;
     }
+    
 }
+$bar->finish() if not $opt_quiet;
 $writer->end();
 
 ######################################################################
@@ -465,8 +479,8 @@
 
 # get channel listing
 sub get_channels {
-    my $bar = new Term::ProgressBar('getting list of channels', 1)
-	if Have_bar && not $opt_quiet;
+    my $bar = new XMLTV::ProgressBar('getting list of channels', 1)
+	if not $opt_quiet;
     my %channels;
     my $url="http://www.port.hu/pls/tv/tv.prog";
     my $local_data=get_nice($url);
@@ -502,7 +516,8 @@
 	}
     }
     die "no channels could be found" if not keys %channels;
-    update $bar if Have_bar && not $opt_quiet;
+    update $bar if not $opt_quiet;
+    $bar->finish() if not $opt_quiet;
     return %channels;
 }
 
Index: grab/it/tv_grab_it.in
===================================================================
RCS file: /cvsroot/xmltv/xmltv/grab/it/tv_grab_it.in,v
retrieving revision 1.23
diff -u -B -b -w -r1.23 tv_grab_it.in
--- grab/it/tv_grab_it.in	1 Feb 2004 20:38:07 -0000	1.23
+++ grab/it/tv_grab_it.in	30 Mar 2004 17:21:45 -0000
@@ -11,7 +11,7 @@
 
 tv_grab_it --help
 
-tv_grab_it [--config-file FILE] --configure
+tv_grab_it [--config-file FILE] --configure [--gui OPTION]
 
 tv_grab_it [--config-file FILE] [--output FILE] [--days N]
            [--offset N] [--quiet] [--slow]
@@ -33,6 +33,11 @@
 default is B<~/.xmltv/tv_grab_it.conf>.  This is the file written
 by B<--configure> and read when grabbing.
 
+B<--gui OPTION> Use this option to enable a graphical interface to be used.
+OPTION may be 'Tk' or 'GDialog', or left blank for the best available choice.
+Additional allowed values of OPTION are 'Term' for normal terminal output
+(default) and 'TermNoProgressBar' to disable the use of XMLTV::ProgressBar.
+
 B<--output FILE> write to FILE rather than standard output.
 
 B<--days N> grab N days.  The default is 7.
@@ -79,6 +84,8 @@
 use Memoize;
 use XMLTV;
 use XMLTV::Memoize;
+use XMLTV::GUI;
+use XMLTV::ProgressBar;
 use XMLTV::Ask;
 use XMLTV::Config_file;
 use XMLTV::Europe_TZ;
@@ -86,7 +93,7 @@
 # Todo: perhaps we should internationalize messages and docs?
 use XMLTV::Usage <<END
 $0: get Italian television listings in XMLTV format
-To configure: $0 --configure [--config-file FILE]
+To configure: $0 --configure [--config-file FILE] [--gui OPTION]
 To grab listings: $0 [--config-file FILE] [--output FILE] [--days N]
         [--offset N] [--quiet] [--slow]
 END
@@ -115,11 +122,6 @@
 my $base="http://$domain/canali.phtml";
 my $rturl="http://$domain/";
 
-######################################################################
-# Use Term::ProgressBar if installed.
-use constant Have_bar => eval { require Term::ProgressBar; 1 };
-
-######################################################################
 # get options
 # Get options, including undocumented --cache option.
 
@@ -135,6 +137,7 @@
     $opt_slow,
     $opt_configure,
     $opt_config_file,
+    $opt_gui,
     $opt_quiet,
     $opt_share,
    );
@@ -150,6 +153,7 @@
 	   'help'          => \$opt_help,
 	   'configure'     => \$opt_configure,
 	   'config-file=s' => \$opt_config_file,
+           'gui:s'         => \$opt_gui,
 	   'output=s'      => \$opt_output,
 	   'quiet'         => \$opt_quiet,
 	   'slow'	   => \$opt_slow,
@@ -160,6 +164,12 @@
   if (defined $opt_days && $opt_days < 0);
 usage(1) if $opt_help;
 
+# Ask the XMLTV::GUI module for the graphics type we will use  
+my $gui_type = XMLTV::GUI::get_gui_type($opt_gui);
+
+XMLTV::Ask::init($gui_type);
+XMLTV::ProgressBar::init($gui_type);
+
 # share/ directory for storing the list of dud channels.  This next line
 # is altered by processing through tv_grab_it.PL.  But we can use the
 # current directory instead of share/tv_grab_it for development.
@@ -206,11 +216,11 @@
 		}
     open(CONF, ">$config_file") or die "cannot write to $config_file: $!";
     # find list of available channels
-    my $bar = new Term::ProgressBar('getting list of channels', 1)
-      if Have_bar && not $opt_quiet;
+    my $bar = new XMLTV::ProgressBar('getting list of channels', 1)
+      if not $opt_quiet;
 	my %channels=get_channels_list($content);
     die "no channels could be found" if (scalar(keys(%channels))==0);
-    update $bar if Have_bar && not $opt_quiet;
+    update $bar if not $opt_quiet;
 
     # Ask about each channel.
     my @chs = sort keys %channels;
@@ -230,6 +240,7 @@
 	# TODO don't store display-name in config file.
     }
 
+    $bar->finish() if not $opt_quiet;
     close CONF or warn "cannot close $config_file: $!";
     say("Finished configuration.");
 
@@ -297,8 +308,8 @@
 }
 t "will get $days2get days from $opt_offset onwards";
 
-my $bar2 = new Term::ProgressBar('getting icons', scalar @channels)
-  if Have_bar && not $opt_quiet;
+my $bar2 = new XMLTV::ProgressBar('getting icons', scalar @channels)
+  if not $opt_quiet;
 foreach my $ch_id (@channels) {
     my $ch_xid=id_to_xid($ch_id);
 
@@ -316,11 +327,12 @@
 		       'display-name' => [ [ $channels{$ch_id} ] ],
 		       icon => [{src => get_icon($url)}]
 		      });
-    update $bar2 if Have_bar && not $opt_quiet;
+    update $bar2 if not $opt_quiet;
 }
+$bar2->finish() if not $opt_quiet;
 
-my $bar = new Term::ProgressBar('getting listings', scalar @to_get)
-  if Have_bar && not $opt_quiet;
+my $bar = new XMLTV::ProgressBar('getting listings', scalar @to_get)
+  if not $opt_quiet;
 foreach (@to_get) {
     my $canale= $_->[1];
     $url   = $_->[0];
@@ -347,9 +359,10 @@
 	$w->write_programme($_) foreach @dati;
     }
     #			else {warn "skipping $canale \n";}
-    update $bar if Have_bar && not $opt_quiet;
+    update $bar if not $opt_quiet;
 }
 $w->end;
+$bar->finish() if not $opt_quiet;
 
 ######################################################################
 # subroutines
Index: grab/jp/tv_grab_jp
===================================================================
RCS file: /cvsroot/xmltv/xmltv/grab/jp/tv_grab_jp,v
retrieving revision 1.2
diff -u -B -b -w -r1.2 tv_grab_jp
--- grab/jp/tv_grab_jp	26 Feb 2004 04:28:26 -0000	1.2
+++ grab/jp/tv_grab_jp	30 Mar 2004 17:21:45 -0000
@@ -10,7 +10,7 @@
 
 tv_grab_jp --help
 
-tv_grab_jp [--config-file FILE] --configure
+tv_grab_jp [--config-file FILE] --configure [--gui OPTION]
 
 tv_grab_jp [--config-file FILE] [--output FILE] [--days N]
            [--offset N] [--enable-readstr] [--no-category-code]
@@ -37,6 +37,11 @@
 default is B<~/.xmltv/tv_grab_jp.conf>.  This is the file written by
 B<--configure> and read when grabbing.
 
+B<--gui OPTION> Use this option to enable a graphical interface to be used.
+OPTION may be 'Tk' or 'GDialog', or left blank for the best available choice.
+Additional allowed values of OPTION are 'Term' for normal terminal output
+(default) and 'TermNoProgressBar' to disable the use of Term::ProgressBar.
+
 B<--output FILE> write to FILE rather than standard output.
 
 B<--days N> grab N days.  The default and the maximum is 7.
@@ -84,6 +89,8 @@
 use XMLTV;
 use XMLTV::Memoize;
 use XMLTV::Ask;
+use XMLTV::GUI;
+use XMLTV::ProgressBar;
 use XMLTV::Config_file;
 use XMLTV::Mode;
 use XMLTV::Get_nice;
@@ -103,9 +110,6 @@
 END
   ;
 
-# Use Term::ProgressBar if installed.
-use constant Have_bar => eval { require Term::ProgressBar; 1 };
-
 # Attributes of the root element in output.
 my $HEAD = { 'source-info-url'     => 'http://www.ontvjapan.com/',
 	     'source-data-url'     => "http://www.ontvjapan.com/program/",
@@ -152,7 +156,7 @@
 
 XMLTV::Memoize::check_argv('XMLTV::Get_nice::get_nice_aux');
 my ($opt_days, $opt_offset, $opt_help, $opt_output,
-    $opt_configure, $opt_config_file, $opt_quiet,
+    $opt_configure, $opt_config_file, $opt_quiet, $opt_gui,
     $opt_readstr, $opt_no_cat_code, $opt_list_channels);
 $opt_days        = 7; # default
 $opt_offset      = 0; # default
@@ -162,6 +166,7 @@
 GetOptions('help'             => \$opt_help,
 	   'configure'        => \$opt_configure,
 	   'config-file=s'    => \$opt_config_file,
+       'gui:s'            => \$opt_gui,
 	   'output=s'         => \$opt_output,
 	   'quiet'            => \$opt_quiet,
 	   'enable-readstr'   => \$opt_readstr,
@@ -177,6 +182,12 @@
     if (defined $opt_offset && $opt_offset < 0);
 usage(1) if $opt_help;
 
+# Ask the XMLTV::GUI module for the graphics type we will use  
+my $gui_type = XMLTV::GUI::get_gui_type($opt_gui);
+
+XMLTV::Ask::init($gui_type);
+XMLTV::ProgressBar::init($gui_type);
+
 my $mode = XMLTV::Mode::mode('grab', # default
 			     $opt_configure => 'configure',
 			     $opt_list_channels => 'list-channels',
@@ -204,9 +215,10 @@
     # Get region information
     my $barmsg = select_lang('地域情報を取得中',
 			     'getting regions');
-    my $bar = new Term::ProgressBar($barmsg, 1) if Have_bar;
+    my $bar = new XMLTV::ProgressBar($barmsg, 1);
     my %regions = get_regions();
-    update $bar if Have_bar;
+    update $bar;
+    $bar->finish();
 
     # Ask region
     my %ask_regions;
@@ -250,9 +262,10 @@
 
     $barmsg = select_lang('チャンネル情報を取得中',
 			     'getting channels');
-    $bar = new Term::ProgressBar($barmsg, 1) if Have_bar;
+    $bar = new XMLTV::ProgressBar($barmsg, 1);
     my @channels = get_channels($regionid);
-    update $bar if Have_bar;
+    update $bar;
+    $bar->finish();
 
     # Ask about each channel.
     Text::Kakasi::getopt_argv('kakasi', '-ieuc', '-Ja', '-Ha', '-Ka', '-Ea');
@@ -287,7 +300,7 @@
 
     close CONF or warn "cannot close $config_file: $!";
     $msg = select_lang("設定完了.\n", "Finished\n");
-    print $msg;
+    say( $msg );
 
     exit();
 }
@@ -314,16 +327,16 @@
     # Get region information
     my $barmsg = select_lang('地域情報を取得中',
 			     'getting regions');
-    my $bar = new Term::ProgressBar($barmsg, 1) if Have_bar;
+    my $bar = new XMLTV::ProgressBar($barmsg, 1);
     my %regions = get_regions();
-    update $bar if Have_bar;
+    update $bar;
+    $bar->finish();
 
     # get channels for all region
     my @ch_regions = grep { $_ ne 'CATV' } sort keys %regions;
     $barmsg = select_lang('チャンネル情報を取得中',
 			  'getting channels');
-    $bar = new Term::ProgressBar($barmsg, scalar keys %regions)
-	if Have_bar;
+    $bar = new XMLTV::ProgressBar($barmsg, scalar keys %regions);
     foreach (@ch_regions) {
 	my $regionid = $_;
 	my @channels = get_channels($regionid);
@@ -342,8 +355,9 @@
 				'callsign' => $callsign };
 	    }
 	}
-	update $bar if Have_bar;
+	update $bar;
     }
+    $bar->finish();
 
     # Write channels mode.
     foreach (@ch_all) {
@@ -385,9 +399,10 @@
 
 my $barmsg = select_lang('チャンネル情報を取得中',
 			 'getting channels');
-my $bar = new Term::ProgressBar($barmsg, 1) if Have_bar;
+my $bar = new XMLTV::ProgressBar($barmsg, 1);
 my @channels = get_channels($regionid);
-update $bar if Have_bar;
+update $bar;
+$bar->finish();
 
 foreach (@channels) {
     my $id = $_->{id};
@@ -409,13 +424,14 @@
 # This progress bar is for both downloading and parsing.
 $barmsg = select_lang('番組表を取得中',
 		      'getting program info');
-$bar = new Term::ProgressBar($barmsg, scalar @ch_all)
-    if Have_bar && not $opt_quiet;
+$bar = new XMLTV::ProgressBar($barmsg, scalar @ch_all)
+    if not $opt_quiet;
 foreach (@ch_all) {
     process_table($_->{'id'});
-    update $bar if Have_bar && not $opt_quiet;
+    update $bar if not $opt_quiet;
 }
 $writer->end();
+$bar->finish();
 
 ######################################################################
 # subroutine definitions
Index: grab/na/tv_grab_na
===================================================================
RCS file: /cvsroot/xmltv/xmltv/grab/na/tv_grab_na,v
retrieving revision 1.98
diff -u -B -b -w -r1.98 tv_grab_na
--- grab/na/tv_grab_na	2 Mar 2004 21:00:03 -0000	1.98
+++ grab/na/tv_grab_na	30 Mar 2004 17:21:48 -0000
@@ -8,7 +8,7 @@
 
 =head1 SYNOPSIS
 
-tv_grab_na [--help] [--debug] [--config-file FILE] [--output FILE] --configure CONFIG_OPTIONS
+tv_grab_na [--help] [--debug] [--config-file FILE] [--output FILE] [--gui OPTION] --configure CONFIG_OPTIONS
 
 tv_grab_na [--help] [--debug] [--config-file FILE] [--output FILE] GRAB_OPTIONS
 
@@ -45,6 +45,11 @@
 written in configure mode and read in grab mode.  The default location
 is B<~/.xmltv/tv_grab_na.conf>.
 
+B<--gui OPTION> Use this option to enable a graphical interface to be used.
+OPTION may be 'Tk' or 'GDialog', or left blank for the best available choice.
+Additional allowed values of OPTION are 'Term' for normal terminal output
+(default) and 'TermNoProgressBar' to disable the use of Term::ProgressBar.
+
 B<--output FILE> Write output to FILE rather than standard output.
 
 Configuration is normally interactive, but it can be run
@@ -510,6 +515,8 @@
 use Date::Manip;
 
 use XMLTV::ZapListings;
+use XMLTV::GUI;
+use XMLTV::ProgressBar;
 use XMLTV::Ask;
 use XMLTV::Config_file;
 use XMLTV;
@@ -522,6 +529,15 @@
 
 our ($opt_debug, $opt_quiet, $opt_stats);
 
+sub termStatusMessage($)
+{
+    my $msg = shift;
+    die 'expected a message' if not defined $msg;
+    if ( !$opt_quiet ) {
+	print STDERR $msg;
+    }
+}
+
 sub statusMessage($)
 {
     my $msg = shift;
@@ -690,6 +706,9 @@
     print $fp "   --config-file <file>\n";
     print $fp "     use <file> as config file\n";
     print $fp "\n";
+    print $fp "   --gui <option>\n";
+    print $fp "     Use a GUI to ask questions\n";
+    print $fp "\n";
     print $fp "   --output <file>\n";
     print $fp "     redirect output to <file> rather than stdout. Currently only effects\n";
     print $fp "      --list-providers or --list-channels\n";
@@ -853,6 +872,7 @@
 	 $opt_zipcode,
 	 $opt_provider,
 	 $opt_config_file,
+         $opt_gui,
 	 $opt_output,
 	 $opt_retry_limit,
 	 $opt_retry_delay,
@@ -874,6 +894,7 @@
 			 quiet
 			 release-check=s
 			 config-file=s
+                         gui:s
 			 output=s
 			 postalcode=s
 			 zipcode=s
@@ -895,6 +916,12 @@
 
     checkCache(); # make sure $opt_cache is a dir or undef
 
+    # Ask the XMLTV::GUI module for the graphics type we will use  
+    my $gui_type = XMLTV::GUI::get_gui_type($opt_gui);
+
+    XMLTV::Ask::init($gui_type);
+    XMLTV::ProgressBar::init($gui_type);
+    
     $opt_release_check=($opt_release_check=~m;^t;i);
 
     $opt_quiet=(defined($opt_quiet));
@@ -1098,9 +1125,13 @@
 );
 
     if ( $opt_release_check ) {
-	statusMessage("checking XMLTV release information..\n");
+            
+        my $bar = new XMLTV::ProgressBar('checking XMLTV release information..', 1);
 
 	my $retval=XMLTV::ZapListings::getCurrentReleaseInfo($xmltvFreshmeatRecord, $opt_debug);
+        $bar->update();
+        $bar->finish();
+
 	if ( !defined($retval) ) {
 	    statusMessage( <<END
 Warning: failed to get current release information from:
@@ -1121,7 +1152,7 @@
 	}
     }
 
-    statusMessage("\nstarting manual configuration process..\n\n");
+    termStatusMessage("\nstarting manual configuration process..\n\n");
 
     while ( !defined($config->{option_retry_limit}) ) {
 	my $res=ask('how many times do you want to retry on www site failures ? (default=2)');
@@ -1185,7 +1216,7 @@
 	$code=$config->{option_postalcode} if ( defined($config->{option_postalcode}) );
 	$code=$config->{option_zipcode} if ( defined($config->{option_zipcode}) );
 
-	statusMessage("\ngetting list of providers for postal/zip code $code, be patient..\n");
+        my $bar = new XMLTV::ProgressBar("getting list of providers for postal/zip code $code, be patient..", 1);
 
 	my @providers;
 	my $failed=0;
@@ -1224,11 +1255,11 @@
 		if ( $p->{id} eq $config->{option_provider} ) {
 		    if ( $config->{option_provider_desc} ne $p->{description} ) {
 			if ( $opt_auto_fail_on_provider_changes ) {
-			    statusMessage("provider:\n\t".$config->{option_provider}." \# ".$config->{option_provider_desc}.
+			    termStatusMessage("provider:\n\t".$config->{option_provider}." \# ".$config->{option_provider_desc}.
 					  "\nhas new description (".$p->{description}."), exiting\n");
 			    exit(1);
 			}
-			statusMessage("updating provider description to: $p->{description}\n");
+			termStatusMessage("updating provider description to: $p->{description}\n");
 			$config->{option_provider_desc}=$p->{description};
 		    }
 		    $still_valid=1;
@@ -1248,7 +1279,7 @@
 				exit(1);
 			    }
 			}
-			statusMessage("updating provider id to: $p->{id}\n");
+			termStatusMessage("updating provider id to: $p->{id}\n");
 			$config->{option_provider}=$p->{id};
 			$still_valid=1;
 		    }
@@ -1275,6 +1306,9 @@
 	    }
 	}
 
+        $bar->update();
+        $bar->finish();
+        
 	while ( !$config->{option_provider} ) {
 	    my @providersWithIds;
 
@@ -1316,18 +1350,19 @@
 	    my $id = $1; die if not defined $id;
 	    $config->{option_provider}=$id;
 	    $config->{option_provider_desc}=$res;
-	    statusMessage("\nyou chose $id \# $res\n");
+	    termStatusMessage("\nyou chose $id \# $res\n");
 	}
     }
 
     # if we're in the configure step, lets refresh the list of channels
     # being careful to warn about additions and deletions
     
+    my $bar;
     if ( $config->haveAnyChannels() ) {
-	statusMessage("\nchecking for changes to channel list, be patient..\n");
+        $bar = new XMLTV::ProgressBar('checking for changes to channel list, be patient..', 1);
     }
     else {
-	statusMessage("\ngetting channel list, be patient..\n");
+        $bar = new XMLTV::ProgressBar('getting channel list, be patient..', 1);
     }
 
     my @channels;
@@ -1339,7 +1374,7 @@
 	}
 
 	@channels=$zl->getChannelList($config->{option_provider});
-	statusMessage("got channel list\n");
+	termStatusMessage("got channel list\n");
 	if ( ! @channels || !defined($channels[0]) ) {
 	    $failed=1;
 	}
@@ -1348,6 +1383,8 @@
 	    last;
 	}
     }
+    $bar->update();
+    $bar->finish();
 
     if ( $failed ) {
 	#errorMessage("$0: failed to get list of channels for postal/zip code $config->{option_postalzipcode}\n");
@@ -1375,11 +1412,11 @@
     if ( defined($opt_auto_new_channels) ) {
 	if ( $opt_auto_new_channels eq "ignore" ) {
 	    @r = map { 0 } @toAsk;
-	    statusMessage("ignoring new channel $_\n") foreach @toAsk;
+	    termStatusMessage("ignoring new channel $_\n") foreach @toAsk;
 	}
 	elsif ( $opt_auto_new_channels eq "add" ) {
 	    @r = map { 1 } @toAsk;
-	    statusMessage("adding new channel $_\n") foreach @toAsk;
+	    termStatusMessage("adding new channel $_\n") foreach @toAsk;
 	}
 	else {
 	    die "invalid auto_new_channels $opt_auto_new_channels\n";
@@ -1415,11 +1452,11 @@
     if ( defined($opt_auto_missing_channels) ) {
 	if ( $opt_auto_missing_channels eq "ignore" ) {
 	    @r = map { 0 } @toAsk;
-	    statusMessage("ignoring missing channel $_...\n") foreach @toAsk;
+	    termStatusMessage("ignoring missing channel $_...\n") foreach @toAsk;
 	}
 	elsif ( $opt_auto_missing_channels eq "remove" ) {
 	    @r = map { 1 } @toAsk;
-	    statusMessage("removing channel $_..\n") foreach @toAsk;
+	    termStatusMessage("removing channel $_..\n") foreach @toAsk;
 	}
 	else {
 	    die "invalid auto_missing_channels $opt_auto_missing_channels\n";
@@ -1442,15 +1479,15 @@
 
     if ( $channelsUpdated == 0 ) {
 	if ( $config->haveAnyChannels() ) {
-	    statusMessage("\nchannel line-up hasn't changed\n");
+	    termStatusMessage("\nchannel line-up hasn't changed\n");
 	}
 	else {
-	    statusMessage("\nno channels added\n");
+	    termStatusMessage("\nno channels added\n");
 	}
     } 
 
     # write out config file
-    statusMessage("\nupdating $configfile..\n");
+    termStatusMessage("\nupdating $configfile..\n");
     if ( $config->save($configfile) != 0 ) {
 	errorMessage("$0: $configfile save failed\n");
 	exit(1);
Index: grab/na_dd/tv_grab_na_dd
===================================================================
RCS file: /cvsroot/xmltv/xmltv/grab/na_dd/tv_grab_na_dd,v
retrieving revision 1.13
diff -u -B -b -w -r1.13 tv_grab_na_dd
--- grab/na_dd/tv_grab_na_dd	19 Mar 2004 12:39:15 -0000	1.13
+++ grab/na_dd/tv_grab_na_dd	30 Mar 2004 17:21:50 -0000
@@ -10,7 +10,7 @@
 
 tv_grab_na_dd --help
 
-tv_grab_na_dd [--config-file FILE] --configure
+tv_grab_na_dd [--config-file FILE] --configure [--gui OPTION]
 
 tv_grab_na_dd [--config-file FILE] [--password-file FILE]
               [--output FILE] [--days N]
@@ -43,6 +43,11 @@
 default is B<~/.xmltv/tv_grab_na_dd.conf>.  This is the file written by
 B<--configure> and read when grabbing.
 
+B<--gui OPTION> Use this option to enable a graphical interface to be used.
+OPTION may be 'Tk' or 'GDialog', or left blank for the best available choice.
+Additional allowed values of OPTION are 'Term' for normal terminal output
+(default) and 'TermNoProgressBar' to disable the use of Term::ProgressBar.
+
 B<--password-file FILE> Read the DataDirect password from FILE.  If
 this option is not given, the grabber prompts for a password.
 
@@ -91,15 +96,16 @@
 use Getopt::Long;
 use XML::Twig 3.10;
 # use Data::Dumper;
-use constant Have_bar => eval { require Term::ProgressBar; 1 };
 
 use XMLTV;
 use XMLTV::Ask;
+use XMLTV::GUI;
+use XMLTV::ProgressBar;
 use XMLTV::Config_file;
 use XMLTV::Version '$Id$';
 use XMLTV::Usage <<END
 $0: get lists from Zap2IT's DataDirect service in XMLTV format
-To configure: $0 --configure [--config-file FILE]
+To configure: $0 --configure [--config-file FILE] [--gui OPTION]
 To grab listings: $0 [--config-file FILE] [--output FILE] [--days N]
         [--offset N] [--quiet] [--lineup "LINEUP"] [--old-chan-id] [--low-mem]
         [--dd-data FILE] [--reprocess]
@@ -139,6 +145,7 @@
 my $opt_help;           # ask for help
 my $opt_configure;      # configure mode
 my $opt_config_file;    # config_file_name
+my $opt_gui;            # what type of gui to use
 my $opt_password_file;  # file containing password
 my $opt_output;         # output name
 my $opt_days       =10; # days to fetch
@@ -160,6 +167,7 @@
 	       'help'          => \$opt_help,
     	   'configure'       => \$opt_configure,
     	   'config-file=s'   => \$opt_config_file,
+           'gui:s'           => \$opt_gui,
            'password-file=s' => \$opt_password_file,
     	   'output=s'        => \$opt_output,
            'days=i'          => \$opt_days,
@@ -176,6 +184,12 @@
 
 usage(1) if $opt_help;
 
+# Ask the XMLTV::GUI module for the graphics type we will use  
+my $gui_type = XMLTV::GUI::get_gui_type($opt_gui);
+
+XMLTV::Ask::init($gui_type);
+XMLTV::ProgressBar::init($gui_type);
+
 die 'number of days must not be negative' if ($opt_days < 0);
 
 die "must specify --dd-data during reprocess\n"
@@ -253,11 +267,8 @@
     #
     require Term::ReadKey;
     $| = 1;
-    print STDERR "Password for $dd_user: ";
-    Term::ReadKey::ReadMode('noecho');
-    chomp($dd_pass = <STDIN>);
-    Term::ReadKey::ReadMode('restore');
-    print STDERR "\n";
+    chomp( $dd_pass = ask( "Password for $dd_user: " ) );
+
 }
 
 
@@ -326,7 +337,7 @@
                                              SUFFIX => '.tmp',
                                              UNLINK=>($DEBUG ? 0 : 1));
     }
-    print STDERR "Fetching from DataDirect\n";
+    say( "Fetching from DataDirect" );
     print STDERR "dd data is in $dd_data_name\n"
       if $DEBUG || defined $opt_dd_data;
 
@@ -349,7 +360,7 @@
                              })
             -> proxy("http://localhost/", options => {compress_threshold => 10000},
                                                       timeout            => 420);
-    print STDERR "about to download\n";
+    say( "about to download" );
     my $got = $soap->download("<startTime>$dd_start</startTime><endTime>$dd_stop</endTime>");
     $dd_data->print($got);
     $dd_data->flush;
@@ -358,7 +369,7 @@
     }
     $dd_data_size= -s $dd_data;
     $time = int(time() - $time);
-    printf STDERR "Fetched %d k/bytes in %d seconds\n",$dd_data_size/1024,$time;
+    say( "Fetched %d k/bytes in %d seconds\n",$dd_data_size/1024,$time );
 } # get data
 
 #
@@ -440,13 +451,12 @@
 
 unless ($opt_quiet)
 {
-    if (Have_bar) { $bar = new Term::ProgressBar('loading data',$dd_data_size+1) }
-    else          { print STDERR "loading data "; }
+    $bar = new XMLTV::ProgressBar('loading data',$dd_data_size+1);
 }
 seek($dd_data,0,0);  #rewind
 $twig->parse( $dd_data );
 $bar->update($dd_data_size+1) if $bar;
-print STDERR "\n" unless ($opt_quiet || $bar );
+$bar->finish();
 
 $twig=undef;  # destroy twig (just in case)
 
@@ -464,7 +474,7 @@
 #
 # store lineup information with stations
 #
-print STDERR "Processing only lineup '$opt_lineup'\n" if defined $opt_lineup;
+say( "Processing only lineup '$opt_lineup'" ) if defined $opt_lineup;
 {
     my $id;
     my $count=0;
@@ -754,13 +764,12 @@
 unless ($opt_quiet)
 {
     my $msg = 'writing schedule';
-    if (Have_bar) { $bar = new Term::ProgressBar($msg,$dd_data_size+1) }
-    else          { print STDERR $msg; }
+    $bar = new XMLTV::ProgressBar($msg,$dd_data_size+1);
 }
 my $time=time();
 $twig->parse( $dd_data );
-print STDERR "\n" unless ($opt_quiet || $bar );
 $bar->update($dd_data_size+1) if $bar;
+$bar->finish();
 
 $writer->end();
 if (@messages)
@@ -768,7 +777,7 @@
     warn join( "\n", @messages ), "\n";
 }
 
-printf STDERR "Downloaded %d programs in %d seconds\n",$prog_count,time()-$start_time
+say( "Downloaded %d programs in %d seconds\n",$prog_count,time()-$start_time )
     unless $opt_quiet;
 
 exit();
Index: grab/nl/tv_grab_nl.in
===================================================================
RCS file: /cvsroot/xmltv/xmltv/grab/nl/tv_grab_nl.in,v
retrieving revision 1.3
diff -u -B -b -w -r1.3 tv_grab_nl.in
--- grab/nl/tv_grab_nl.in	7 Jan 2004 19:55:26 -0000	1.3
+++ grab/nl/tv_grab_nl.in	30 Mar 2004 17:21:50 -0000
@@ -82,6 +82,8 @@
 use Date::Manip;
 use XMLTV;
 use XMLTV::Memoize;
+use XMLTV::GUI;
+use XMLTV::ProgressBar;
 use XMLTV::Ask;
 use XMLTV::Config_file;
 use XMLTV::Europe_TZ;
@@ -111,9 +113,6 @@
     }
 }
 
-# Use Term::ProgressBar if installed.
-use constant Have_bar => eval { require Term::ProgressBar; 1 };
-
 # Function prototypes.
 sub time_tot_str( $ );
 sub time_van_str( $ );
@@ -139,7 +138,7 @@
 # Get options, including undocumented --cache option.
 XMLTV::Memoize::check_argv('XMLTV::Get_nice::get_nice_aux');
 my ($opt_days, $opt_offset, $opt_help, $opt_output,
-    $opt_configure, $opt_config_file, $opt_quiet,
+    $opt_configure, $opt_config_file, $opt_gui, $opt_quiet,
     $opt_list_channels, $opt_slow, $opt_share);
 $opt_days   = 7; # default
 $opt_offset = 0; # default
@@ -148,6 +147,7 @@
 	   'help'          => \$opt_help,
 	   'configure'     => \$opt_configure,
 	   'config-file=s' => \$opt_config_file,
+           'gui:s'         => \$opt_gui,
 	   'output=s'      => \$opt_output,
 	   'quiet'         => \$opt_quiet,
 	   'slow'	   => \$opt_slow,
@@ -172,6 +172,12 @@
 my $OUR_SHARE_DIR = (defined $SHARE_DIR) ? "$SHARE_DIR/tv_grab_nl" : '.';
 (my $CHANNELS_FILE = "$OUR_SHARE_DIR/channels") =~ tr!/!/!s;
 
+# Ask the XMLTV::GUI module for the graphics type we will use  
+my $gui_type = XMLTV::GUI::get_gui_type($opt_gui);
+
+XMLTV::Ask::init($gui_type);
+XMLTV::ProgressBar::init($gui_type);
+
 my $mode = XMLTV::Mode::mode('grab', # default
 			     $opt_configure => 'configure',
 			     $opt_list_channels => 'list-channels',
@@ -189,11 +195,12 @@
     # configuring it's useful to download at least something as a
     # check that the site is reachable etc.
     #
-    my $bar = new Term::ProgressBar('getting list of channels', 1)
-      if Have_bar && not $opt_quiet;
+    my $bar = new XMLTV::ProgressBar('getting list of channels', 1)
+      if not $opt_quiet;
     my %channels = get_channels(1);
     die 'no channels could be found' if (scalar(keys(%channels)) == 0);
-    update $bar if Have_bar && not $opt_quiet;
+    update $bar if not $opt_quiet;
+    $bar->finish();
 
     # Ask about each channel.
     my @chs = sort { $a <=> $b } keys %channels;
@@ -246,15 +253,15 @@
    });
 
 if ($mode eq 'list-channels') {
-    my $bar = new Term::ProgressBar('getting list of channels', 1)
-      if Have_bar && not $opt_quiet;
+    my $bar = new XMLTV::ProgressBar('getting list of channels', 1)
+      if not $opt_quiet;
 
     # Don't do any page fetches - assume the contents of the channels
     # file has already been checked enough.
     #
     my %channels = get_channels(0);
     die 'no channels could be found' if (scalar(keys(%channels)) == 0);
-    update $bar if Have_bar && not $opt_quiet;
+    update $bar if not $opt_quiet;
 
     foreach my $ch_did (sort(keys %channels)) {
 	my $ch_name = $channels{$ch_did};
@@ -263,6 +270,7 @@
 				 'display-name' => [ [ $ch_name ] ] });
     }
     $writer->end();
+    $bar->finish();
     exit();
 }
 
@@ -335,16 +343,17 @@
 my %warned_ch_name; # suppress duplicate warnings
 
 my @summary_page_data;
-my $bar = new Term::ProgressBar('downloading summary', scalar @to_get)
-  if Have_bar && not $opt_quiet;
+my $bar = new XMLTV::ProgressBar('downloading summary', scalar @to_get)
+  if not $opt_quiet;
 foreach (@to_get) {
     my ($url, $day, $ch_xmltv_id, $ch_tvgids_id) = @$_;
     die if ref $url;
     push @summary_page_data,
       [ $url, $ch_xmltv_id,
 	[ process_summary_page($url, $day, $day, \$channels{$ch_tvgids_id}) ] ];
-    update $bar if Have_bar and not $opt_quiet;
+    update $bar if not $opt_quiet;
 }
+$bar->finish();
 
 # Now we've fetched the descriptions, we know the channel names.
 foreach $ch_did (@channels) {
@@ -467,8 +476,8 @@
 
 my @to_write;
 if ($opt_slow) {
-    $bar = new Term::ProgressBar('getting details', scalar @summary_programmes)
-      if Have_bar && not $opt_quiet;
+    $bar = new XMLTV::ProgressBar('getting details', scalar @summary_programmes)
+      if not $opt_quiet;
     foreach my $s (@summary_programmes) {
 	my $urls = delete $s->{url};
 	if (not defined $urls) {
@@ -563,6 +572,7 @@
 	push @to_write, $detailed;
 	update $bar if $bar;
     }
+    $bar->finish();
 }
 else {
     @to_write = @summary_programmes;
Index: grab/se/tv_grab_se
===================================================================
RCS file: /cvsroot/xmltv/xmltv/grab/se/tv_grab_se,v
retrieving revision 1.5
diff -u -B -b -w -r1.5 tv_grab_se
--- grab/se/tv_grab_se	17 Mar 2004 17:43:49 -0000	1.5
+++ grab/se/tv_grab_se	30 Mar 2004 17:21:52 -0000
@@ -10,7 +10,7 @@
 
 tv_grab_se --help
 
-tv_grab_se [--config-file FILE] --configure
+tv_grab_se [--config-file FILE] --configure [--gui OPTION]
 
 tv_grab_se [--config-file FILE] [--output FILE] [--days N]
            [--offset N] [--quiet] [--debug]
@@ -34,6 +34,11 @@
 default is B<~/.xmltv/tv_grab_se.conf>.  This is the file written by
 B<--configure> and read when grabbing.
 
+B<--gui OPTION> Use this option to enable a graphical interface to be used.
+OPTION may be 'Tk' or 'GDialog', or left blank for the best available choice.
+Additional allowed values of OPTION are 'Term' for normal terminal output
+(default) and 'TermNoProgressBar' to disable the use of Term::ProgressBar.
+
 B<--output FILE> When grabbing, write output to FILE rather than
 standard output.
 
@@ -84,6 +89,8 @@
 
 use XMLTV;
 use XMLTV::Ask;
+use XMLTV::GUI;
+use XMLTV::ProgressBar;
 use XMLTV::Config_file;
 use XMLTV::Get_nice qw(get_nice);
 use XMLTV::Memoize;
@@ -96,9 +103,6 @@
 use Getopt::Long;
 use URI;
 
-# Use Term::ProgressBar if installed.
-use constant Have_bar => eval { require Term::ProgressBar; 1 };
-
 # The timezone for Sweden during wintertime.
 use constant LOCAL_TZ => "+0100";
 
@@ -147,6 +151,7 @@
 my $opt = { days => 5,
             offset => 0,
             "config-file" => undef,
+            gui => undef,
             configure => 0,
             help => 0,
             quiet => 0,
@@ -158,6 +163,7 @@
                       days=i
                       offset=i
                       config-file=s
+                      gui
                       configure
                       help|h
                       quiet
@@ -172,7 +178,7 @@
   print << 'EOH';
 tv_grab_se --help
 
-tv_grab_se [--config-file FILE] --configure
+tv_grab_se [--config-file FILE] --configure [--gui OPTION]
 
 tv_grab_se [--config-file FILE] [--output FILE] [--days N]
            [--offset N] [--quiet] [--debug]
@@ -182,6 +188,12 @@
   exit(1);
 }
 
+# Ask the XMLTV::GUI module for the graphics type we will use  
+my $gui_type = XMLTV::GUI::get_gui_type($opt->{'gui'});
+
+XMLTV::Ask::init($gui_type);
+XMLTV::ProgressBar::init($gui_type);
+
 # XMLTV::Europe_TZ says that we should do this...
 Date_Init('TZ=UTC');
 
@@ -237,10 +249,10 @@
 my $date = increase_date( $today, $opt->{offset} );
 
 my $bar = undef;
-$bar = new Term::ProgressBar( {
+$bar = new XMLTV::ProgressBar( {
     name => 'downloading listings',
     count => $opt->{days} * @channel_list
-    }) if Have_bar && (not $opt->{quiet}) && (not $opt->{debug});
+    }) if (not $opt->{quiet}) && (not $opt->{debug});
 
 for( my $i=0; $i < $opt->{days}; $i++ )
 {
@@ -255,7 +267,7 @@
 
     $date = increase_date( $date, 1 );
 }
-
+$bar->finish();
 $w->end();
 
 # Signal that something went wrong if there were warnings.
Index: grab/uk_rt/tv_grab_uk_rt.in
===================================================================
RCS file: /cvsroot/xmltv/xmltv/grab/uk_rt/tv_grab_uk_rt.in,v
retrieving revision 1.55
diff -u -B -b -w -r1.55 tv_grab_uk_rt.in
--- grab/uk_rt/tv_grab_uk_rt.in	19 Mar 2004 21:12:39 -0000	1.55
+++ grab/uk_rt/tv_grab_uk_rt.in	30 Mar 2004 17:21:53 -0000
@@ -11,7 +11,7 @@
 
 tv_grab_uk_rt --help
 
-tv_grab_uk_rt [--config-file FILE] --configure
+tv_grab_uk_rt [--config-file FILE] --configure [--gui OPTION]
 
 tv_grab_uk_rt [--config-file FILE] [--output FILE] [--quiet]
               [--days N] [--offset N]
@@ -36,6 +36,11 @@
 default is B<~/.xmltv/tv_grab_uk_rt.conf>.  This is the file written by
 B<--configure> and read when grabbing.
 
+B<--gui OPTION> Use this option to enable a graphical interface to be used.
+OPTION may be 'Tk' or 'GDialog', or left blank for the best available choice.
+Additional allowed values of OPTION are 'Term' for normal terminal output
+(default) and 'TermNoProgressBar' to disable the use of Term::ProgressBar.
+
 B<--output FILE> When grabbing, write output to FILE rather than
 standard output.
 
@@ -89,6 +94,8 @@
 require HTML::Entities;
 use XMLTV;
 use XMLTV::Memoize;
+use XMLTV::GUI;
+use XMLTV::ProgressBar;
 use XMLTV::Ask;
 use XMLTV::Europe_TZ;
 use XMLTV::Config_file;
@@ -177,6 +184,7 @@
     $opt_share,
     $opt_configure,
     $opt_config_file,
+    $opt_gui,
     $opt_offset,
     $opt_quiet,
     $opt_slow,
@@ -192,6 +200,7 @@
 	   'help'          => \$opt_help,
 	   'configure'     => \$opt_configure,
 	   'config-file=s' => \$opt_config_file,
+           'gui:s'         => \$opt_gui,
 	   'output=s'      => \$opt_output,
 	   'share=s'       => \$opt_share, # also undocumented
            'offset=i'      => \$opt_offset,
@@ -208,6 +217,12 @@
 die "--limit-details makes no sense without --slow\n"
   if defined $opt_detailtimerange and not $opt_slow;
 
+# Ask the XMLTV::GUI module for the graphics type we will use  
+my $gui_type = XMLTV::GUI::get_gui_type($opt_gui);
+
+XMLTV::Ask::init($gui_type);
+XMLTV::ProgressBar::init($gui_type);
+
 # Date::Manip has a bug where 'now' will be wrong if you change the
 # timezone.  It won't be correctly converted from the system timezone
 # to the new one.  So we call parse_date('today midnight') _before_
@@ -387,8 +402,11 @@
 	}
     }
     my $days = @dates_to_get > 1 ? 'days' : 'day';
-    say('getting ' . (scalar @dates_to_get) . " $days of listings\n")
+    
+    my $bar = new XMLTV::ProgressBar('getting ' . (scalar @dates_to_get)
+        . " $days of listings", 1)
       unless $opt_quiet;
+    
     t 'getting dates:' . d \@dates_to_get;
 
     $writer->start({ 'source-info-url'     => "$BASE_URL/",
@@ -641,6 +659,9 @@
 	$writer->write_programme($_);
     }
     $writer->end();
+    
+    $bar->finish();
+    
 }
 
 
@@ -1758,10 +1779,12 @@
 
     XMLTV::Config_file::check_no_overwrite($config_file);
 
-    # FIXME turn into progress bar.
-    print STDERR "finding channels:\t";
+    my $bar = new XMLTV::ProgressBar('finding channels', 1)
+	  if not $opt_quiet;
+    
     my %channels = get_channels();
-    print STDERR "got ". (scalar keys %channels) . " done.\n" unless $opt_quiet;
+    
+    $bar->finish();
 
     # FIXME need to make directory
     open(CONF, ">$config_file") or die "cannot write to $config_file: $!";
Index: lib/Ask.pm
===================================================================
RCS file: /cvsroot/xmltv/xmltv/lib/Ask.pm,v
retrieving revision 1.12
diff -u -B -b -w -r1.12 Ask.pm
--- lib/Ask.pm	3 Jan 2004 14:52:53 -0000	1.12
+++ lib/Ask.pm	30 Mar 2004 17:21:53 -0000
@@ -7,9 +7,18 @@
 
 package XMLTV::Ask;
 use strict;
-use Carp qw(croak carp);
 
-# Use Log::TraceMessages if installed, and choose graphical or not.
+use vars qw(@ISA @EXPORT);
+use Exporter;
+@ISA = qw(Exporter);
+@EXPORT = qw(ask
+             askQuestion
+             askBooleanQuestion
+             askManyBooleanQuestions
+             say
+             );
+
+# Use Log::TraceMessages if installed.
 BEGIN {
     eval { require Log::TraceMessages };
     if ($@) {
@@ -21,21 +30,51 @@
 	*d = \&Log::TraceMessages::d;
     }
 
-    # For now we do graphical configuration only if the undocumented
-    # XMLTV_TK environment variable is set to a true value.
-    #
-    if ($ENV{XMLTV_TK}
-	and (defined($ENV{DISPLAY}) || $^O eq 'MSWin32')
-	and eval { require Tk }) {
-	require XMLTV::AskTk; XMLTV::AskTk->import;
-	*XMLTV::Ask:: = *XMLTV::AskTk::;
-    }
-    else {
-	require XMLTV::AskTerm; XMLTV::AskTerm->import;
-	*XMLTV::Ask:: = *XMLTV::AskTerm::;
+
+    
     }
+
+my $real_class = 'XMLTV::Ask::Term';
+
+sub AUTOLOAD {
+        
+        use vars qw($AUTOLOAD);
+        
+        (my $method_name = $AUTOLOAD) =~ s/.*::(.*?)/$1/;
+        
+        (my $real_class_path = $real_class.".pm") =~ s/::/\//g;
+        
+        require $real_class_path;
+        import $real_class_path;
+        
+        $real_class->$method_name(@_);
+        
 }
 
 
+# Must be called before we use this module if we want to use a gui.
+sub init( $ ) {
+        
+        my $gui_type = shift;
+        
+        if($gui_type =~ /^term/) {
+                
+                $real_class = 'XMLTV::Ask::Term';
+                
+        } elsif($gui_type eq 'tk') {
+                
+                $real_class = 'XMLTV::Ask::Tk';
+                
+        } elsif($gui_type eq 'gdialog') {
+                
+                $real_class = 'XMLTV::Ask::GDialog';
+                
+        } else {
+                
+                die "Unknown gui type: '$gui_type'.";
+                
+        }
+        
+}
 
 1;
