Index: MANIFEST
===================================================================
RCS file: /cvsroot/xmltv/xmltv/MANIFEST,v
retrieving revision 1.108
diff -u -B -b -w -r1.108 MANIFEST
--- MANIFEST	23 May 2004 18:15:53 -0000	1.108
+++ MANIFEST	5 Jun 2004 14:51:54 -0000
@@ -5,8 +5,14 @@
 README
 README.cygwin
 lib/Ask.pm
-lib/AskTerm.pm
-lib/AskTk.pm
+lib/Ask/GDialog.pm
+lib/Ask/Term.pm
+lib/Ask/Tk.pm
+lib/ProgressBar.pm
+lib/ProgressBar/GDialog.pm
+lib/ProgressBar/None.pm
+lib/ProgressBar/Term.pm
+lib/ProgressBar/Tk.pm
 lib/XMLTV.pm.PL
 lib/XMLTV.pm.in
 lib/TZ.pm
Index: Makefile.PL
===================================================================
RCS file: /cvsroot/xmltv/xmltv/Makefile.PL,v
retrieving revision 1.214
diff -u -B -b -w -r1.214 Makefile.PL
--- Makefile.PL	23 May 2004 19:18:13 -0000	1.214
+++ Makefile.PL	5 Jun 2004 14:51:55 -0000
@@ -129,9 +129,16 @@
      'lib/Usage.pm'               => '$(INST_LIBDIR)/XMLTV/Usage.pm',
      'lib/Date.pm'                => '$(INST_LIBDIR)/XMLTV/Date.pm',
      'lib/Version.pm'             => '$(INST_LIBDIR)/XMLTV/Version.pm',
+     'lib/GUI.pm'                 => '$(INST_LIBDIR)/XMLTV/GUI.pm',
      'lib/Ask.pm'                 => '$(INST_LIBDIR)/XMLTV/Ask.pm',
-     'lib/AskTk.pm'               => '$(INST_LIBDIR)/XMLTV/AskTk.pm',
-     'lib/AskTerm.pm'             => '$(INST_LIBDIR)/XMLTV/AskTerm.pm',
+     'lib/Ask/GDialog.pm'         => '$(INST_LIBDIR)/XMLTV/Ask/GDialog.pm',
+     'lib/Ask/Term.pm'            => '$(INST_LIBDIR)/XMLTV/Ask/Term.pm',
+     'lib/Ask/Tk.pm'              => '$(INST_LIBDIR)/XMLTV/Ask/Tk.pm',
+     'lib/ProgressBar.pm'         => '$(INST_LIBDIR)/XMLTV/ProgressBar.pm',
+     'lib/ProgressBar/GDialog.pm' => '$(INST_LIBDIR)/XMLTV/ProgressBar/GDialog.pm',
+     'lib/ProgressBar/None.pm'    => '$(INST_LIBDIR)/XMLTV/ProgressBar/None.pm',
+     'lib/ProgressBar/Term.pm'    => '$(INST_LIBDIR)/XMLTV/ProgressBar/Term.pm',
+     'lib/ProgressBar/Tk.pm'      => '$(INST_LIBDIR)/XMLTV/ProgressBar/Tk.pm',
      'lib/Summarize.pm'           => '$(INST_LIBDIR)/XMLTV/Summarize.pm',
      'lib/IMDB.pm'                => '$(INST_LIBDIR)/XMLTV/IMDB.pm',
      'lib/Gunzip.pm'              => '$(INST_LIBDIR)/XMLTV/Gunzip.pm',
@@ -198,11 +205,11 @@
 # And Log::TraceMessages is 'suggested' but we don't warn about that.
 
 if ($opt_yes) {
-    *ask = sub { print "$_[0] yes\n"; 1 };
+    *ask = sub { print "$_[1] yes\n"; 1 };
 }
 else {
-    do 'lib/AskTerm.pm' or die 'could not load lib/AskTerm.pm, aborting';
-    *ask = \&XMLTV::AskTerm::askBooleanQuestion;
+    do 'lib/Ask/Term.pm' or die 'could not load lib/Ask/Term.pm, aborting';
+    *ask = \&XMLTV::Ask::Term::ask_boolean;
 }
 
 # Weird shit happens when you change things like PREFIX without
@@ -216,7 +223,7 @@
 
 END
   ;
-    if (! ask("Do you wish to continue anyway?", 0)) {
+    if (! ask(0, "Do you wish to continue anyway?", 0)) {
 	exit(1);
     }
 }
@@ -245,6 +252,11 @@
        prereqs => { 'HTML::Entities' => 1.27 } ,
      },
 
+     { name => 'tv_grab_uk_bleb',
+       blurb => 'Fast alternative grabber for the UK',
+       exes => [ 'grab/uk_bleb/tv_grab_uk_bleb' ],
+       prereqs => { 'IO::Scalar' => 0, 'Archive::Zip' => 0 } },
+     
      { name => 'tv_grab_it',
        blurb => 'Grabber for Italy',
        exes => [ 'grab/it/tv_grab_it' ],
@@ -455,7 +467,7 @@
 		      "\n");
     }
     print STDERR "\n";
-    if (not ask('Do you want to proceed with this configuration?', 1)) {
+    if (not ask(0, 'Do you want to proceed with this configuration?', 1)) {
 	# Need to set {install} for each component by prompting.
 	foreach my $info (@opt_components) {
 	    my $missing = $info->{missing};
@@ -485,8 +497,9 @@
 	    }
 	
 	    $info->{install} =
-	      ask($msg, not $missing);
+	      ask(0, $msg, not $missing);
 	}
+
     }
 }
 else {
Index: grab/Config_file.pm
===================================================================
RCS file: /cvsroot/xmltv/xmltv/grab/Config_file.pm,v
retrieving revision 1.8
diff -u -B -b -w -r1.8 Config_file.pm
--- grab/Config_file.pm	13 May 2004 19:32:51 -0000	1.8
+++ grab/Config_file.pm	5 Jun 2004 14:51:55 -0000
@@ -46,7 +46,7 @@
 sub check_no_overwrite( $ ) {
     my $f = shift;
     if (-e $f) {
-	if (not askBooleanQuestion <<END
+	if (not ask_boolean <<END
 The configuration file $f
 already exists.  There is currently no support for altering an
 existing configuration: you have to reconfigure from scratch.
Index: grab/Grab_XML.pm
===================================================================
RCS file: /cvsroot/xmltv/xmltv/grab/Grab_XML.pm,v
retrieving revision 1.11
diff -u -B -b -w -r1.11 Grab_XML.pm
--- grab/Grab_XML.pm	3 Jan 2004 14:52:53 -0000	1.11
+++ grab/Grab_XML.pm	5 Jun 2004 14:51:55 -0000
@@ -6,6 +6,8 @@
 use XMLTV;
 use XMLTV::Usage;
 use XMLTV::Memoize;
+use XMLTV::ProgressBar;
+use XMLTV::Ask;
 use XMLTV::TZ qw(parse_local_date);
 use XMLTV::Get_nice qw();
 use XMLTV::Date;
@@ -24,9 +26,6 @@
     }
 }
 
-# Use Term::ProgressBar if installed.
-use constant Have_bar => eval { require Term::ProgressBar; 1 };
-
 =pod
 
 =head1 NAME
@@ -74,8 +73,10 @@
 date can be downloaded.  This method is abstract, you must override
 it.
 
+Arguments: the command line options for --config-file and --quiet.
+
 =cut
-sub urls_by_date( $ ) {
+sub urls_by_date( $$$ ) {
     my $pkg = shift;
     die 'abstract class method: override in subclass';
 }
@@ -95,6 +96,21 @@
     return shift; # leave unchanged
 }
 
+=pod
+
+=item XMLTV::Grab_XML->configure()
+
+Configure the grabber if needed.
+
+=cut
+sub configure( $ ) {
+    # Default is no config file
+    print STDERR "no configuration necessary\n";
+    exit();
+}
+
+=pod
+
 =item XMLTV::Grab_XML->nextday(day)
 
 Bump a YYYYMMDD date by one.  You probably shouldnE<39>t override this.
@@ -167,6 +183,7 @@
     my ($opt_days,
 	$opt_help,
 	$opt_output,
+    $opt_gui,
 	$opt_offset,
 	$opt_quiet,
 	$opt_configure,
@@ -177,6 +194,7 @@
     GetOptions('days=i'        => \$opt_days,
 	       'help'          => \$opt_help,
 	       'output=s'      => \$opt_output,
+           'gui:s'         => \$opt_gui,
 	       'offset=i'      => \$opt_offset,
 	       'quiet'         => \$opt_quiet,
 	       'configure'     => \$opt_configure,
@@ -187,13 +205,13 @@
       if (defined $opt_days && $opt_days < 0);
     usage(1, $pkg->usage_msg()) if $opt_help;
     usage(0, $pkg->usage_msg()) if @ARGV;
+    
+    XMLTV::Ask::init($opt_gui);
+    XMLTV::ProgressBar::init($opt_gui);
+    
     if ($opt_configure) {
-	print STDERR "no configuration necessary\n";
-	exit();
-    }
-    for ($opt_config_file) {
-	warn "this grabber has no configuration, so ignoring --config-file $_\n"
-	  if defined;
+        $pkg->configure( $opt_config_file, $opt_quiet );
+        exit;
     }
 
     # Need to call parse_local_date() before any resetting of
@@ -204,7 +222,7 @@
     $pkg->date_init();
     my $today = UnixDate($now, '%Q');
 
-    my %urls = $pkg->urls_by_date();
+    my %urls = $pkg->urls_by_date( $opt_config_file, $opt_quiet );
     t 'URLs by date: ' . d \%urls;
 
     my @to_get;
@@ -227,8 +245,8 @@
 	warn "couldn't get any listings from the site for today or later\n";
     }
 
-    my $bar = new Term::ProgressBar('downloading listings', scalar @to_get)
-      if Have_bar && not $opt_quiet;
+    my $bar = new XMLTV::ProgressBar('downloading listings', scalar @to_get)
+      if not $opt_quiet;
     my @listingses;
     foreach my $url (@to_get) {
 	my $xml;
@@ -263,8 +281,9 @@
 	}
 
 	push @listingses, XMLTV::parse($xml);
-	update $bar if Have_bar && not $opt_quiet;
+	update $bar if not $opt_quiet;
     }
+    $bar->finish();
     my %w_args = ();
     if (defined $opt_output) {
 	my $fh = new IO::File ">$opt_output";
@@ -290,6 +309,45 @@
 
 =pod
 
+=item XMLTV::Grab_XML->remove_early_stop_times()
+
+Checks each stop time and removes it if it's before the start time.
+
+Argument: the XML to correct
+Returns: the corrected XML
+
+=cut
+
+my $warned_bad_stop_time = 0;
+sub remove_early_stop_times( $$ ) {
+    my $pkg = shift;
+    my @lines = split /\n/, shift;
+    foreach (@lines) {
+	if (/<programme/) {
+	    # First change to numeric timezones.
+	    s{(start|stop)="(\d+) ([A-Z]+)"}
+	    {qq'$1="$2 ' . tz_to_num($3) . '"'}eg;
+	    
+	    # Now remove stop times before start.  Only worry about
+	    # cases where the timezone is the same - we hope the
+	    # upstream data will be fixed by the next TZ changeover.
+	    #
+	    /start="(\d+) (\S+)"/ or next;
+	    my ($start, $tz) = ($1, $2);
+	    /stop="(\d+) \Q$tz\E"/ or next;
+	    my $stop = $1;
+	    if ($stop lt $start) {
+		warn "removing stop time before start time: $_"
+		  unless $warned_bad_stop_time++;
+		s/stop="[^""]+"\s*// or die;
+	    }
+	}
+    }
+    return join('', @lines);
+}
+
+=pod
+
 =back
 
 =head1 AUTHOR
Index: grab/de_tvtoday/tv_grab_de_tvtoday.in
===================================================================
RCS file: /cvsroot/xmltv/xmltv/grab/de_tvtoday/tv_grab_de_tvtoday.in,v
retrieving revision 1.13
diff -u -B -b -w -r1.13 tv_grab_de_tvtoday.in
--- grab/de_tvtoday/tv_grab_de_tvtoday.in	9 May 2004 17:49:11 -0000	1.13
+++ grab/de_tvtoday/tv_grab_de_tvtoday.in	5 Jun 2004 14:51:56 -0000
@@ -11,7 +11,7 @@
 
 tv_grab_de_tvtoday --help
 
-tv_grab_de_tvtoday [--config-file FILE] --configure
+tv_grab_de_tvtoday [--config-file FILE] --configure [--gui OPTION]
 
 tv_grab_de_tvtoday [--config-file FILE] [--output FILE] 
                    [--days N] [--offset N]
@@ -37,6 +37,11 @@
 default is B<~/.xmltv/tv_grab_de_tvtoday.conf>.  This is the file 
 written by B<--configure> and read when grabbing.
 
+B<--gui OPTION> Use this option to enable a graphical interface to be used.
+OPTION may be 'Tk' or 'GDialog', or left blank for the best available choice.
+Additional allowed values of OPTION are 'Term' for normal terminal output
+(default) and 'TermNoProgressBar' to disable the use of Term::ProgressBar.
+
 B<--output FILE> write to FILE rather than standard output.
 
 B<--days N> grab N days.  The default is seven.
@@ -90,6 +95,8 @@
 use URI::Escape;
 use XMLTV;
 use XMLTV::Ask;
+use XMLTV::ProgressBar;
+use XMLTV::Europe_TZ;
 use XMLTV::DST;
 use XMLTV::Config_file;
 use XMLTV::Mode;
@@ -97,7 +104,7 @@
 use XMLTV::Memoize;
 use XMLTV::Usage <<END
 $0: get German television listings from www.tvtoday.de in XMLTV format
-To configure: $0 --configure [--config-file FILE]
+To configure: $0 --configure [--config-file FILE] [--gui OPTION]
 To grab data: $0 [--config-file FILE] [--output FILE] 
                  [--days N] [--offset N]
                  [--quiet] [--slow] [--nosqueezeout]
@@ -139,9 +146,6 @@
 my $debug = 0;
 $XMLTV::Get_nice::Delay = 0 if($debug);
 
-#-- Use Term::ProgressBar if installed.
-use constant Have_bar => eval { require Term::ProgressBar; 1; };
-
 #-- attributes of xmltv root element
 my $head = { 
     'source-data-url'      => 'http://www.tvtoday.de/tv/programm/programm.php',
@@ -159,6 +163,7 @@
 
 my $opt_configure;
 my $opt_config_file;
+my $opt_gui;
 my $opt_output;
 my $opt_days;
 my $opt_offset = 0;
@@ -172,6 +177,7 @@
 GetOptions(
     'configure'      => \$opt_configure,
     'config-file=s'  => \$opt_config_file,
+    'gui:s'          => \$opt_gui,
     'output=s'       => \$opt_output,
     'days=i'         => \$opt_days,
     'offset=i'       => \$opt_offset,
@@ -185,6 +191,9 @@
 
 usage(1) if $opt_help;
 
+XMLTV::Ask::init($opt_gui);
+XMLTV::ProgressBar::init($opt_gui);
+
 #-- make sure offset+days arguments are within range
 die "offset mustn't be larger than six"
   if($opt_offset > 6);
@@ -287,7 +296,7 @@
     my @chs = sort keys %channels;
     my @names = map { $channels{$_} } @chs;
     my @qs = map { "add channel $_?" } @names;
-    my @want = askManyBooleanQuestions(1, @qs);
+    my @want = ask_many_boolean(1, @qs);
 
     foreach (@chs) {
 	my $w = shift @want;
@@ -355,16 +364,18 @@
 
 #-- write out <programme> tags
 my $numdays = $opt_days + $opt_offset - 1;
-my $bar = new Term::ProgressBar('grabbing', scalar(@requests) * $opt_days)
-  if Have_bar && not $opt_quiet;
+  
+my $bar = new XMLTV::ProgressBar('grabbing', scalar(@requests) * $opt_days)
+  if not $opt_quiet;
 
 foreach my $channel (@requests) {
     for (my $day = $opt_offset; $day <= $numdays; $day ++) {
 	grab_data($channel, $day, $day == $numdays);
-	update $bar if Have_bar && not $opt_quiet;
+	update $bar if not $opt_quiet;
     }
 }
 
+$bar->finish();
 
 #-- hey, looks like we've finished ...
 $writer->end();
@@ -389,7 +400,7 @@
     my $lday = shift @_;    #- true: last day to grab in row
 
     print STDERR "grabbing channel $channels{$ch} (offset=$offset) ...\n"
-      unless(Have_bar || $opt_quiet);
+      unless($opt_quiet);
 	    
     #- we got to send ztag=8 to retrieve data for yesterday
     $offset = 8 if($offset < 0);
Index: grab/dk/tv_grab_dk
===================================================================
RCS file: /cvsroot/xmltv/xmltv/grab/dk/tv_grab_dk,v
retrieving revision 1.16
diff -u -B -b -w -r1.16 tv_grab_dk
--- grab/dk/tv_grab_dk	23 May 2004 16:23:05 -0000	1.16
+++ grab/dk/tv_grab_dk	5 Jun 2004 14:51:56 -0000
@@ -10,7 +10,7 @@
 
 tv_grab_dk --help
 
-tv_grab_dk [--config-file FILE] --configure
+tv_grab_dk [--config-file FILE] --configure [--gui OPTION]
 
 tv_grab_dk [--config-file FILE] [--output FILE] [--days N]
            [--offset N] [--quiet]
@@ -32,6 +32,11 @@
 default is B<~/.xmltv/tv_grab_dk.conf>.  This is the file written by
 B<--configure> and read when grabbing.
 
+B<--gui OPTION> Use this option to enable a graphical interface to be used.
+OPTION may be 'Tk' or 'GDialog', or left blank for the best available choice.
+Additional allowed values of OPTION are 'Term' for normal terminal output
+(default) and 'TermNoProgressBar' to disable the use of Term::ProgressBar.
+
 B<--output FILE> write to FILE rather than standard output.
 
 B<--days N> grab N days.  The default is one week.
@@ -81,6 +86,7 @@
 
 use XMLTV;
 use XMLTV::Memoize;
+use XMLTV::ProgressBar;
 use XMLTV::Ask;
 use XMLTV::Mode;
 use XMLTV::Config_file;
@@ -111,8 +117,7 @@
     }
 }
 
-# Use Term::ProgressBar if installed.
-use constant Have_bar => eval { require Term::ProgressBar; 1 };
+
 
 # Whether zero-length programmes should be included in the output.
 my $WRITE_ZERO_LENGTH = 0;
@@ -132,8 +137,8 @@
 # Get options
 XMLTV::Memoize::check_argv('XMLTV::Get_nice::get_nice_aux');
 my ($opt_days, $opt_offset, $opt_help, $opt_output,
-    $opt_configure, $opt_config_file, $opt_quiet,
-    $opt_list_channels);
+    $opt_configure, $opt_config_file, $opt_gui,
+    $opt_quiet, $opt_list_channels);
 $opt_days   = 7; # default
 $opt_offset = 0; # default
 GetOptions('days=i'        => \$opt_days,
@@ -141,6 +146,7 @@
 	   'help'          => \$opt_help,
 	   'configure'     => \$opt_configure,
 	   'config-file=s' => \$opt_config_file,
+	   'gui:s'         => \$opt_gui,
 	   'output=s'      => \$opt_output,
 	   'quiet'         => \$opt_quiet,
 	   'list-channels' => \$opt_list_channels,
@@ -150,6 +156,9 @@
   if (defined $opt_days && $opt_days < 0);
 usage(1) if $opt_help;
 
+XMLTV::Ask::init($opt_gui);
+XMLTV::ProgressBar::init($opt_gui);
+
 my $mode = XMLTV::Mode::mode('grab', # default
 			     $opt_configure => 'configure',
 			     $opt_list_channels => 'list-channels',
@@ -163,17 +172,18 @@
     XMLTV::Config_file::check_no_overwrite($config_file);
     open(CONF, ">$config_file") or die "cannot write to $config_file: $!";
     # find list of available channels
-    my $bar = new Term::ProgressBar('getting list of channels', 1)
-      if Have_bar && not $opt_quiet;
+    my $bar = new XMLTV::ProgressBar('getting list of channels', 1)
+      if not $opt_quiet;
     my %channels = get_channels();
     die 'no channels could be found' if (scalar(keys(%channels)) == 0);
-    update $bar if Have_bar && not $opt_quiet;
+    update $bar if not $opt_quiet;
+    $bar->finish() if not $opt_quiet;
 
     # Ask about each channel.
     my @chs = sort keys %channels;
     my @names = map { $channels{$_} } @chs;
     my @qs = map { "add channel $_?" } @names;
-    my @want = askManyBooleanQuestions(1, @qs);
+    my @want = ask_many_boolean(1, @qs);
     foreach (@chs) {
 	my $w = shift @want;
 	warn("cannot read input, stopping channel questions"), last
@@ -221,11 +231,11 @@
    });
 
 if ($opt_list_channels) {
-    my $bar = new Term::ProgressBar('getting list of channels', 1)
-      if Have_bar && not $opt_quiet;
+    my $bar = new XMLTV::ProgressBar('getting list of channels', 1)
+      if not $opt_quiet;
     my %channels = get_channels();
     die 'no channels could be found' if (scalar(keys(%channels)) == 0);
-    update $bar if Have_bar && not $opt_quiet;
+    update $bar if not $opt_quiet;
 
     foreach my $ch_did (sort(keys %channels)) {
 	my $ch_name = $channels{$ch_did};
@@ -233,6 +243,7 @@
 	$writer->write_channel({ id => $ch_xid,
 				 'display-name' => [ [ $ch_name ] ] });
     }
+    $bar->finish() if not $opt_quiet;
     $writer->end();
     exit();
 }
@@ -292,16 +303,17 @@
 
 my %warned_ch_name; # suppress duplicate warnings
 
-my $bar = new Term::ProgressBar('fetching data', scalar @to_get)
-  if Have_bar && not $opt_quiet;
+my $bar = new XMLTV::ProgressBar('fetching data', scalar @to_get)
+  if not $opt_quiet;
 my @to_get_detailed;
 my $num_detailed = 0;
 foreach (@to_get) {
     my ($url, $date, $ch_xmltv_id, $ch_tvgids_id) = @$_;
     t "going to get $ch_xmltv_id for $date";
     process_listings_page($writer, $ch_xmltv_id, $url, $date);
-    update $bar if Have_bar && not $opt_quiet;
+    update $bar if not $opt_quiet;
 }
+$bar->finish() if not $opt_quiet;
 $writer->end();
 
 ######################################################################
Index: grab/es/tv_grab_es
===================================================================
RCS file: /cvsroot/xmltv/xmltv/grab/es/tv_grab_es,v
retrieving revision 1.27
diff -u -B -b -w -r1.27 tv_grab_es
--- grab/es/tv_grab_es	9 May 2004 17:49:11 -0000	1.27
+++ grab/es/tv_grab_es	5 Jun 2004 14:51:57 -0000
@@ -10,7 +10,7 @@
 
 tv_grab_es --help
 
-tv_grab_es [--config-file FILE] --configure
+tv_grab_es [--config-file FILE] --configure [--gui OPTION]
 
 tv_grab_es [--config-file FILE] [--output FILE] [--days N]
            [--offset N] [--quiet]
@@ -38,6 +38,11 @@
 default is B<~/.xmltv/tv_grab_es.conf>.  This is the file written by
 B<--configure> and read when grabbing.
 
+B<--gui OPTION> Use this option to enable a graphical interface to be used.
+OPTION may be 'Tk' or 'GDialog', or left blank for the best available choice.
+Additional allowed values of OPTION are 'Term' for normal terminal output
+(default) and 'TermNoProgressBar' to disable the use of XMLTV::ProgressBar.
+
 B<--output FILE> write to FILE rather than standard output.
 
 B<--days N> grab N days.  The default is 3.
@@ -99,6 +104,7 @@
 
 use XMLTV;
 use XMLTV::Memoize;
+use XMLTV::ProgressBar;
 use XMLTV::Ask;
 use XMLTV::Config_file;
 use XMLTV::DST;
@@ -115,9 +121,6 @@
 END
   ;
 
-# Use Term::ProgressBar if installed.
-use constant Have_bar => eval { require Term::ProgressBar; 1 };
-
 # Attributes of the root element in output.
 my $HEAD = { 'source-info-url'     => 'http://www.elpais.es/parrillatv/portada.html',
 	     'source-data-url'     => "http://www.elpais.es/parrillatv/resultados.html",
@@ -140,8 +143,8 @@
 # Get options, including undocumented --cache option.
 XMLTV::Memoize::check_argv('XMLTV::Get_nice::get_nice_aux');
 my ($opt_days, $opt_offset, $opt_help, $opt_output,
-    $opt_configure, $opt_config_file, $opt_quiet,
-    $opt_list_channels);
+    $opt_configure, $opt_config_file, $opt_gui,
+    $opt_quiet, $opt_list_channels);
 $opt_days  = 3; # default
 $opt_offset = 0; # default
 $opt_quiet  = 0; # default
@@ -150,6 +153,7 @@
 	   'help'          => \$opt_help,
 	   'configure'     => \$opt_configure,
 	   'config-file=s' => \$opt_config_file,
+           'gui:s'         => \$opt_gui,
 	   'output=s'      => \$opt_output,
 	   'quiet'         => \$opt_quiet,
 	   'list-channels' => \$opt_list_channels
@@ -159,6 +163,9 @@
   if (defined $opt_days && $opt_days < 0);
 usage(1) if $opt_help;
 
+XMLTV::Ask::init($opt_gui);
+XMLTV::ProgressBar::init($opt_gui);
+
 my $mode = XMLTV::Mode::mode('grab', # default
 			     $opt_configure => 'configure',
 			     $opt_list_channels => 'list-channels',
@@ -194,7 +201,7 @@
     my @chs = sort keys %channels;
     my @names = map { $channels{$_} } @chs;
     my @qs = map { "add channel $_?" } @names;
-    my @want = askManyBooleanQuestions(1, @qs);
+    my @want = ask_many_boolean(1, @qs);
     foreach (@chs) {
 	my $w = shift @want;
 	warn("cannot read input, stopping channel questions"), last
@@ -287,14 +294,15 @@
 # This progress bar is for both downloading and parsing.  Maybe
 # they could be separate.
 #
-my $bar = new Term::ProgressBar('getting listings', scalar @to_get)
-  if Have_bar && not $opt_quiet;
+my $bar = new XMLTV::ProgressBar('getting listings', scalar @to_get)
+  if not $opt_quiet;
 foreach (@to_get) {
 	foreach (process_table($_->[0], $_->[1], $_->[2])) {
 		$writer->write_programme($_);
 	}
-	update $bar if Have_bar && not $opt_quiet;
+	update $bar if not $opt_quiet;
 }
+$bar->finish() if not $opt_quiet;
 $writer->end();
 
 ######################################################################
@@ -493,8 +501,8 @@
 
 # get channel listing
 sub get_channels {
-    my $bar = new Term::ProgressBar('getting list of channels', 1)
-	if Have_bar && not $opt_quiet;
+    my $bar = new XMLTV::ProgressBar('getting list of channels', 1)
+	if not $opt_quiet;
     my %channels;
     my $url="http://www.elpais.es/parrillatv/portada.html";
     t $url;
@@ -529,7 +537,8 @@
 	}
     }
     die "no channels could be found" if not keys %channels;
-    update $bar if Have_bar && not $opt_quiet;
+    update $bar if not $opt_quiet;
+    $bar->finish() if not $opt_quiet;
     return %channels;
 }
 
Index: grab/es_digital/tv_grab_es_digital
===================================================================
RCS file: /cvsroot/xmltv/xmltv/grab/es_digital/tv_grab_es_digital,v
retrieving revision 1.4
diff -u -B -b -w -r1.4 tv_grab_es_digital
--- grab/es_digital/tv_grab_es_digital	9 May 2004 17:49:12 -0000	1.4
+++ grab/es_digital/tv_grab_es_digital	5 Jun 2004 14:51:58 -0000
@@ -10,7 +10,7 @@
 
 tv_grab_es_digital --help
 
-tv_grab_es_digital [--config-file FILE] --configure
+tv_grab_es_digital [--config-file FILE] --configure [--gui OPTION]
 
 tv_grab_es_digital [--config-file FILE] [--output FILE] [--days N]
            [--offset N] [--quiet]
@@ -36,6 +36,11 @@
 default is B<~/.xmltv/tv_grab_es_digital.conf>.  This is the file written by
 B<--configure> and read when grabbing.
 
+B<--gui OPTION> Use this option to enable a graphical interface to be used.
+OPTION may be 'Tk' or 'GDialog', or left blank for the best available choice.
+Additional allowed values of OPTION are 'Term' for normal terminal output
+(default) and 'TermNoProgressBar' to disable the use of Term::ProgressBar.
+
 B<--output FILE> write to FILE rather than standard output.
 
 B<--days N> grab N days.  The default is 3.
@@ -95,6 +100,7 @@
 
 use XMLTV;
 use XMLTV::Memoize;
+use XMLTV::ProgressBar;
 use XMLTV::Ask;
 use XMLTV::Config_file;
 use XMLTV::DST;
@@ -110,9 +116,6 @@
 END
   ;
 
-# Use Term::ProgressBar if installed.
-use constant Have_bar => eval { require Term::ProgressBar; 1 };
-
 # Attributes of the root element in output.
 my $HEAD = { 'source-info-url'     => 'http://http://www.plus.es/codigo/television/rejilla_dia_dia.asp',
 	     'source-data-url'     => "http://http://www.plus.es/codigo/television/rejilla_dia_dia.asp",
@@ -135,8 +138,8 @@
 # Get options, including undocumented --cache option.
 XMLTV::Memoize::check_argv('XMLTV::Get_nice::get_nice_aux');
 my ($opt_days, $opt_offset, $opt_help, $opt_output,
-    $opt_configure, $opt_config_file, $opt_quiet,
-    $opt_list_channels);
+    $opt_configure, $opt_config_file, $opt_gui,
+    $opt_quiet, $opt_list_channels);
 $opt_days  = 3; # default
 $opt_offset = 0; # default
 $opt_quiet  = 0; # default
@@ -145,6 +148,7 @@
 	   'help'          => \$opt_help,
 	   'configure'     => \$opt_configure,
 	   'config-file=s' => \$opt_config_file,
+       'gui:s'         => \$opt_gui,
 	   'output=s'      => \$opt_output,
 	   'quiet'         => \$opt_quiet,
 	   'list-channels' => \$opt_list_channels
@@ -154,6 +158,9 @@
   if (defined $opt_days && $opt_days < 0);
 usage(1) if $opt_help;
 
+XMLTV::Ask::init($opt_gui);
+XMLTV::ProgressBar::init($opt_gui);
+
 my $mode = XMLTV::Mode::mode('grab', # default
 			     $opt_configure => 'configure',
 			     $opt_list_channels => 'list-channels',
@@ -189,7 +196,7 @@
     my @chs = sort keys %channels;
     my @names = map { $channels{$_} } @chs;
     my @qs = map { "add channel $_?" } @names;
-    my @want = askManyBooleanQuestions(1, @qs);
+    my @want = ask_many_boolean(1, @qs);
     foreach (@chs) {
 	my $w = shift @want;
 	warn("cannot read input, stopping channel questions"), last
@@ -280,16 +287,18 @@
 # This progress bar is for both downloading and parsing.  Maybe
 # they could be separate.
 #
-my $bar = new Term::ProgressBar('getting listings', scalar @to_get)
-  if Have_bar && not $opt_quiet;
+my $bar = new XMLTV::ProgressBar('getting listings', scalar @to_get)
+  if not $opt_quiet;
 foreach (@to_get) {
 	foreach (process_table($_->[0], $_->[1], $_->[2])) {
 		$writer->write_programme($_);
 	}
-	update $bar if Have_bar && not $opt_quiet;
+	update $bar if not $opt_quiet;
 }
 $writer->end();
 
+$bar->finish();
+
 ######################################################################
 # subroutine definitions
 
@@ -482,8 +491,8 @@
 
 # get channel listing
 sub get_channels {
-    my $bar = new Term::ProgressBar('getting list of channels', 1)
-	if Have_bar && not $opt_quiet;
+    my $bar = new XMLTV::ProgressBar('getting list of channels', 1)
+	if not $opt_quiet;
     my %channels;
     my $url="http://www.plus.es/codigo/television/plataformas/digitalplus/portada2.asp";
     t $url;
@@ -519,7 +528,8 @@
 	}
     }
     die "no channels could be found" if not keys %channels;
-    update $bar if Have_bar && not $opt_quiet;
+    update $bar if not $opt_quiet;
+    $bar->finish();
     return %channels;
 }
 
Index: grab/fi/tv_grab_fi
===================================================================
RCS file: /cvsroot/xmltv/xmltv/grab/fi/tv_grab_fi,v
retrieving revision 1.34
diff -u -B -b -w -r1.34 tv_grab_fi
--- grab/fi/tv_grab_fi	23 May 2004 16:23:05 -0000	1.34
+++ grab/fi/tv_grab_fi	5 Jun 2004 14:51:58 -0000
@@ -10,7 +10,7 @@
 
 tv_grab_fi --help
 
-tv_grab_fi [--config-file FILE] --configure
+tv_grab_fi [--config-file FILE] --configure [--gui OPTION]
 
 tv_grab_fi [--config-file FILE] [--output FILE] [--days N]
            [--offset N] [--quiet]
@@ -35,6 +35,11 @@
 default is B<~/.xmltv/tv_grab_fi.conf>.  This is the file written by
 B<--configure> and read when grabbing.
 
+B<--gui OPTION> Use this option to enable a graphical interface to be used.
+OPTION may be 'Tk' or 'GDialog', or left blank for the best available choice.
+Additional allowed values of OPTION are 'Term' for normal terminal output
+(default) and 'TermNoProgressBar' to disable the use of XMLTV::ProgressBar.
+
 B<--output FILE> write to FILE rather than standard output.
 
 B<--days N> grab N days.  The default is ten.
@@ -79,6 +84,7 @@
 
 use XMLTV;
 use XMLTV::Memoize;
+use XMLTV::ProgressBar;
 use XMLTV::Ask;
 use XMLTV::Config_file;
 use XMLTV::DST;
@@ -95,9 +101,6 @@
 END
   ;
 
-# Use Term::ProgressBar if installed.
-use constant Have_bar => eval { require Term::ProgressBar; 1 };
-
 # Attributes of the root element in output.
 my $HEAD = { 'source-info-url'     => 'http://www.katso.fi/',
 	     'source-data-url'     => "http://www.katso.fi/tvopas",
@@ -124,8 +127,8 @@
 # Get options, including undocumented --cache option.
 XMLTV::Memoize::check_argv('XMLTV::Get_nice::get_nice_aux');
 my ($opt_days, $opt_offset, $opt_help, $opt_output,
-    $opt_configure, $opt_config_file, $opt_quiet,
-    $opt_list_channels);
+    $opt_configure, $opt_config_file, $opt_gui,
+    $opt_quiet, $opt_list_channels);
 $opt_days  = 10; # default
 $opt_offset = 0; # default
 $opt_quiet  = 0; # default
@@ -134,6 +137,7 @@
 	   'help'          => \$opt_help,
 	   'configure'     => \$opt_configure,
 	   'config-file=s' => \$opt_config_file,
+           'gui:s'         => \$opt_gui,
 	   'output=s'      => \$opt_output,
 	   'quiet'         => \$opt_quiet,
 	   'list-channels' => \$opt_list_channels,
@@ -143,6 +147,9 @@
   if (defined $opt_days && $opt_days < 0);
 usage(1) if $opt_help;
 
+XMLTV::Ask::init($opt_gui);
+XMLTV::ProgressBar::init($opt_gui);
+
 my $mode = XMLTV::Mode::mode('grab', # default
 			     $opt_configure => 'configure',
 			     $opt_list_channels => 'list-channels',
@@ -178,7 +185,7 @@
     my @chs = sort keys %channels;
     my @names = map { $channels{$_} } @chs;
     my @qs = map { "add channel $_?" } @names;
-    my @want = askManyBooleanQuestions(1, @qs);
+    my @want = ask_many_boolean(1, @qs);
     foreach (@chs) {
 	my $w = shift @want;
 	warn("cannot read input, stopping channel questions"), last
@@ -268,14 +275,15 @@
 # This progress bar is for both downloading and parsing.  Maybe
 # they could be separate stages.
 #
-my $bar = new Term::ProgressBar('getting listings', scalar @to_get)
-  if Have_bar && not $opt_quiet;
+my $bar = new XMLTV::ProgressBar('getting listings', scalar @to_get)
+  if not $opt_quiet;
 foreach (@to_get) {
     foreach (process_table($_->[0], $_->[1], $_->[2])) {
 	$writer->write_programme($_);
     }
-    update $bar if Have_bar && not $opt_quiet;
+    update $bar if not $opt_quiet;
 }
+$bar->finish() if not $opt_quiet;
 $writer->end();
 
 ######################################################################
@@ -503,8 +511,8 @@
 
 # get channel listing
 sub get_channels {
-    my $bar = new Term::ProgressBar('getting list of channels', 1)
-      if Have_bar && not $opt_quiet;
+    my $bar = new XMLTV::ProgressBar('getting list of channels', 1)
+      if not $opt_quiet;
     my %channels;
     my $url="http://www.katso.fi/tvopas";
     my $local_data=get_nice($url);
@@ -546,7 +554,8 @@
 	}
     }
     die "no channels could be found" if not keys %channels;
-    update $bar if Have_bar && not $opt_quiet;
+    update $bar if not $opt_quiet;
+    $bar->finish() if not $opt_quiet;
     return %channels;
 }
 
Index: grab/fr/tv_grab_fr
===================================================================
RCS file: /cvsroot/xmltv/xmltv/grab/fr/tv_grab_fr,v
retrieving revision 1.6
diff -u -B -b -w -r1.6 tv_grab_fr
--- grab/fr/tv_grab_fr	23 May 2004 16:23:13 -0000	1.6
+++ grab/fr/tv_grab_fr	5 Jun 2004 14:51:59 -0000
@@ -6,7 +6,7 @@
 
 =head1 SYNOPSIS
 
-To configure: tv_grab_fr --configure [--config-file FILE]
+To configure: tv_grab_fr --configure [--config-file FILE] [--gui OPTION]
 To grab listings: tv_grab_fr [--output FILE] [--quiet]
 Slower, detailed grab: tv_grab_fr --slow [--output FILE] [--days N] [--offset N] [--quiet]
 Help: tv_grab_fr --help
@@ -23,6 +23,11 @@
 B<--configure> Grab channels informations from the website and ask for
 channel type and names.
 
+B<--gui OPTION> Use this option to enable a graphical interface to be used.
+OPTION may be 'Tk' or 'GDialog', or left blank for the best available choice.
+Additional allowed values of OPTION are 'Term' for normal terminal output
+(default) and 'TermNoProgressBar' to disable the use of Term::ProgressBar.
+
 B<--output FILE> write to FILE rather than standard output.
 
 B<--days N> grab N days starting from today, rather than as many as
@@ -62,9 +67,6 @@
 END
   ;
 
-# Use Term::ProgressBar if installed.
-use constant Have_bar => eval { require Term::ProgressBar; 1 };
-
 use warnings;
 use strict;
 use XMLTV::Version '$Id$ ';
@@ -77,6 +79,7 @@
 use XMLTV;
 use XMLTV::Memoize;
 use XMLTV::Ask;
+use XMLTV::ProgressBar;
 use XMLTV::Mode;
 use XMLTV::Config_file;
 use XMLTV::DST;
@@ -108,7 +111,7 @@
 # Global variables allocation according to options
 #***************************************************************************
 XMLTV::Memoize::check_argv('get_page_aux');
-my ($opt_days,  $opt_help,  $opt_output,  $opt_offset,  $opt_quiet,  $opt_list_channels, $opt_config_file, $opt_configure, $opt_slow, $opt_licons);
+my ($opt_days,  $opt_help,  $opt_output,  $opt_offset,  $opt_gui, $opt_quiet,  $opt_list_channels, $opt_config_file, $opt_configure, $opt_slow, $opt_licons);
 $opt_quiet  = 0;
 # The website is able to store up to nine days from now
 my $default_opt_days = 9;
@@ -120,6 +123,7 @@
      'quiet'     => \$opt_quiet,
      'configure' => \$opt_configure,
      'config-file=s' => \$opt_config_file,
+     'gui:s'     => \$opt_gui,
      'list-channels' => \$opt_list_channels,
      'slow' => \$opt_slow
     )
@@ -132,6 +136,9 @@
 die 'Cannot get more than one day before current day' if (defined $opt_offset && $opt_offset < -1);
 usage(1) if $opt_help;
 
+XMLTV::Ask::init($opt_gui);
+XMLTV::ProgressBar::init($opt_gui);
+
 if (not $opt_slow) {
     # Certain options are ignored in fast mode.
     my %slow_options = (days => $opt_days,
@@ -253,11 +260,11 @@
     my @gts = sort keys %GridType;
     my @gtnames = map { $GridType{$_} } @gts;
     my @gtqs = map { "Get channels type : $_?" } @gts;
-    my @gtwant = askManyBooleanQuestions(1, @gtqs);
+    my @gtwant = ask_many_boolean(1, @gtqs);
 
-    my $bar = new Term::ProgressBar('getting channel lists',
+    my $bar = new XMLTV::ProgressBar('getting channel lists',
                                     scalar grep { $_ } @gtwant)
-                    if Have_bar && not $opt_quiet;
+                    if not $opt_quiet;
     my %channels_for;
     foreach my $i (0 .. $#gts) {
         my ($gt, $gtw, $gtname) = ($gts[$i], $gtwant[$i], $gtnames[$i]);
@@ -265,8 +272,9 @@
         my %channels = get_channels( $gtname );
         die 'No channels could be found' if not %channels;
         $channels_for{$gt} = \%channels;
-        update $bar if Have_bar && not $opt_quiet;
+        update $bar if not $opt_quiet;
     }
+    $bar->finish();
 
     my %asked;
     foreach (@gts) {
@@ -280,7 +288,7 @@
             my @chs = grep { not $asked{$_}++ } sort keys %channels;
             my @names = map { $channels{$_}{name} } @chs;
             my @qs = map { "add channel $_?" } @names;
-            my @want = askManyBooleanQuestions(1, @qs);
+            my @want = ask_many_boolean(1, @qs);
             foreach (@chs) {
                 my $w = shift @want;
                 warn("cannot read input, stopping channel questions"), last if not defined $w;
@@ -332,7 +340,7 @@
     my @gts = sort keys %GridType;
     my @gtnames = map { $GridType{$_} } @gts;
     my @gtqs = map { "List channels for grid : $_?" } @gts;
-    my @gtwant = askManyBooleanQuestions(1, @gtqs);
+    my @gtwant = ask_many_boolean(1, @gtqs);
 
     foreach (@gts) {
         my $gtw = shift @gtwant;
@@ -393,13 +401,14 @@
         push @to_get, [ $url, $chid, $_ ];
     }
 }
-my $bar = new Term::ProgressBar('getting listings', scalar @to_get)  if Have_bar && not $opt_quiet;
+my $bar = new XMLTV::ProgressBar('getting listings', scalar @to_get)  if not $opt_quiet;
 foreach (@to_get) {
     my ($url, $chid, $slot ) = @$_;
     process_channel_grid_page($writer, $chid, $url, $slot);
-    update $bar if Have_bar && not $opt_quiet;
+    update $bar if not $opt_quiet;
 }
 $writer->end();
+$bar->finish();
 
 #***************************************************************************
 # Specific functions for grabbing information
Index: grab/huro/tv_grab_huro
===================================================================
RCS file: /cvsroot/xmltv/xmltv/grab/huro/tv_grab_huro,v
retrieving revision 1.3
diff -u -B -b -w -r1.3 tv_grab_huro
--- grab/huro/tv_grab_huro	9 May 2004 17:49:12 -0000	1.3
+++ grab/huro/tv_grab_huro	5 Jun 2004 14:52:00 -0000
@@ -10,7 +10,7 @@
 
 tv_grab_huro --help
 
-tv_grab_huro [--config-file FILE] --configure
+tv_grab_huro [--config-file FILE] --configure [--gui OPTION]
 
 tv_grab_huro [--config-file FILE] [--output FILE] [--days N]
            [--offset N] [--quiet]
@@ -34,6 +34,11 @@
 default is B<~/.xmltv/tv_grab_huro.conf>.  This is the file written by
 B<--configure> and read when grabbing.
 
+B<--gui OPTION> Use this option to enable a graphical interface to be used.
+OPTION may be 'Tk' or 'GDialog', or left blank for the best available choice.
+Additional allowed values of OPTION are 'Term' for normal terminal output
+(default) and 'TermNoProgressBar' to disable the use of Term::ProgressBar.
+
 B<--output FILE> write to FILE rather than standard output.
 
 B<--days N> grab N days.  The default is eight.
@@ -79,6 +84,7 @@
 
 use XMLTV;
 use XMLTV::Memoize;
+use XMLTV::ProgressBar;
 use XMLTV::Ask;
 use XMLTV::DST;
 use XMLTV::Get_nice;
@@ -88,7 +94,7 @@
 # Todo: perhaps we should internationalize messages and docs?
 use XMLTV::Usage <<END
 $0: get Hungarian or Romanian television listings in XMLTV format
-To configure: $0 --configure [--config-file FILE]
+To configure: $0 --configure [--config-file FILE] [--gui OPTION]
 To grab listings: $0 [--config-file FILE] [--output FILE] [--days N]
         [--offset N] [--quiet]
 To list channels: $0 --list-channels --loc [hu | ro]
@@ -108,9 +114,6 @@
 sub get_channels( $ );
 sub nextday( $ );
 
-# Use Term::ProgressBar if installed.
-use constant Have_bar => eval { require Term::ProgressBar; 1 };
-
 # Attributes of the root element in output.
 sub domain( $ ) { "port.$_[0]" };
 sub xhead( $ ) {
@@ -138,7 +141,7 @@
 # Get options, including undocumented --cache option.
 XMLTV::Memoize::check_argv('XMLTV::Get_nice::get_nice_aux');
 my ($opt_days, $opt_offset, $opt_help, $opt_output,
-    $opt_configure, $opt_config_file, $opt_quiet,
+    $opt_configure, $opt_config_file, $opt_gui, $opt_quiet,
     $opt_list_channels, $opt_loc);
 $opt_days = 8; # default
 $opt_offset = 0; # default
@@ -148,6 +151,7 @@
 	   'help'          => \$opt_help,
 	   'configure'     => \$opt_configure,
 	   'config-file=s' => \$opt_config_file,
+	   'gui:s'         => \$opt_gui,
 	   'output=s'      => \$opt_output,
 	   'quiet'         => \$opt_quiet,
 	   'list-channels' => \$opt_list_channels,
@@ -162,6 +166,9 @@
 			     $opt_list_channels => 'list-channels',
 			    );
 
+XMLTV::Ask::init($opt_gui);
+XMLTV::ProgressBar::init($opt_gui);
+          
 # File that stores which channels to download.
 my $config_file
   = XMLTV::Config_file::filename($opt_config_file, 'tv_grab_huro', $opt_quiet, 'tv_grab_hu');
@@ -185,7 +192,7 @@
     open(CONF, ">$config_file") or die "cannot write to $config_file: $!";
 
     my $default_cn = 'Hungary';
-    my $cn = askQuestion('Grab listings for which country?',
+    my $cn = ask_choice('Grab listings for which country?',
 			 $default_cn,
 			 sort keys %COUNTRIES);
     my $c = $COUNTRIES{$cn}[0];
@@ -196,7 +203,7 @@
     my @chs = sort keys %channels;
     my @names = map { $channels{$_} } @chs;
     my @qs = map { "add channel $_?" } @names;
-    my @want = askManyBooleanQuestions(1, @qs);
+    my @want = ask_many_boolean(1, @qs);
     foreach (@chs) {
 	my $w = shift @want;
 	warn("cannot read input, stopping channel questions"), last
@@ -327,8 +334,8 @@
 # This progress bar is for both downloading and parsing.  Maybe
 # they could be separate stages.
 #
-my $bar = new Term::ProgressBar('getting listings', @days * @channels)
-  if Have_bar && not $opt_quiet;
+my $bar = new XMLTV::ProgressBar('getting listings', @days * @channels)
+  if not $opt_quiet;
 foreach my $d (@days) {
     my ($day, $i) = @$d;
     my $some_success = 0;
@@ -336,13 +343,14 @@
 	my @ps = process_table($country, $day, xid($country, $ch_did), $ch_did, $i);
 	$some_success = 1 if @ps;
 	$writer->write_programme($_) foreach @ps;
-	update $bar if Have_bar && not $opt_quiet;
+	update $bar if not $opt_quiet;
     }
     if (@channels and not $some_success) {
 	warn "failed to get any listings for day $i, stopping\n";
 	last;
     }
 }
+$bar->finish() if not $opt_quiet;
 $writer->end();
 
 ######################################################################
@@ -534,8 +542,8 @@
 sub get_channels( $ ) {
     my $country = shift;
     my $d = domain($country);
-    my $bar = new Term::ProgressBar('getting list of channels', 1)
-	if Have_bar && not $opt_quiet;
+    my $bar = new XMLTV::ProgressBar('getting list of channels', 1)
+	if not $opt_quiet;
     my %channels;
     my $url="http://www.$d/pls/tv/tv.prog";
     my $local_data=get_nice($url);
@@ -572,7 +580,8 @@
 	}
     }
     die "no channels could be found" if not keys %channels;
-    update $bar if Have_bar && not $opt_quiet;
+    update $bar if not $opt_quiet;
+    $bar->finish() if not $opt_quiet;
     return %channels;
 }
 
Index: grab/it/tv_grab_it.in
===================================================================
RCS file: /cvsroot/xmltv/xmltv/grab/it/tv_grab_it.in,v
retrieving revision 1.27
diff -u -B -b -w -r1.27 tv_grab_it.in
--- grab/it/tv_grab_it.in	23 May 2004 16:23:13 -0000	1.27
+++ grab/it/tv_grab_it.in	5 Jun 2004 14:52:00 -0000
@@ -11,7 +11,7 @@
 
 tv_grab_it --help
 
-tv_grab_it [--config-file FILE] --configure
+tv_grab_it [--config-file FILE] --configure [--gui OPTION]
 
 tv_grab_it [--config-file FILE] [--output FILE] [--days N]
            [--offset N] [--quiet] [--slow]
@@ -33,6 +33,11 @@
 default is B<~/.xmltv/tv_grab_it.conf>.  This is the file written
 by B<--configure> and read when grabbing.
 
+B<--gui OPTION> Use this option to enable a graphical interface to be used.
+OPTION may be 'Tk' or 'GDialog', or left blank for the best available choice.
+Additional allowed values of OPTION are 'Term' for normal terminal output
+(default) and 'TermNoProgressBar' to disable the use of XMLTV::ProgressBar.
+
 B<--output FILE> write to FILE rather than standard output.
 
 B<--days N> grab N days.  The default is 7.
@@ -81,6 +86,7 @@
 use Memoize;
 use XMLTV;
 use XMLTV::Memoize;
+use XMLTV::ProgressBar;
 use XMLTV::Ask;
 use XMLTV::Config_file;
 use XMLTV::DST;
@@ -88,7 +94,7 @@
 # Todo: perhaps we should internationalize messages and docs?
 use XMLTV::Usage <<END
 $0: get Italian television listings in XMLTV format
-To configure: $0 --configure [--config-file FILE]
+To configure: $0 --configure [--config-file FILE] [--gui OPTION]
 To grab listings: $0 [--config-file FILE] [--output FILE] [--days N]
         [--offset N] [--quiet] [--slow]
 END
@@ -117,11 +123,6 @@
 my $base="http://$domain/canali.phtml";
 my $rturl="http://$domain/";
 
-######################################################################
-# Use Term::ProgressBar if installed.
-use constant Have_bar => eval { require Term::ProgressBar; 1 };
-
-######################################################################
 # get options
 # Get options, including undocumented --cache option.
 
@@ -137,6 +138,7 @@
     $opt_slow,
     $opt_configure,
     $opt_config_file,
+    $opt_gui,
     $opt_quiet,
     $opt_share,
    );
@@ -152,6 +154,7 @@
        'help'          => \$opt_help,
        'configure'     => \$opt_configure,
        'config-file=s' => \$opt_config_file,
+       'gui:s'         => \$opt_gui,
        'output=s'      => \$opt_output,
        'quiet'         => \$opt_quiet,
        'slow'      => \$opt_slow,
@@ -162,6 +165,9 @@
   if (defined $opt_days && $opt_days < 0);
 usage(1) if $opt_help;
 
+XMLTV::Ask::init($opt_gui);
+XMLTV::ProgressBar::init($opt_gui);
+
 # share/ directory for storing channel mapping files.  This next line
 # is altered by processing through tv_grab_it.PL.  But we can use the
 # current directory instead of share/tv_grab_it for development.
@@ -195,17 +201,17 @@
         }
     open(CONF, ">$config_file") or die "cannot write to $config_file: $!";
     # find list of available channels
-    my $bar = new Term::ProgressBar('getting list of channels', 1)
-      if Have_bar && not $opt_quiet;
+    my $bar = new XMLTV::ProgressBar('getting list of channels', 1)
+      if not $opt_quiet;
     my %channels=get_channels_list($content);
     die "no channels could be found" if (scalar(keys(%channels))==0);
-    update $bar if Have_bar && not $opt_quiet;
+    update $bar if not $opt_quiet;
 
     # Ask about each channel.
     my @chs = sort keys %channels;
     my @names = map { $channels{$_} } @chs;
     my @qs = map { "add channel $_?" } @names;
-    my @want = askManyBooleanQuestions(1, @qs);
+    my @want = ask_many_boolean(1, @qs);
     foreach (@chs) {
     my $w = shift @want;
     warn("cannot read input, stopping channel questions"), last
@@ -218,6 +224,7 @@
         print CONF "channel $_\n";
     }
 
+    $bar->finish() if not $opt_quiet;
     close CONF or warn "cannot close $config_file: $!";
     say("Finished configuration.");
 
@@ -296,8 +303,8 @@
 }
 t "will get $days2get days from $opt_offset onwards";
 
-my $bar2 = new Term::ProgressBar('getting icons', scalar @channels)
-  if Have_bar && not $opt_quiet;
+my $bar2 = new XMLTV::ProgressBar('getting icons', scalar @channels)
+  if not $opt_quiet;
 foreach my $ch_id (@channels) {
     my $ch_xid=xmltv_chanid($ch_id);
 
@@ -315,11 +322,12 @@
                'display-name' => [ [ $ch_id ] ],
                icon => [{src => get_icon($url)}]
               });
-    update $bar2 if Have_bar && not $opt_quiet;
+    update $bar2 if not $opt_quiet;
 }
+$bar2->finish() if not $opt_quiet;
 
-my $bar = new Term::ProgressBar('getting listings', scalar @to_get)
-  if Have_bar && not $opt_quiet;
+my $bar = new XMLTV::ProgressBar('getting listings', scalar @to_get)
+  if not $opt_quiet;
 foreach (@to_get) {
     my $canale= $_->[1];
     $url   = $_->[0];
@@ -346,9 +354,10 @@
     $w->write_programme($_) foreach @dati;
     }
     #           else {warn "skipping $canale \n";}
-    update $bar if Have_bar && not $opt_quiet;
+    update $bar if not $opt_quiet;
 }
 $w->end;
+$bar->finish() if not $opt_quiet;
 
 ######################################################################
 # subroutines
Index: grab/jp/tv_grab_jp
===================================================================
RCS file: /cvsroot/xmltv/xmltv/grab/jp/tv_grab_jp,v
retrieving revision 1.4
diff -u -B -b -w -r1.4 tv_grab_jp
--- grab/jp/tv_grab_jp	9 May 2004 17:49:13 -0000	1.4
+++ grab/jp/tv_grab_jp	5 Jun 2004 14:52:01 -0000
@@ -10,7 +10,7 @@
 
 tv_grab_jp --help
 
-tv_grab_jp [--config-file FILE] --configure
+tv_grab_jp [--config-file FILE] --configure [--gui OPTION]
 
 tv_grab_jp [--config-file FILE] [--output FILE] [--days N]
            [--offset N] [--enable-readstr] [--no-category-code]
@@ -37,6 +37,11 @@
 default is B<~/.xmltv/tv_grab_jp.conf>.  This is the file written by
 B<--configure> and read when grabbing.
 
+B<--gui OPTION> Use this option to enable a graphical interface to be used.
+OPTION may be 'Tk' or 'GDialog', or left blank for the best available choice.
+Additional allowed values of OPTION are 'Term' for normal terminal output
+(default) and 'TermNoProgressBar' to disable the use of Term::ProgressBar.
+
 B<--output FILE> write to FILE rather than standard output.
 
 B<--days N> grab N days.  The default and the maximum is 7.
@@ -86,6 +91,7 @@
 use XMLTV;
 use XMLTV::Memoize;
 use XMLTV::Ask;
+use XMLTV::ProgressBar;
 use XMLTV::Config_file;
 use XMLTV::Mode;
 use XMLTV::Get_nice;
@@ -106,9 +112,6 @@
 END
   ;
 
-# Use Term::ProgressBar if installed.
-use constant Have_bar => eval { require Term::ProgressBar; 1 };
-
 # Attributes of the root element in output.
 my $HEAD = { 'source-info-url'     => 'http://www.ontvjapan.com/',
 	     'source-data-url'     => "http://www.ontvjapan.com/program/",
@@ -162,7 +165,7 @@
 
 XMLTV::Memoize::check_argv('XMLTV::Get_nice::get_nice_aux');
 my ($opt_days, $opt_offset, $opt_help, $opt_output,
-    $opt_configure, $opt_config_file, $opt_quiet,
+    $opt_configure, $opt_config_file, $opt_quiet, $opt_gui,
     $opt_readstr, $opt_no_cat_code, $opt_list_channels);
 $opt_days        = 7; # default
 $opt_offset      = 0; # default
@@ -172,6 +175,7 @@
 GetOptions('help'             => \$opt_help,
 	   'configure'        => \$opt_configure,
 	   'config-file=s'    => \$opt_config_file,
+       'gui:s'            => \$opt_gui,
 	   'output=s'         => \$opt_output,
 	   'quiet'            => \$opt_quiet,
 	   'enable-readstr'   => \$opt_readstr,
@@ -187,6 +191,9 @@
     if (defined $opt_offset && $opt_offset < 0);
 usage(1) if $opt_help;
 
+XMLTV::Ask::init($opt_gui);
+XMLTV::ProgressBar::init($opt_gui);
+
 my $mode = XMLTV::Mode::mode('grab', # default
 			     $opt_configure => 'configure',
 			     $opt_list_channels => 'list-channels',
@@ -214,9 +221,10 @@
     # Get region information
     my $barmsg = select_lang('地域情報を取得中',
 			     'getting regions');
-    my $bar = new Term::ProgressBar($barmsg, 1) if Have_bar;
+    my $bar = new XMLTV::ProgressBar($barmsg, 1);
     my %regions = get_regions();
-    update $bar if Have_bar;
+    update $bar;
+    $bar->finish();
 
     # Ask region
     my %ask_regions;
@@ -243,7 +251,7 @@
 
     $msg = select_lang("\n地域を指定してください: ",
 		       "\nSelect your region: ");
-    $req_region = askQuestion($msg, $req_region, @r);
+    $req_region = ask_choice($msg, $req_region, @r);
     $regionid = $ask_regions{$req_region};
 
     if ($regionid eq 'CATV') {
@@ -260,9 +268,10 @@
 
     $barmsg = select_lang('チャンネル情報を取得中',
 			     'getting channels');
-    $bar = new Term::ProgressBar($barmsg, 1) if Have_bar;
+    $bar = new XMLTV::ProgressBar($barmsg, 1);
     my @channels = get_channels($regionid);
-    update $bar if Have_bar;
+    update $bar;
+    $bar->finish();
 
     # Ask about each channel.
     Text::Kakasi::getopt_argv('kakasi', '-ieuc', '-Ja', '-Ha', '-Ka', '-Ea');
@@ -281,7 +290,7 @@
     }
     Text::Kakasi::close_kanwadict();
 
-    my @want = askManyBooleanQuestions(1, @qs);
+    my @want = ask_many_boolean(1, @qs);
     foreach (@channels) {
 	my $w = shift @want;
 	my $id = $_->{id};
@@ -297,7 +306,7 @@
 
     close CONF or warn "cannot close $config_file: $!";
     $msg = select_lang("設定完了.\n", "Finished\n");
-    print $msg;
+    say( $msg );
 
     exit();
 }
@@ -324,16 +333,16 @@
     # Get region information
     my $barmsg = select_lang('地域情報を取得中',
 			     'getting regions');
-    my $bar = new Term::ProgressBar($barmsg, 1) if Have_bar;
+    my $bar = new XMLTV::ProgressBar($barmsg, 1);
     my %regions = get_regions();
-    update $bar if Have_bar;
+    update $bar;
+    $bar->finish();
 
     # get channels for all region
     my @ch_regions = grep { $_ ne 'CATV' } sort keys %regions;
     $barmsg = select_lang('チャンネル情報を取得中',
 			  'getting channels');
-    $bar = new Term::ProgressBar($barmsg, scalar keys %regions)
-	if Have_bar;
+    $bar = new XMLTV::ProgressBar($barmsg, scalar keys %regions);
     foreach (@ch_regions) {
 	my $regionid = $_;
 	my @channels = get_channels($regionid);
@@ -352,8 +361,9 @@
 				'callsign' => $callsign };
 	    }
 	}
-	update $bar if Have_bar;
+	update $bar;
     }
+    $bar->finish();
 
     # Write channels mode.
     foreach (@ch_all) {
@@ -395,9 +405,10 @@
 
 my $barmsg = select_lang('チャンネル情報を取得中',
 			 'getting channels');
-my $bar = new Term::ProgressBar($barmsg, 1) if Have_bar;
+my $bar = new XMLTV::ProgressBar($barmsg, 1);
 my @channels = get_channels($regionid);
-update $bar if Have_bar;
+update $bar;
+$bar->finish();
 
 foreach (@channels) {
     my $id = $_->{id};
@@ -419,13 +430,14 @@
 # This progress bar is for both downloading and parsing.
 $barmsg = select_lang('番組表を取得中',
 		      'getting program info');
-$bar = new Term::ProgressBar($barmsg, scalar @ch_all)
-    if Have_bar && not $opt_quiet;
+$bar = new XMLTV::ProgressBar($barmsg, scalar @ch_all)
+    if not $opt_quiet;
 foreach (@ch_all) {
     process_table($_->{'id'});
-    update $bar if Have_bar && not $opt_quiet;
+    update $bar if not $opt_quiet;
 }
 $writer->end();
+$bar->finish();
 
 ######################################################################
 # subroutine definitions
Index: grab/na_dd/tv_grab_na_dd.in
===================================================================
RCS file: /cvsroot/xmltv/xmltv/grab/na_dd/tv_grab_na_dd.in,v
retrieving revision 1.5
diff -u -B -b -w -r1.5 tv_grab_na_dd.in
--- grab/na_dd/tv_grab_na_dd.in	22 May 2004 20:22:30 -0000	1.5
+++ grab/na_dd/tv_grab_na_dd.in	5 Jun 2004 14:52:03 -0000
@@ -14,6 +14,7 @@
     
     tv_grab_na_dd --configure [--config-file FILE] [--dd-data FILE]
                               [--reprocess] [--auto-config add|ignore]
+                              [--gui OPTION]
 
     tv_grab_na_dd --list-lineups [--config-file FILE] [--dd-data FILE]
                               [--reprocess]
@@ -50,6 +51,13 @@
 
 Activates configure mode,  If a config file already exists the values are used as defaults.
 
+=item --gui OPTION
+
+Use this option to enable a graphical interface to be used.
+OPTION may be 'Tk' or 'GDialog', or left blank for the best available choice.
+Additional allowed values of OPTION are 'Term' for normal terminal output
+(default) and 'TermNoProgressBar' to disable the use of Term::ProgressBar.
+
 =item --list-lineups
 
 Lists available lineups.  Only requires username in the config file. Used by programs that
@@ -275,11 +283,11 @@
 use File::Temp qw(tempfile);
 use Getopt::Long;
 use XML::Twig 3.10;
-use constant Have_bar => eval { require Term::ProgressBar; 1 };
 
 use XMLTV;
 use XMLTV::Ask;
 use XMLTV::Config_file;
+use XMLTV::ProgressBar;
 use XMLTV::Version '$Id$ ';
 use XMLTV::Usage <<END
 $0: get lists from Zap2IT's DataDirect service in XMLTV format
@@ -287,7 +295,7 @@
     $0 [--version] [--help]
   
 To Configure
-    $0 --configure [--config-file FILE] [--dd-data FILE]
+    $0 --configure [--config-file FILE] [--dd-data FILE] [--gui OPTION]
                               [--reprocess] [--auto-config add|ignore]
 To get listings
     $0 [--config-file FILE] [--dd-data FILE]
@@ -339,6 +347,7 @@
 my $opt_help;           # ask for help
 my $opt_configure;      # configure mode
 my $opt_config_file ;   # config_file_name
+my $opt_gui ;           # use a gui for configuration
 my $opt_output;         # output name
 my $opt_days       =10; # days to fetch
 my $opt_offset     =0;  # day to start
@@ -365,7 +374,7 @@
 	       'help'          => \$opt_help,
     	   'configure'     => \$opt_configure,
     	   'config-file=s' => \$opt_config_file,
-    	   'config_file=s' => \$opt_config_file,
+    	   'gui:s'         => \$opt_gui,
     	   'output=s'      => \$opt_output,
            'days=i'        => \$opt_days,
     	   'offset=i'      => \$opt_offset,
@@ -396,6 +405,10 @@
 die "--down-only without --dd-data is pointless!\n" if $opt_down_only && ! $opt_dd_data;
 
 $opt_days    = 0 if $opt_configure || $opt_list_channels || $opt_list_lineups;
+
+XMLTV::Ask::init($opt_gui);
+XMLTV::ProgressBar::init($opt_gui);
+
 $config_file = XMLTV::Config_file::filename($opt_config_file, 'tv_grab_na_dd', $opt_quiet);
 
 
@@ -489,7 +502,7 @@
 
     ");
         $dd_user=ask("Username ($dd_user):")|| $dd_user || die "DataDirect Username Required\n";
-        $dd_pass=ask("
+        $dd_pass=ask_password("
 WARNING: Storing the password in the config file is not secure
 If password is blank, it will be prompted as needed(more secure)
 Unsecured password ('x':delete,default:<keep>,):")|| $dd_pass;
@@ -575,7 +588,7 @@
 #
 # Fetch data
 # 
-    print STDERR "Fetching from DataDirect\n";
+    my $bar = new XMLTV::ProgressBar( 'Fetching from DataDirect', 1 );
     print STDERR "    dd_data is in $dd_data_name\n" if $DEBUG || $opt_dd_data;   
 
     sub SOAP::Transport::HTTP::Client::get_basic_credentials
@@ -612,6 +625,8 @@
     die 'got empty result from SOAP call' if $dd_data_size == 0;
     $time = int(time() - $time);
     printf STDERR "    Fetched %d k/bytes in %d seconds\n",$dd_data_size/1024,$time;
+    $bar->update();
+    $bar->finish();
 } # get data
 
 #
@@ -733,8 +748,7 @@
 
 unless ($opt_quiet)
 {
-    if (Have_bar) { $bar = new Term::ProgressBar('loading data',$dd_data_size+1) }
-    else          { print STDERR "loading data "; }
+    $bar = new XMLTV::ProgressBar('loading data',$dd_data_size+1);
 }
 seek($dd_data,0,0);  #rewind
 eval { $twig->parse( $dd_data ) };
@@ -752,7 +766,7 @@
 }
 
 $bar->update($dd_data_size+1) if $bar;
-print STDERR "\n" unless ($opt_quiet || $bar );
+$bar->finish() if $bar;
 
 $twig=undef;  # destroy twig (just in case)
 
@@ -810,7 +824,6 @@
 if ($opt_configure)
 {
     my %chan_found=();
-    my $key;
     $dd_lineup=$old_lineups{$dd_lineup} if exists $old_lineups{$dd_lineup};
     $dd_lineup=(sort keys %lineups)[0]  unless exists $lineups{$dd_lineup};
     if (! $opt_auto_config)
@@ -823,51 +836,110 @@
                                    $lineups{$dd_lineup}{name},
                                    $lineups{$dd_lineup}{type});
                                            
-        $val = askQuestion("\nWhich Lineup? ($dd_lineup)",$val,@choices);
+        $val = ask_choice("\nWhich Lineup? ($dd_lineup)",$val,@choices);
         $dd_lineup = (split(/\|/,$val))[0];
     }
     print CONF "lineup: $dd_lineup\n";
 
-        $opt_auto_config='add' if !$opt_auto_config && !askBooleanQuestion("
+        $opt_auto_config='add' if !$opt_auto_config && !ask_boolean("
 The preferred method for controlling the channel lineup is through
 the DataDirect web site, but you can omit channels here as well.
 Do you want to skip some channels?",0);
 
     print "\n";
     
-    foreach (@{$lineups{$dd_lineup}{map}})
-    {
-       my $res='yes';
-       my $key=sprintf("%s %s",$_->{channel},$station{$_->{station}}{callSign});
-       $chan_found{$key}=1;
+    #foreach (@{$lineups{$dd_lineup}{map}})
+    #{
+    #   my $res='yes';
+    #   my $key=sprintf("%s %s",$_->{channel},$station{$_->{station}}{callSign});
+    #   $chan_found{$key}=1;
+    #   if ($opt_auto_config)
+    #   {
+    #        if (exists $chan_config{$key})
+    #        {
+    #        }
+    #        elsif ($opt_auto_config eq 'add')
+    #        {
+    #            print STDERR "Adding new channel: $key\n";
+    #            $chan_config{$key}=1;
+    #        }
+    #        else
+    #        {
+    #            print STDERR "Ignoring new channel: (see docs about bandwidth issues) $key\n";
+    #            $chan_config{$key}=0;
+    #        }
+    #   }
+    #   else
+    #   {
+    #        $res='no' if exists $chan_config{$key} and ! $chan_config{$key};
+    #        $res = ask_choice("Add channel $key?",$res,qw(yes no all none ));
+    #        $chan_config{$key}=1       if $res=~/^[ya]/i;
+    #        $chan_config{$key}=0       if $res=~/^[n]/i;
+    #        $opt_auto_config='add'     if $res=~/^all/i;
+    #        $opt_auto_config='ign'     if $res=~/^none/i;
+    #   } # ask question (or --auto-new )
+
+    #   print CONF ( $chan_config{$key} ? '' : 'not ' ), "channel: $key\n";
+    #} # channel loop
+
+    # If the user expressed a default preference
        if ($opt_auto_config)
        {
-            if (exists $chan_config{$key})
+        # Either add all the channels
+        if ($opt_auto_config eq 'add')
             {
-            }
-            elsif ($opt_auto_config eq 'add')
+            foreach (@{$lineups{$dd_lineup}{map}})
             {
-                print STDERR "Adding new channel: $key\n";
-                $chan_config{$key}=1;
+                my $key1=sprintf("%s %s",$_->{channel}, 
+                    $station{$_->{station}}{callSign});
+                print STDERR "Adding new channel: $key1\n";
+                $chan_config{$key1}=1;
+            }
             }
+        # or ignore them all
             else
             {
-                print STDERR "Ignoring new channel: (see docs about bandwidth issues) $key\n";
-                $chan_config{$key}=0;
+            foreach (@{$lineups{$dd_lineup}{map}})
+            {  
+                my $key2=sprintf("%s %s",$_->{channel}, 
+                    $station{$_->{station}}{callSign});
+                print STDERR "Ignoring new channel: (see docs about bandwidth issues) $key2\n";
+                $chan_config{$key2}=0;
             }
        }
-       else
+    }
+    else  # There was no default for channels, so we ask the user
+    {
+        # Construct the questions
+        my @questions;
+        foreach (@{$lineups{$dd_lineup}{map}})
        {
-            $res='no' if exists $chan_config{$key} and ! $chan_config{$key};
-            $res = askQuestion("Add channel $key?",$res,qw(yes no all none ));
-            $chan_config{$key}=1       if $res=~/^[ya]/i;
-            $chan_config{$key}=0       if $res=~/^[n]/i;
-            $opt_auto_config='add'     if $res=~/^all/i;
-            $opt_auto_config='ign'     if $res=~/^none/i;
-       } # ask question (or --auto-new )
+            my $key3=sprintf("%s %s",$_->{channel}, 
+                $station{$_->{station}}{callSign});
 
-       print CONF ( $chan_config{$key} ? '' : 'not ' ), "channel: $key\n";
-    } # channel loop
+            push @questions, ["Add channel $key3?"];
+        }
+        # Ask the questions
+        my @answers = ask_many_boolean( 1, @questions );
+        # Save the answers
+        my $i=0;
+        foreach (@{$lineups{$dd_lineup}{map}})
+        {  
+            my $key4=sprintf("%s %s",$_->{channel}, 
+                $station{$_->{station}}{callSign});
+            
+            $chan_config{$key4} = $answers[$i];
+            $i++;
+        }
+    }
+    
+    # Write the choices out to the config file
+    foreach (@{$lineups{$dd_lineup}{map}})
+    {  
+        my $key5=sprintf("%s %s",$_->{channel}, 
+            $station{$_->{station}}{callSign});
+        print CONF ( $chan_config{$key5} ? '' : 'not ' ), "channel: $key5\n";
+    }
 
     foreach (sort keys %chan_config)
     {
@@ -875,6 +947,7 @@
         print STDERR "Channel '$_' no longer exists\n";
     }
     close CONF;
+    say( 'Configuration complete!' );
     exit 0;
 } # --configure channel list
 
@@ -1256,13 +1329,12 @@
     seek($dd_data,0,0);  #rewind
     unless ($opt_quiet)
     {
-        if (Have_bar) { $bar = new Term::ProgressBar('Writing schedule',$dd_data_size+1) }
-        else          { print STDERR "Writing schedule"; }
+        $bar = new XMLTV::ProgressBar('Writing schedule',$dd_data_size+1);
     }
     my $time=time();
     $twig->parse( $dd_data );
-    print STDERR "\n" unless ($opt_quiet || $bar );
     $bar->update($dd_data_size+1) if $bar;
+    $bar->finish() if $bar;
 
     $writer->end();
     if (@messages)
Index: grab/na_icons/tv_grab_na_icons.in
===================================================================
RCS file: /cvsroot/xmltv/xmltv/grab/na_icons/tv_grab_na_icons.in,v
retrieving revision 1.2
diff -u -B -b -w -r1.2 tv_grab_na_icons.in
--- grab/na_icons/tv_grab_na_icons.in	23 May 2004 17:38:53 -0000	1.2
+++ grab/na_icons/tv_grab_na_icons.in	5 Jun 2004 14:52:03 -0000
@@ -139,7 +139,7 @@
     map {s/\xa0//g} @names;  # remove non-breaking spaces
 
     my $name=$names[0];    
-    $name=askQuestion("\nLineup?",$name,@names);
+    $name=ask_choice("\nLineup?",$name,@names);
 
     foreach (0..$#names)
     {
Index: grab/nl/tv_grab_nl.in
===================================================================
RCS file: /cvsroot/xmltv/xmltv/grab/nl/tv_grab_nl.in,v
retrieving revision 1.9
diff -u -B -b -w -r1.9 tv_grab_nl.in
--- grab/nl/tv_grab_nl.in	23 May 2004 16:23:14 -0000	1.9
+++ grab/nl/tv_grab_nl.in	5 Jun 2004 14:52:04 -0000
@@ -84,6 +84,7 @@
 use Date::Manip;
 use XMLTV;
 use XMLTV::Memoize;
+use XMLTV::ProgressBar;
 use XMLTV::Ask;
 use XMLTV::Config_file;
 use XMLTV::DST;
@@ -113,9 +114,6 @@
     }
 }
 
-# Use Term::ProgressBar if installed.
-use constant Have_bar => eval { require Term::ProgressBar; 1 };
-
 # Function prototypes.
 sub time_tot_str( $ );
 sub time_van_str( $ );
@@ -141,7 +139,7 @@
 # Get options, including undocumented --cache option.
 XMLTV::Memoize::check_argv('XMLTV::Get_nice::get_nice_aux');
 my ($opt_days, $opt_offset, $opt_help, $opt_output,
-    $opt_configure, $opt_config_file, $opt_quiet,
+    $opt_configure, $opt_config_file, $opt_gui, $opt_quiet,
     $opt_list_channels, $opt_slow, $opt_share);
 $opt_days   = 7; # default
 $opt_offset = 0; # default
@@ -150,6 +148,7 @@
 	   'help'          => \$opt_help,
 	   'configure'     => \$opt_configure,
 	   'config-file=s' => \$opt_config_file,
+           'gui:s'         => \$opt_gui,
 	   'output=s'      => \$opt_output,
 	   'quiet'         => \$opt_quiet,
 	   'slow'	   => \$opt_slow,
@@ -174,6 +173,9 @@
 my $OUR_SHARE_DIR = (defined $SHARE_DIR) ? "$SHARE_DIR/tv_grab_nl" : '.';
 (my $CHANNELS_FILE = "$OUR_SHARE_DIR/channels") =~ tr!/!/!s;
 
+XMLTV::Ask::init($opt_gui);
+XMLTV::ProgressBar::init($opt_gui);
+
 my $mode = XMLTV::Mode::mode('grab', # default
 			     $opt_configure => 'configure',
 			     $opt_list_channels => 'list-channels',
@@ -191,17 +193,18 @@
     # configuring it's useful to download at least something as a
     # check that the site is reachable etc.
     #
-    my $bar = new Term::ProgressBar('getting list of channels', 1)
-      if Have_bar && not $opt_quiet;
+    my $bar = new XMLTV::ProgressBar('getting list of channels', 1)
+      if not $opt_quiet;
     my %channels = get_channels(1);
     die 'no channels could be found' if (scalar(keys(%channels)) == 0);
-    update $bar if Have_bar && not $opt_quiet;
+    update $bar if not $opt_quiet;
+    $bar->finish();
 
     # Ask about each channel.
     my @chs = sort { $a <=> $b } keys %channels;
     my @names = map { $channels{$_} } @chs;
     my @qs = map { "add channel $_?" } @names;
-    my @want = askManyBooleanQuestions(1, @qs);
+    my @want = ask_many_boolean(1, @qs);
     foreach (@chs) {
 	my $w = shift @want;
 	warn("cannot read input, stopping channel questions"), last
@@ -248,15 +251,15 @@
    });
 
 if ($mode eq 'list-channels') {
-    my $bar = new Term::ProgressBar('getting list of channels', 1)
-      if Have_bar && not $opt_quiet;
+    my $bar = new XMLTV::ProgressBar('getting list of channels', 1)
+      if not $opt_quiet;
 
     # Don't do any page fetches - assume the contents of the channels
     # file has already been checked enough.
     #
     my %channels = get_channels(0);
     die 'no channels could be found' if (scalar(keys(%channels)) == 0);
-    update $bar if Have_bar && not $opt_quiet;
+    update $bar if not $opt_quiet;
 
     foreach my $ch_did (sort(keys %channels)) {
 	my $ch_name = $channels{$ch_did};
@@ -265,6 +268,7 @@
 				 'display-name' => [ [ $ch_name ] ] });
     }
     $writer->end();
+    $bar->finish();
     exit();
 }
 
@@ -337,16 +341,17 @@
 my %warned_ch_name; # suppress duplicate warnings
 
 my @summary_page_data;
-my $bar = new Term::ProgressBar('downloading summary', scalar @to_get)
-  if Have_bar && not $opt_quiet;
+my $bar = new XMLTV::ProgressBar('downloading summary', scalar @to_get)
+  if not $opt_quiet;
 foreach (@to_get) {
     my ($url, $day, $ch_xmltv_id, $ch_tvgids_id) = @$_;
     die if ref $url;
     push @summary_page_data,
       [ $url, $ch_xmltv_id,
 	[ process_summary_page($url, $day, $day, \$channels{$ch_tvgids_id}) ] ];
-    update $bar if Have_bar and not $opt_quiet;
+    update $bar if not $opt_quiet;
 }
+$bar->finish();
 
 # Now we've fetched the descriptions, we know the channel names.
 foreach $ch_did (@channels) {
@@ -469,8 +474,8 @@
 
 my @to_write;
 if ($opt_slow) {
-    $bar = new Term::ProgressBar('getting details', scalar @summary_programmes)
-      if Have_bar && not $opt_quiet;
+    $bar = new XMLTV::ProgressBar('getting details', scalar @summary_programmes)
+      if not $opt_quiet;
     foreach my $s (@summary_programmes) {
 	my $urls = delete $s->{url};
 	if (not defined $urls) {
@@ -565,6 +570,7 @@
 	push @to_write, $detailed;
 	update $bar if $bar;
     }
+    $bar->finish();
 }
 else {
     @to_write = @summary_programmes;
Index: grab/nl_wolf/tv_grab_nl_wolf
===================================================================
RCS file: /cvsroot/xmltv/xmltv/grab/nl_wolf/tv_grab_nl_wolf,v
retrieving revision 1.12
diff -u -B -b -w -r1.12 tv_grab_nl_wolf
--- grab/nl_wolf/tv_grab_nl_wolf	9 May 2004 17:49:13 -0000	1.12
+++ grab/nl_wolf/tv_grab_nl_wolf	5 Jun 2004 14:52:04 -0000
@@ -85,7 +85,7 @@
 my $url_base = "$URL_HOST$URL_DIR";
 
 # Returns a hash mapping YYYMMDD to URL.
-sub urls_by_date( $ ) {
+sub urls_by_date( $$$ ) {
     my $pkg = shift;
     my $index = $pkg->get($url_base);
     die "could not get index page $url_base, aborting\n"
@@ -139,29 +139,9 @@
 my $warned_bad_stop_time = 0;
 sub xml_from_data( $$ ) {
     my $pkg = shift;
-    my @lines = split /\n/, shift;
-    foreach (@lines) {
-	if (/<programme/) {
-	    # First change to numeric timezones.
-	    s{(start|stop)="(\d+) ([A-Z]+)"}
-	    {qq'$1="$2 ' . tz_to_num($3) . '"'}eg;
+    my $xml = shift;
 	    
-	    # Now remove stop times before start.  Only worry about
-	    # cases where the timezone is the same - we hope the
-	    # upstream data will be fixed by the next TZ changeover.
-	    #
-	    /start="(\d+) (\S+)"/ or next;
-	    my ($start, $tz) = ($1, $2);
-	    /stop="(\d+) \Q$tz\E"/ or next;
-	    my $stop = $1;
-	    if ($stop lt $start) {
-		warn "removing stop time before start time: $_"
-		  unless $warned_bad_stop_time++;
-		s/stop="[^""]+"\s*// or die;
-	    }
-	}
-    }
-    return join('', @lines);
+    return pkg->remove_early_stop_times( $xml );
 }
 
 Grab_XML_nl_wolf->go();
Index: grab/no/tv_grab_no
===================================================================
RCS file: /cvsroot/xmltv/xmltv/grab/no/tv_grab_no,v
retrieving revision 1.10
diff -u -B -b -w -r1.10 tv_grab_no
--- grab/no/tv_grab_no	23 May 2004 16:23:14 -0000	1.10
+++ grab/no/tv_grab_no	5 Jun 2004 14:52:05 -0000
@@ -10,7 +10,7 @@
 
 tv_grab_no --help
 
-tv_grab_no [--config-file FILE] --configure
+tv_grab_no [--config-file FILE] --configure [--gui OPTION]
 
 tv_grab_no [--config-file FILE] [--output FILE] [--days N]
            [--offset N] [--quiet]
@@ -32,6 +32,11 @@
 default is B<~/.xmltv/tv_grab_no.conf>.  This is the file written by
 B<--configure> and read when grabbing.
 
+B<--gui OPTION> Use this option to enable a graphical interface to be used.
+OPTION may be 'Tk' or 'GDialog', or left blank for the best available choice.
+Additional allowed values of OPTION are 'Term' for normal terminal output
+(default) and 'TermNoProgressBar' to disable the use of Term::ProgressBar.
+
 B<--output FILE> write to FILE rather than standard output.
 
 B<--days N> grab N days.  The default is as many as the source carries.
@@ -76,6 +81,7 @@
 
 use XMLTV;
 use XMLTV::Memoize;
+use XMLTV::ProgressBar;
 use XMLTV::Ask;
 use XMLTV::Mode;
 use XMLTV::Config_file;
@@ -85,7 +91,7 @@
 # Todo: perhaps we should internationalize messages and docs?
 use XMLTV::Usage <<END
 $0: get Norwegian television listings in XMLTV format
-To configure: $0 --configure [--config-file FILE]
+To configure: $0 --configure [--config-file FILE] [--gui OPTION]
 To grab listings: $0 [--config-file FILE] [--output FILE] [--days N]
         [--offset N] [--quiet]
 END
@@ -108,9 +114,6 @@
     }
 }
 
-# Use Term::ProgressBar if installed.
-use constant Have_bar => eval { require Term::ProgressBar; 1 };
-
 # Whether zero-length programmes should be included in the output.
 my $WRITE_ZERO_LENGTH = 0;
 
@@ -129,7 +132,7 @@
 # Get options
 XMLTV::Memoize::check_argv('XMLTV::Get_nice::get_nice_aux');
 my ($opt_days, $opt_offset, $opt_help, $opt_output,
-    $opt_configure, $opt_config_file, $opt_quiet,
+    $opt_configure, $opt_config_file, $opt_gui, $opt_quiet,
     $opt_list_channels);
 $opt_days   = 4; # default
 $opt_offset = 0; # default
@@ -138,6 +141,7 @@
 	   'help'          => \$opt_help,
 	   'configure'     => \$opt_configure,
 	   'config-file=s' => \$opt_config_file,
+	   'gui:s'         => \$opt_gui,
 	   'output=s'      => \$opt_output,
 	   'quiet'         => \$opt_quiet,
 	   'list-channels' => \$opt_list_channels,
@@ -146,6 +150,10 @@
 die 'number of days must not be negative'
   if (defined $opt_days && $opt_days < 0);
 usage(1) if $opt_help;
+
+XMLTV::Ask::init($opt_gui);
+XMLTV::ProgressBar::init($opt_gui);
+
 if ($opt_days > 7) {
     print "WARNING: This grabber can only grab 7 days ahead!\n-------: I will grab 7 days and then quit.\n";
     $opt_days = 7;
@@ -164,17 +172,18 @@
     XMLTV::Config_file::check_no_overwrite($config_file);
     open(CONF, ">$config_file") or die "cannot write to $config_file: $!";
     # find list of available channels
-    my $bar = new Term::ProgressBar('getting list of channels', 1)
-      if Have_bar && not $opt_quiet;
+    my $bar = new XMLTV::ProgressBar('getting list of channels', 1)
+      if not $opt_quiet;
     my %channels = get_channels();
     die 'no channels could be found' if (scalar(keys(%channels)) == 0);
-    update $bar if Have_bar && not $opt_quiet;
+    update $bar if not $opt_quiet;
+    $bar->finish() if not $opt_quiet;
 
     # Ask about each channel.
     my @chs = sort keys %channels;
     my @names = map { $channels{$_} } @chs;
     my @qs = map { "add channel $_?" } @names;
-    my @want = askManyBooleanQuestions(1, @qs);
+    my @want = ask_many_boolean(1, @qs);
     foreach (@chs) {
 	my $w = shift @want;
 	warn("cannot read input, stopping channel questions"), last
@@ -222,11 +231,12 @@
    });
 
 if ($opt_list_channels) {
-    my $bar = new Term::ProgressBar('getting list of channels', 1)
-      if Have_bar && not $opt_quiet;
+    my $bar = new XMLTV::ProgressBar('getting list of channels', 1)
+      if not $opt_quiet;
     my %channels = get_channels();
     die 'no channels could be found' if (scalar(keys(%channels)) == 0);
-    update $bar if Have_bar && not $opt_quiet;
+    update $bar if not $opt_quiet;
+    $bar->finish() if not $opt_quiet;
 
     foreach my $ch_did (sort(keys %channels)) {
 	my $ch_name = $channels{$ch_did};
@@ -253,10 +263,11 @@
     }
 }
 
-my $configbar = new Term::ProgressBar('fetching channel names', 1)
-  if Have_bar && not $opt_quiet;
+my $configbar = new XMLTV::ProgressBar('fetching channel names', 1)
+  if not $opt_quiet;
 %channels = getchandisplaynames(@channels);
-update $configbar if Have_bar && not $opt_quiet;
+update $configbar if not $opt_quiet;
+$configbar->finish() if not $opt_quiet;
 
 ######################################################################
 # begin main program
@@ -276,8 +287,8 @@
 }
 
 my %warned_ch_name; # suppress duplicate warnings
-my $bar = new Term::ProgressBar('fetching data', @channels * $opt_days)
-  if Have_bar && not $opt_quiet;
+my $bar = new XMLTV::ProgressBar('fetching data', @channels * $opt_days)
+  if not $opt_quiet;
 foreach my $d (0 .. $opt_days - 1) {
     my $i = $opt_offset + $d;
     my $day = UnixDate(DateCalc($today, "+ $i days"), '%Y-%m-%d');
@@ -295,7 +306,7 @@
 	warn "no listings for channel $ch_xid on day $d, $url\n"
 	  if $got == 0;
 	$num_this_day += $got;
-	update $bar if Have_bar && not $opt_quiet;
+	update $bar if not $opt_quiet;
     }
     if ($num_this_day == 0) {
 	die "could not get any listings\n" if $d == 0;
@@ -304,6 +315,7 @@
 	last;
     }
 }
+$bar->finish() if not $opt_quiet;
 $writer->end();
 
 ######################################################################
Index: grab/pt/tv_grab_pt
===================================================================
RCS file: /cvsroot/xmltv/xmltv/grab/pt/tv_grab_pt,v
retrieving revision 1.6
diff -u -B -b -w -r1.6 tv_grab_pt
--- grab/pt/tv_grab_pt	23 May 2004 16:23:14 -0000	1.6
+++ grab/pt/tv_grab_pt	5 Jun 2004 14:52:05 -0000
@@ -10,7 +10,7 @@
 
 tv_grab_pt --help
 
-tv_grab_pt [--config-file FILE] --configure
+tv_grab_pt [--config-file FILE] --configure [--gui OPTION]
 
 tv_grab_pt [--config-file FILE] [--output FILE] [--days N]
            [--offset N] [--quiet]
@@ -29,6 +29,11 @@
 B<--configure> Prompt for which channels,
 and write the configuration file.
 
+B<--gui OPTION> Use this option to enable a graphical interface to be used.
+OPTION may be 'Tk' or 'GDialog', or left blank for the best available choice.
+Additional allowed values of OPTION are 'Term' for normal terminal output
+(default) and 'TermNoProgressBar' to disable the use of Term::ProgressBar.
+
 B<--config-file FILE> Set the name of the configuration file, the
 default is B<~/.xmltv/tv_grab_pt.conf>.  This is the file written by
 B<--configure> and read when grabbing.
@@ -72,6 +77,7 @@
 
 use XMLTV;
 use XMLTV::Memoize;
+use XMLTV::ProgressBar;
 use XMLTV::Ask;
 use XMLTV::Config_file;
 use XMLTV::DST;
@@ -80,16 +86,13 @@
 # Todo: perhaps we should internationalize messages and docs?
 use XMLTV::Usage <<END
 $0: get Portuguese television listings in XMLTV format
-To configure: $0 --configure [--config-file FILE]
+To configure: $0 --configure [--config-file FILE] [--gui OPTION]
 To grab listings: $0 [--config-file FILE] [--output FILE] [--days N]
         [--offset N] [--quiet]
 To list channels: $0 --list-channels
 END
   ;
 
-# Use Term::ProgressBar if installed.
-use constant Have_bar => eval { require Term::ProgressBar; 1 };
-
 # Attributes of the root element in output.
 my $HEAD = { 'source-info-url'     => 'http://tv.clix.pt',
 	     'source-data-url'     => "http://tv.clix.pt",
@@ -109,7 +112,7 @@
 # Get options, including undocumented --cache option.
 XMLTV::Memoize::check_argv('XMLTV::Get_nice::get_nice_aux');
 my ($opt_days, $opt_offset, $opt_help, $opt_output,
-    $opt_configure, $opt_config_file, $opt_quiet,
+    $opt_configure, $opt_config_file, $opt_gui, $opt_quiet,
     $opt_list_channels);
 $opt_days  = 2; # default
 $opt_offset = 0; # default
@@ -119,6 +122,7 @@
 	   'help'          => \$opt_help,
 	   'configure'     => \$opt_configure,
 	   'config-file=s' => \$opt_config_file,
+	   'gui:s'         => \$opt_gui,
 	   'output=s'      => \$opt_output,
 	   'quiet'         => \$opt_quiet,
 	   'list-channels' => \$opt_list_channels
@@ -130,6 +134,9 @@
   if $opt_days + $opt_offset > 2;
 usage(1) if $opt_help;
 
+XMLTV::Ask::init($opt_gui);
+XMLTV::ProgressBar::init($opt_gui);
+
 my $mode = XMLTV::Mode::mode('grab', # default
 			     $opt_configure => 'configure',
 			     $opt_list_channels => 'list-channels',
@@ -165,7 +172,7 @@
     my @chs = sort keys %channels;
     my @names = map { $channels{$_} } @chs;
     my @qs = map { "add channel $_?" } @names;
-    my @want = askManyBooleanQuestions(1, @qs);
+    my @want = ask_many_boolean(1, @qs);
     foreach (@chs) {
 	my $w = shift @want;
 	warn("cannot read input, stopping channel questions"), last
@@ -239,9 +246,9 @@
 # This progress bar is for both downloading and parsing.  Maybe
 # they could be separate.
 #
-my $bar = new Term::ProgressBar('getting listings',
+my $bar = new XMLTV::ProgressBar('getting listings',
 				scalar(@channels) * $opt_days)
-  if Have_bar && not $opt_quiet;
+  if not $opt_quiet;
 foreach my $ch_did (@channels) {
     my $ch_name=$channels{$ch_did};
     $writer->write_channel({ id => $ch_did,
@@ -256,7 +263,7 @@
 	    $writer->write_programme($_);
 	    $some = 1;
 	}
-	update $bar;
+	update $bar if $bar;
     }
     if (not $some) {
 	die "no programmes found\n" if $i == 0;
@@ -266,7 +273,7 @@
     }
     $date = nextday($date); die if not defined $date;
 }
-
+$bar->finish() if $bar;
 $writer->end();
 
 ######################################################################
@@ -416,8 +423,8 @@
 
 # get channel listing
 sub get_channels {
-    my $bar = new Term::ProgressBar('getting list of channels', 1)
-	if Have_bar && not $opt_quiet;
+    my $bar = new XMLTV::ProgressBar('getting list of channels', 1)
+	if not $opt_quiet;
     my %channels;
     my $url=$HEAD->{'source-info-url'};
     t $url;
@@ -455,7 +462,8 @@
 	}
 	die "no channels could be found in $url\n";
     }
-    update $bar if Have_bar && not $opt_quiet;
+    update $bar if not $opt_quiet;
+    $bar->finish() if not $opt_quiet;
     return %channels;
 }
 
Index: grab/se/tv_grab_se
===================================================================
RCS file: /cvsroot/xmltv/xmltv/grab/se/tv_grab_se,v
retrieving revision 1.12
diff -u -B -b -w -r1.12 tv_grab_se
--- grab/se/tv_grab_se	23 May 2004 18:19:38 -0000	1.12
+++ grab/se/tv_grab_se	5 Jun 2004 14:52:07 -0000
@@ -10,7 +10,7 @@
 
 tv_grab_se --help
 
-tv_grab_se [--config-file FILE] --configure
+tv_grab_se [--config-file FILE] --configure [--gui OPTION]
 
 tv_grab_se [--config-file FILE] [--output FILE] [--days N]
            [--offset N] [--quiet] [--debug]
@@ -34,6 +34,11 @@
 default is B<~/.xmltv/tv_grab_se.conf>.  This is the file written by
 B<--configure> and read when grabbing.
 
+B<--gui OPTION> Use this option to enable a graphical interface to be used.
+OPTION may be 'Tk' or 'GDialog', or left blank for the best available choice.
+Additional allowed values of OPTION are 'Term' for normal terminal output
+(default) and 'TermNoProgressBar' to disable the use of Term::ProgressBar.
+
 B<--output FILE> When grabbing, write output to FILE rather than
 standard output.
 
@@ -89,6 +94,7 @@
 
 use XMLTV;
 use XMLTV::Ask;
+use XMLTV::ProgressBar;
 use XMLTV::Config_file;
 use XMLTV::Get_nice qw(get_nice);
 use XMLTV::Memoize;
@@ -101,9 +107,6 @@
 use Getopt::Long;
 use URI;
 
-# Use Term::ProgressBar if installed.
-use constant Have_bar => eval { require Term::ProgressBar; 1 };
-
 # The timezone for Sweden during wintertime.
 use constant LOCAL_TZ => "+0100";
 
@@ -361,6 +364,7 @@
 my $opt = { days => 5,
             offset => 0,
             "config-file" => undef,
+            gui => undef,
             configure => 0,
             help => 0,
             quiet => 0,
@@ -372,6 +376,7 @@
                       days=i
                       offset=i
                       config-file=s
+                      gui
                       configure
                       help|h
                       quiet
@@ -386,7 +391,7 @@
   print << 'EOH';
 tv_grab_se --help
 
-tv_grab_se [--config-file FILE] --configure
+tv_grab_se [--config-file FILE] --configure [--gui OPTION]
 
 tv_grab_se [--config-file FILE] [--output FILE] [--days N]
            [--offset N] [--quiet] [--debug]
@@ -396,6 +401,9 @@
   exit(1);
 }
 
+XMLTV::Ask::init($opt->{'gui'});
+XMLTV::ProgressBar::init($opt->{'gui'});
+
 # XMLTV::DST says that we should do this...
 Date_Init('TZ=UTC');
 
@@ -451,10 +459,10 @@
 my $date = increase_date( $today, $opt->{offset} );
 
 my $bar = undef;
-$bar = new Term::ProgressBar( {
+$bar = new XMLTV::ProgressBar( {
     name => 'downloading listings',
     count => $opt->{days} * @channel_list
-    }) if Have_bar && (not $opt->{quiet}) && (not $opt->{debug});
+    }) if (not $opt->{quiet}) && (not $opt->{debug});
 
 for( my $i=0; $i < $opt->{days}; $i++ )
 {
@@ -469,7 +477,7 @@
 
     $date = increase_date( $date, 1 );
 }
-
+$bar->finish();
 $w->end();
 
 # Signal that something went wrong if there were warnings.
@@ -582,7 +590,7 @@
     # Ask about Swedish channels first.
     my @all = (grep( /\.se$/, @chan ), grep( !/\.se$/, @chan ));
 
-    my @wanted = askManyBooleanQuestions(1,
+    my @wanted = ask_many_boolean(1,
                     map { "get channel $channels{$_}->[0] ($_)?" }
                     @all );
     foreach (@all) {
Index: grab/uk_rt/tv_grab_uk_rt.in
===================================================================
RCS file: /cvsroot/xmltv/xmltv/grab/uk_rt/tv_grab_uk_rt.in,v
retrieving revision 1.65
diff -u -B -b -w -r1.65 tv_grab_uk_rt.in
--- grab/uk_rt/tv_grab_uk_rt.in	23 May 2004 16:23:15 -0000	1.65
+++ grab/uk_rt/tv_grab_uk_rt.in	5 Jun 2004 14:52:08 -0000
@@ -11,7 +11,7 @@
 
 tv_grab_uk_rt --help
 
-tv_grab_uk_rt [--config-file FILE] --configure
+tv_grab_uk_rt [--config-file FILE] --configure [--gui OPTION]
 
 tv_grab_uk_rt [--config-file FILE] [--output FILE] [--quiet]
               [--days N] [--offset N]
@@ -36,6 +36,11 @@
 default is B<~/.xmltv/tv_grab_uk_rt.conf>.  This is the file written by
 B<--configure> and read when grabbing.
 
+B<--gui OPTION> Use this option to enable a graphical interface to be used.
+OPTION may be 'Tk' or 'GDialog', or left blank for the best available choice.
+Additional allowed values of OPTION are 'Term' for normal terminal output
+(default) and 'TermNoProgressBar' to disable the use of Term::ProgressBar.
+
 B<--output FILE> When grabbing, write output to FILE rather than
 standard output.
 
@@ -95,6 +100,7 @@
 require HTML::Entities;
 use XMLTV;
 use XMLTV::Memoize;
+use XMLTV::ProgressBar;
 use XMLTV::Ask;
 use XMLTV::DST;
 use XMLTV::Config_file;
@@ -184,6 +190,7 @@
     $opt_share,
     $opt_configure,
     $opt_config_file,
+    $opt_gui,
     $opt_offset,
     $opt_quiet,
     $opt_slow,
@@ -200,6 +207,7 @@
 	   'help'          => \$opt_help,
 	   'configure'     => \$opt_configure,
 	   'config-file=s' => \$opt_config_file,
+           'gui:s'         => \$opt_gui,
 	   'output=s'      => \$opt_output,
 	   'share=s'       => \$opt_share, # also undocumented
            'offset=i'      => \$opt_offset,
@@ -217,6 +225,9 @@
 die "--limit-details makes no sense without --slow\n"
   if defined $opt_detailtimerange and not $opt_slow;
 
+XMLTV::Ask::init($opt_gui);
+XMLTV::ProgressBar::init($opt_gui);
+
 # Date::Manip has a bug where 'now' will be wrong if you change the
 # timezone.  It won't be correctly converted from the system timezone
 # to the new one.  So we call parse_date('today midnight') _before_
@@ -404,8 +415,11 @@
 	}
     }
     my $days = @dates_to_get > 1 ? 'days' : 'day';
-    say('getting ' . (scalar @dates_to_get) . " $days of listings\n")
+    
+    my $bar = new XMLTV::ProgressBar('getting ' . (scalar @dates_to_get)
+        . " $days of listings", 1)
       unless $opt_quiet;
+    
     t 'getting dates:' . d \@dates_to_get;
 
     $writer->start({ 'source-info-url'     => "$BASE_URL/",
@@ -669,6 +683,9 @@
 	$writer->write_programme($_);
     }
     $writer->end();
+    
+    $bar->finish();
+    
 }
 
 
@@ -1812,10 +1829,12 @@
 
     XMLTV::Config_file::check_no_overwrite($config_file);
 
-    # FIXME turn into progress bar.
-    print STDERR "finding channels:\t";
+    my $bar = new XMLTV::ProgressBar('finding channels', 1)
+	  if not $opt_quiet;
+    
     my %channels = get_channels();
-    print STDERR "got ". (scalar keys %channels) . " done.\n" unless $opt_quiet;
+    
+    $bar->finish();
 
     # FIXME need to make directory
     open(CONF, ">$config_file") or die "cannot write to $config_file: $!";
@@ -1869,7 +1888,7 @@
 	}
 	elsif (@poss == 1) {
 	    my $ch = $poss[0];
-	    if (askBooleanQuestion('Add channel ' .
+	    if (ask_boolean('Add channel ' .
 				   $ch->{'display-name'}->[0]->[0] . '?', 1)) {
 		my $xmltv_id = $ch->{id};
 		unless ($chose_ch{$xmltv_id}++) {
@@ -1887,7 +1906,7 @@
 	    }
 	    my $none_option = 'None of the above are what I wanted';
 	    die 'silly channel name' if exists $dn_to_ch{$none_option};
-	    my $r = askQuestion('Which channel to add?',
+	    my $r = ask_choice('Which channel to add?',
 				$poss[0]->{'display-name'}->[0]->[0],
 				(sort keys %dn_to_ch), $none_option);
 	    # treat EOF as same as none-option
Index: lib/Ask.pm
===================================================================
RCS file: /cvsroot/xmltv/xmltv/lib/Ask.pm,v
retrieving revision 1.12
diff -u -B -b -w -r1.12 Ask.pm
--- lib/Ask.pm	3 Jan 2004 14:52:53 -0000	1.12
+++ lib/Ask.pm	5 Jun 2004 14:52:08 -0000
@@ -7,9 +7,20 @@
 
 package XMLTV::Ask;
 use strict;
-use Carp qw(croak carp);
+use XMLTV::GUI;
 
-# Use Log::TraceMessages if installed, and choose graphical or not.
+use vars qw(@ISA @EXPORT);
+use Exporter;
+@ISA = qw(Exporter);
+@EXPORT = qw(ask
+             ask_password
+             ask_choice
+             ask_boolean
+             ask_many_boolean
+             say
+             );
+
+# Use Log::TraceMessages if installed.
 BEGIN {
     eval { require Log::TraceMessages };
     if ($@) {
@@ -20,22 +31,38 @@
 	*t = \&Log::TraceMessages::t;
 	*d = \&Log::TraceMessages::d;
     }
-
-    # For now we do graphical configuration only if the undocumented
-    # XMLTV_TK environment variable is set to a true value.
-    #
-    if ($ENV{XMLTV_TK}
-	and (defined($ENV{DISPLAY}) || $^O eq 'MSWin32')
-	and eval { require Tk }) {
-	require XMLTV::AskTk; XMLTV::AskTk->import;
-	*XMLTV::Ask:: = *XMLTV::AskTk::;
-    }
-    else {
-	require XMLTV::AskTerm; XMLTV::AskTerm->import;
-	*XMLTV::Ask:: = *XMLTV::AskTerm::;
     }
+
+my $real_class = 'XMLTV::Ask::Term';
+
+sub AUTOLOAD {
+        use vars qw($AUTOLOAD);
+        (my $method_name = $AUTOLOAD) =~ s/.*::(.*?)/$1/;
+        (my $real_class_path = $real_class.".pm") =~ s/::/\//g;
+        
+        require $real_class_path;
+        import $real_class_path;
+        
+        $real_class->$method_name(@_);       
 }
 
 
+# Must be called before we use this module if we want to use a gui.
+sub init( $ ) {
+        my $opt_gui = shift;
+        
+        # Ask the XMLTV::GUI module for the graphics type we will use  
+        my $gui_type = XMLTV::GUI::get_gui_type($opt_gui);
+        
+        if($gui_type =~ /^term/) {        
+                $real_class = 'XMLTV::Ask::Term';
+        } elsif($gui_type eq 'tk') {
+                $real_class = 'XMLTV::Ask::Tk';
+        } elsif($gui_type eq 'gdialog') {
+                $real_class = 'XMLTV::Ask::GDialog';
+        } else {
+                die "Unknown gui type: '$gui_type'.";
+        }
+}
 
 1;
Index: lib/exe_opt.pl
===================================================================
RCS file: /cvsroot/xmltv/xmltv/lib/exe_opt.pl,v
retrieving revision 1.8
diff -u -B -b -w -r1.8 exe_opt.pl
--- lib/exe_opt.pl	14 May 2004 22:21:26 -0000	1.8
+++ lib/exe_opt.pl	5 Jun 2004 14:52:08 -0000
@@ -14,7 +14,10 @@
 #
 print '-nologo
 -force
--add="XMLTV::Ask::Term;XMLTV::ProgressBar::Term"
+-add="XMLTV::Ask::Term;XMLTV::Ask::Tk"
+-add="XMLTV::ProgressBar::Term;XMLTV::ProgressBar::Tk;XMLTV::ProgressBar::None;"
+-add="XMLTV::GUI"
+-add="Tk::ProgressBar"
 -bind=libexpat.dll[file=\perl\site\lib\auto\XML\Parser\Expat\libexpat.dll,extract]
 -bind=libxml2.dll[file=\perl\bin\libxml2.dll,extract]
 -trim="Convert::EBCDIC;DB_File;Encode;HASH;HTML::FromText;Text::Iconv;Unicode::Map8;v5;URI/urn::isbn.pm;URI/urn::oid.pm;PerlIO/gzip.pm;HTML::FormatText"
